---
title: "ChIPseq analysis"
author: "Johan"
date: "10/25/2019"
output: html_document

params:
  workingDir: /Users/johanreimegard/git/flyEmbryogenesis
  proSeqDir: results/PROseq
  chipSeqDir: results/chipSeq
  RscriptsDir: R_Analysis/Rscripts  
  dataDir: data/ChIPseq
  sampleInfo: information/ChIP_ATAC_sampleInfo.tab.txt
  count_K27ac: K27ac.featureCount.count.tab.txt 
  count_K27me3: K27me3.featureCount.count.tab.txt
  count_H3K4me1: H3K4me1.featureCount.count.tab.txt 
  count_zld: zld.featureCount.count.tab.txt
  Antibody: zld

  RfunctionsDirectory: /Users/johanreimegard/GoogleDrive/git/RNAmappingPipeline/R
  
  
---

```{r setup, include=FALSE}

source(paste(params$RfunctionsDirectory,"ExpressionAnalysisFunctions.R",sep = "/"))
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)

library(gplots)

library(knitr)
#install.packages("kableExtra")
library(kableExtra)


gtfFile = paste(params$workingDir, "annotations/Drosophila_melanogaster.BDGP6.28.99.gtf", sep = "/")





trancriptData.QC.pauseIndex.knowGenes = paste(params$workingDir, "annotations/Drosophila_melanogaster.BDGP6.28.99.gtf", sep = "/")

```

```{r PROseq annotation}



PROseqInfo = read.table( file = paste( params$workingDir, params$proSeqDir, 
                          "selectedGenes.tsv" ,
                          sep = "/"),
                         quote = "", sep = "\t", 
                         header = TRUE, stringsAsFactors = F) 
annotation2 =  PROseqInfo %>% filter(!phenotype %in% c("Early1","Early2","removed"))
annotation2$pheno2 = "Gd7"
annotation2$pheno2[grep(pattern = "Toll9",  annotation2$phenotype)] = "Toll9"
annotation2$pheno2[grep(pattern = "Toll10b",  annotation2$phenotype)] = "Toll10b"

annotation2$direction = "Down"
annotation2$direction[grep(pattern = "Up",  annotation2$phenotype)] = "Up"



# ChIPseq analysis
```


## a
Add gene to transcript correlation

```{r  add  transcript to gene Information}

# Load in the fastq file that is generated as data from running multiqc on all samples. 
gtfFile = paste(params$workingDir, "annotations/Drosophila_melanogaster.BDGP6.28.99.gtf", sep = "/")
gtfInfo = read.table(file = gtfFile, header = F, sep = "\t", quote = "", stringsAsFactors = F)

gtfInfomRNA = gtfInfo %>% dplyr::filter(V3 == "transcript") 
gtfInfomRNA = gtfInfomRNA[grep(pattern = "gene_biotype \"protein_coding\"",x = gtfInfomRNA$V9 ),  ] 

gtfInfomRNA_DF= gtfInfomRNA %>% separate(col = V9,sep = ";",into =  c("geneID1", "transcriptID1"))
gtfInfomRNA_DF= gtfInfomRNA_DF %>% separate(col = geneID1,sep = "\"",into =  c("irrelevant", "geneID"))
gtfInfomRNA_DF= gtfInfomRNA_DF %>% separate(col = transcriptID1,sep = "\"",into =  c("irrelevant1", "transcriptID"))

gene2transcriptInfo  = gtfInfomRNA_DF %>% dplyr::select(geneID,transcriptID)



```


## Reads mapping


## Comparing mapping pattern between the samples

### Using feature count to get the difference between the samples

```{r Adding the first analysis of the data}
# install RColorBrewer if missing

library(RColorBrewer)
library(gplots)
library(GenomicFeatures)


featureCountFile = paste(params$workingDir, params$dataDir,
                         paste(params$Antibody , "featureCount.count.tab.txt", sep = "."), sep = "/")

FCdata = read.table( file = featureCountFile, header = TRUE, 
                     sep = "\t", quote = "",  
                     stringsAsFactors = FALSE)

peakInfo  = FCdata [,1:6]
countData = FCdata[,-1:-6]
rownames(countData) = peakInfo$Geneid

```



```{r handle the samples}

sampleInfoFile =paste(params$workingDir, params$sampleInfo, sep = "/")
sampleInfo = read.table( file = sampleInfoFile, header = TRUE, 
                     sep = "\t", quote = "",  
                     stringsAsFactors = FALSE)


sampleInfo$bamFiles2  =  gsub(pattern = "/",replacement = ".",x = sampleInfo$bamFiles)
sampleInfo$bamFiles2  =  gsub(pattern = "-",replacement = ".",x = sampleInfo$bamFiles2)

sampleInfo =sampleInfo %>% filter(Antibody == params$Antibody) 

sampleInfoCountTable = data.frame(bamFiles2 = colnames(countData))


sampleInfo = inner_join(sampleInfo,sampleInfoCountTable)

countData = countData[,sampleInfo$bamFiles2]
colnames(countData) = paste(sampleInfo$Genotype, sampleInfo$Replicate, sep = "_")

```


```{r add annotation}

txdb <- makeTxDbFromGFF(file=gtfFile, dataSource="ensemblgenomes",format = "gtf")

TR <- transcripts(txdb)

PR <- promoters(txdb, upstream=2000, downstream=400)
peakInfo$Chr = gsub(pattern = "chr",replacement = "", peakInfo$Chr)

chipPeakData = makeGRangesFromDataFrame(peakInfo,
                                      seqnames.field = colnames(peakInfo[2]), 
                                      start.field = colnames(peakInfo[3]), 
                                      end.field = colnames(peakInfo[4]), 
                                      strand.field = colnames(peakInfo[5]))



overlapPromoter = findOverlaps(PR,chipPeakData)
OP = data.frame(overlapPromoter)
PR.df = data.frame(PR)
PR.df$queryHits = 1:nrow(PR.df)
peakInfo$subjectHits = 1:nrow(peakInfo)


peakInfo = inner_join(peakInfo, OP)
peakInfo = inner_join(peakInfo, PR.df)

promoterRegions = unique(peakInfo$Geneid)

countData.expressed = countData[promoterRegions, ]



```





## Filtering





```{r differential gene expression analysis}
library(DESeq2)

exp.data = countData.expressed
metaInfo = sampleInfo

rownames(metaInfo) = paste(sampleInfo$Genotype, sampleInfo$Replicate, sep = "_")

metaInfo$Antibody = as.factor(metaInfo$Antibody)
metaInfo$Genotype = as.factor(metaInfo$Genotype)

dds <- DESeqDataSetFromMatrix(countData = exp.data,
                               colData = metaInfo,
                               design = ~Genotype )


#Remove rows with low counts and normalise samples for visualisation
dds <- dds[ rowSums(counts(dds)) > 100, ]

dim(dds)
#Normalizing and change to 2 fold log changes. 
rld <- rlog(dds)


#' ### Visualising the data


#+ save1, include=FALSE
#png(paste(params$workingDirectory, params$resultsDirectory,"SampleDistance.png", sep = "/"))
#plotSample2SampleDistance(assay(rld))
#dev.off()

#' 
#'

pdf(paste( params$Antibody,"distance_analysis.pdf", sep ="_"))
plotSample2SampleDistance(assay(rld))
dev.off()
#'  __Figure 1 Plotting sample to sample distance__ . 
#'  0 means that they are identical and 1 means that they are totally different.  
#'  The darker the blue the more similair. Also dendogram shows how similair they are. 
#' 
#' Samples do not cluster according to pre an post op. Most likely more due to difference in mapping (technical problem)
#' 
#' #### PCA analysis
#' Running PCA on the samples and plotting the different variables to see which of the parameterrs that fit the different components the best.
#' 
#' First checking how much the different PC contribute. 


mir.pca <- prcomp(t(assay(rld)), center = TRUE, scale = FALSE) 
e.var = (mir.pca[['sdev']]^2 / sum(mir.pca[['sdev']]^2))
e.var = as.data.frame( e.var )
e.var$PC = as.factor(1:nrow(e.var)) 
e.var$Variance = e.var$e.var*100
e.var$CumulativeVariance = cumsum(e.var$Variance)
qplot(PC, Variance, data=e.var, geom = c("point")
      , ylim = c(0,max(e.var$Variance)+5) 
      ,ylab = "Variance (%)")

#'  __Figure 2 Plotting PCA variance__.  
#'  Displays how much each of the PC contributes to the overall expression.    
#'  This suggest that the two first PC explains most of the variation.
#'
#+ save2, include=FALSE
#ggsave(paste(params$workingDirectory,params$resultsDirectory, "PCAvariance.png", sep = "/"))
#'
#'
pctable = as.data.frame(mir.pca$x)[, 1:3]
pctable = cbind(pctable, metaInfo)


test  = plotPCAplot(PCAinfo = pctable, n.comp = 3,
            varianceInfo = e.var, 
            colorComponent = "Genotype"
)

test
ggsave(paste( params$Antibody,"PCA_analysis.pdf", sep ="_"))



```




```{r loadings analysis}


loadings =mir.pca$rotation[,1:3]

loadings1 = as.data.frame(t((t(loadings)-colMeans(loadings))/colSds(loadings)))
loadings1$Geneid = rownames(loadings)

loadingsDF = gather(loadings1, key = PC, value = Loading, -Geneid )

loadingsDF = inner_join(loadingsDF, peakInfo)

loadingsDF = loadingsDF %>% rename(transcriptID = tx_name )
loadingsDF = inner_join(loadingsDF, gene2transcriptInfo)
loadingsDF = loadingsDF %>% distinct()




loadingsDF = left_join(loadingsDF,annotation2)
loadingsDF$phenotype[is.na(loadingsDF$phenotype)] = "background"
loadingsDF$direction[is.na(loadingsDF$direction)] = "None"
loadingsDF$pheno2[is.na(loadingsDF$pheno2)] = "Background"



ggplot(loadingsDF , mapping = aes(x = Loading, color = pheno2, linetype = direction))+ geom_density()+ facet_grid(pheno2~PC)
loadingsDF = loadingsDF %>% distinct()


loadingsDF2= loadingsDF %>% spread(key = PC, value = Loading )

ggplot(loadingsDF2 , mapping = aes(x = PC1, y = PC2, color = direction))+ geom_point()+ facet_wrap(pheno2~.)



pctable$test = "PC"
ggplot(loadingsDF2 , mapping = aes(x = PC1, y = PC2, color = direction))+ geom_point()+ facet_wrap(pheno2~.)+geom_point(data = pctable,mapping =  aes(x = PC1, y = PC2,fill=Genotype), shape=23, color="darkred", size=3)+ggtitle(label = paste("ChIPseq data ", params$Antibody, sep = " : "))



```
```{r join the data f}
