---
title: "PROseq analysis"
author: "Johan"
date: "10/25/2019"
output: html_document

params:
  workingDir: /Users/johanreimegard/git/flyEmbryogenesis
  gtfFile: R_Analysis/Rscripts  
  gtfFilePRO: data/PROseq
  RfunctionsDirectory: /Users/johanreimegard/GoogleDrive/git/RNAmappingPipeline/R
  
  
---

```{r setup, include=FALSE}

#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")

#BiocManager::install("Rsubread")
install.packages("ggsci")
cbPalette <- c( "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7","#999999")


library(Rsubread)
library(tidyverse)



```

# PROseq analysis


## Reads mapping



## Reads mapping as table
```{r  add fastqc}

# Load in the fastq file that is generated as data from running multiqc on all samples. 
gtfFile = paste(params$workingDir, "annotations/Drosophila_melanogaster.BDGP6.28.99.gtf", sep = "/")
gtfInfo = read.table(file = gtfFile, header = F, sep = "\t", quote = "", stringsAsFactors = F)


head(gtfInfo)


gtfInfoGenes = gtfInfo %>% dplyr::filter(V3 == "gene")


gtfInfomRNA = gtfInfo %>% dplyr::filter(V3 == "transcript") 
gtfInfomRNA = gtfInfomRNA[grep(pattern = "gene_biotype \"protein_coding\"",x = gtfInfomRNA$V9 ),  ] 

head(gtfInfomRNA)

dim(gtfInfomRNA)

test= gtfInfomRNA %>% separate(col = V9,sep = ";",into =  c("geneID1", "transcriptID1"))
test= test %>% separate(col = geneID1,sep = "\"",into =  c("irrelevant", "geneID"))
test= test %>% separate(col = transcriptID1,sep = "\"",into =  c("irrelevant1", "transcriptID"))

gene2transcriptInfo  = test %>% dplyr::select(geneID,transcriptID)




gtfInfomRNAtrimmed = gtfInfomRNA %>% distinct(V4,V5,V7, .keep_all = TRUE)
gtfInfomRNAtrimmed = gtfInfomRNAtrimmed %>% filter((V5 - V4) > 1000)



gtfInfomRNA = gtfInfomRNAtrimmed
#getPPC correct

head(gtfInfomRNA)

ppc_upstream = 50
ppc_downstream = 100

geneBody_dsTSS = 500
geneBody_usTES = 100

#Plus strand
gtfInfomRNAplus = gtfInfomRNA[gtfInfomRNA$V7 == "+",]
#Fix pauseindex
gtfInfomRNAplus$ppcLeft = gtfInfomRNAplus$V4 - ppc_upstream
gtfInfomRNAplus$ppcRight = gtfInfomRNAplus$V4 + ppc_downstream

#Fix geneBodyCoverage
gtfInfomRNAplus$gbLeft = gtfInfomRNAplus$V4 + geneBody_dsTSS
gtfInfomRNAplus$gbRight = gtfInfomRNAplus$V5 -  geneBody_usTES

gtfInfomRNAplus$left = gtfInfomRNAplus$ppcLeft
gtfInfomRNAplus$right = gtfInfomRNAplus$V5


#Negative strand
gtfInfomRNAminus = gtfInfomRNA[gtfInfomRNA$V7 == "-",]
#Fix pauseindex
gtfInfomRNAminus$ppcLeft = gtfInfomRNAminus$V5 - ppc_downstream 
gtfInfomRNAminus$ppcRight = gtfInfomRNAminus$V5 + ppc_upstream

#Fix geneBodyCoverage
gtfInfomRNAminus$gbLeft = gtfInfomRNAminus$V4 + geneBody_usTES
gtfInfomRNAminus$gbRight = gtfInfomRNAminus$V5 -  geneBody_dsTSS

#Fix left and right
gtfInfomRNAminus$left = gtfInfomRNAminus$V4
gtfInfomRNAminus$right = gtfInfomRNAminus$ppcRight


gtfInfomRNA2 = rbind(gtfInfomRNAminus,gtfInfomRNAplus)

head(gtfInfomRNA2)
gtfInfomRNA2.mRNA =  gtfInfomRNA2 %>% dplyr::select(V1,V2,V3,left,right,V6,V7,V8,V9)
colnames(gtfInfomRNA2.mRNA) = c("V1","V2","V3","left","right","V6","V7","V8","V9")
gtfInfomRNA2.ppc =  gtfInfomRNA2 %>% dplyr::select(V1,V2,V3,ppcLeft,ppcRight,V6,V7,V8,V9)
gtfInfomRNA2.ppc$V3 = "ppc"
colnames(gtfInfomRNA2.ppc) = c("V1","V2","V3","left","right","V6","V7","V8","V9")
gtfInfomRNA2.gbc =  gtfInfomRNA2 %>% dplyr::select(V1,V2,V3,gbLeft,gbRight,V6,V7,V8,V9)
gtfInfomRNA2.gbc$V3 = "gbc"
colnames(gtfInfomRNA2.gbc) = c("V1","V2","V3","left","right","V6","V7","V8","V9")


gtfInfomRNA3 = rbind(gtfInfomRNA2.mRNA,
                     gtfInfomRNA2.ppc,
                     gtfInfomRNA2.gbc)


head(gtfInfomRNA3)


gtfInfomRNA3 = gtfInfomRNA3 %>% arrange(V1,left)

min(gtfInfomRNA3$right - gtfInfomRNA3$left)

gtfFile_ProSeq = paste(params$workingDir, "annotations/Drosophila_melanogaster.BDGP6.28.99.PROseq.gtf", sep = "/")
write.table(x = gtfInfomRNA3, file =  gtfFile_ProSeq,row.names = FALSE, col.names = FALSE,
                      sep = "\t",quote = F)


head(gtfInfomRNA3, n = 100)
rbind()

# Use only some of the columns .
fastqc_info = fastqc %>% dplyr::select(Sample,avg_sequence_length,Total.Sequences )

# Spread the data from idxSummary agains so that each species gets its own column 
sampleInfo =   idxSummary %>% dplyr::select(species, Sample, TotalCount) %>% spread(key = species, value = TotalCount)

# Join idx stats and fastqc stats 
sampleInfo = inner_join(sampleInfo,fastqc_info) 

# Nr unmapped by subtraction
unmapped  = sampleInfo %>% dplyr::group_by(Sample) %>% dplyr::summarize(unmapped = Total.Sequences-Drosophila-Mouse)
sampleInfo = inner_join(sampleInfo,unmapped)


# Plot the data in a html format table. 


sampleInfo %>% dplyr::select(Sample,avg_sequence_length,Total.Sequences,Drosophila,Mouse,unmapped) %>%
  kable() %>%
  kable_styling()


```


## Comparing mapping pattern between the samples

### Using feature count to get the difference between the samples

```{r Adding the first analysis of the data}
# install RColorBrewer if missing
if (!require("RColorBrewer")) {
    install.packages("RColorBrewer")
}
library(RColorBrewer)
library(gplots)
featureCountFile = paste(params$workingDir, params$dataDir,"counts.CDS.txt", sep = "/")

FCdata = read.table( file = featureCountFile, header = TRUE, 
                     sep = "\t", quote = "",  
                     stringsAsFactors = FALSE)

geneInfo  = FCdata [,1:6]
geneInfo  = geneInfo[,c(1,6)]
countData = FCdata[,-1:-6]
rownames(countData) = geneInfo$Geneid
```


```{r handle the samples}

sampleInfoOld = sampleInfo
sampleInfo = data.frame(bamFile =  colnames(countData))
sampleInfo= sampleInfo %>% separate(col = bamFile,sep = "\\.",into =  c("dir", "mutant","sampleInfo"), remove = FALSE)
sampleInfo = sampleInfo %>% separate(col = sampleInfo,sep = "_",into =  c("irrelevant", "time","replicate"))

sampleInfo = sampleInfo %>% separate(col = mutant,sep = "_",into =  c("mutant"))

sampleInfo = sampleInfo %>% dplyr::select(bamFile,mutant,time,replicate)
sampleInfo = sampleInfo %>% unite("sampleName", mutant:time:replicate, remove = FALSE )
sampleInfo = sampleInfo %>% dplyr::select(bamFile,sampleName,mutant,time,replicate)


rownames(sampleInfo) = sampleInfo$bamFile
sampleInfo = sampleInfo[colnames(countData), ]
colnames(countData) = sampleInfo$sampleName
rownames(sampleInfo) = sampleInfo$sampleName

sampleInfo = sampleInfo %>% dplyr::select(sampleName,mutant,time,replicate,bamFile)



```


```{r first qc analysis}

geneExpression = log(rowSums(countData))

hist(geneExpression)

geneInfo$expression  =log(rowSums(countData))

geneInfo.QC = geneInfo %>% filter(expression > 1) 
countData.QC =  countData[geneInfo.QC$Geneid,]

distance = cor((countData.QC))

distance_matrix <- as.matrix(distance)
heatmap.2(distance_matrix, 
              col=brewer.pal(9,"Blues"), 
              density.info = "none",
              trace="none")


```


```{r script to check the known genes. }

knownGenes = read.table(file = paste(params$workingDir,"information/FlyBase_IDs.txt", sep = "/"),
                        sep = "\t", header = F, quote = "", stringsAsFactors = F)


colnames(knownGenes) = c("Submitted ID", "Current ID","Converted ID","Related record")

geneInfo.QC$knownGene = "No"
geneInfo.QC$knownGene[geneInfo.QC$Geneid %in% knownGenes$`Current ID`] = "Yes"

table(geneInfo.QC$knownGene)

countData.QC.knownTargets =  countData[geneInfo.QC$Geneid[geneInfo.QC$knownGene == "Yes"],]
distance = cor((countData.QC.knownTargets))

distance_matrix <- as.matrix(distance)
pdf(file = "PROseqData.HeatMap.knownGenes.pdf")
heatmap.2(distance_matrix, 
              col=brewer.pal(9,"Blues"), 
              margins = c(8, 8),
              density.info = "none",
              trace="none")

dev.off()


```


### sample swat
```{r change the two files that differ}
# Change filenames 
test = sampleInfo["toll10b_5h_rep1", ]
sampleInfo["toll10b_5h_rep1", ] = sampleInfo["gd7_5h_rep1", ]  
sampleInfo["gd7_5h_rep1", ] = test 
colnames(countData) = sampleInfo$sampleName
rownames(sampleInfo) = sampleInfo$sampleName
countData.QC =  countData[geneInfo.QC$Geneid,]

countData.QC.knownTargets =  countData[geneInfo.QC$Geneid[geneInfo.QC$knownGene == "Yes"],]
distance = cor((countData.QC.knownTargets))

distance_matrix <- as.matrix(distance)
pdf(file = "PROseqData.HeatMap.knownGenes.pdf")
heatmap.2(distance_matrix, 
              col=brewer.pal(9,"Blues"), 
              margins = c(8, 8),
              density.info = "none",
              trace="none")

dev.off()



```




```{r differential gene expression analysis}
library(DESeq2)

exp.data = countData.QC
metaInfo = sampleInfo

colnames(exp.data)
dim(metaInfo)
rownames(metaInfo)
rownames(exp.data)

metaInfo$mutant = as.factor(metaInfo$mutant)
metaInfo$time = as.factor(metaInfo$time)

dds <- DESeqDataSetFromMatrix(countData = exp.data,
                               colData = metaInfo,
                               design = ~time + mutant )


#Remove rows with low counts and normalise samples for visualisation
dds <- dds[ rowSums(counts(dds)) > 100, ]

dim(dds)
#Normalizing and change to 2 fold log changes. 
rld <- rlog(dds)


#' ### Visualising the data


#+ save1, include=FALSE
#png(paste(params$workingDirectory, params$resultsDirectory,"SampleDistance.png", sep = "/"))
#plotSample2SampleDistance(assay(rld))
#dev.off()

#' 
#'
plotSample2SampleDistance(assay(rld))
#'  __Figure 1 Plotting sample to sample distance__ . 
#'  0 means that they are identical and 1 means that they are totally different.  
#'  The darker the blue the more similair. Also dendogram shows how similair they are. 
#' 
#' Samples do not cluster according to pre an post op. Most likely more due to difference in mapping (technical problem)
#' 
#' #### PCA analysis
#' Running PCA on the samples and plotting the different variables to see which of the parameterrs that fit the different components the best.
#' 
#' First checking how much the different PC contribute. 


mir.pca <- prcomp(t(assay(rld)), center = TRUE, scale = FALSE) 
e.var = (mir.pca[['sdev']]^2 / sum(mir.pca[['sdev']]^2))
e.var = as.data.frame( e.var )
e.var$PC = as.factor(1:nrow(e.var)) 
e.var$Variance = e.var$e.var*100
e.var$CumulativeVariance = cumsum(e.var$Variance)
qplot(PC, Variance, data=e.var, geom = c("point")
      , ylim = c(0,max(e.var$Variance)+5) 
      ,ylab = "Variance (%)")

#'  __Figure 2 Plotting PCA variance__.  
#'  Displays how much each of the PC contributes to the overall expression.    
#'  This suggest that the two first PC explains most of the variation.
#'
#+ save2, include=FALSE
#ggsave(paste(params$workingDirectory,params$resultsDirectory, "PCAvariance.png", sep = "/"))
#'
#'
pctable = as.data.frame(mir.pca$x)[, 1:4]
pctable = cbind(pctable, metaInfo)


test  = plotPCAplot(PCAinfo = pctable, n.comp = 3,
            varianceInfo = e.var, 
            colorComponent = "mutant", 
            pchComponent = "time"
)



test

#ggsave("PCA_analysis.pdf")

ggplot(data = pctable, mapping = aes(x = PC1, y =  PC2, color = mutant, point = time)) + geom_point()
ggplot(data = pctable, mapping = aes(x = PC1, y =  PC2, color = mutant, point = time)) + geom_point()

loadings = mir.pca$rotation[,1:3]

colMeans(loadings)
colSds(loadings)

loadings1 = as.data.frame(t((t(loadings)-colMeans(loadings))/colSds(loadings)))

loadings1$geneID = rownames(loadings1)

loadingsDF = gather(loadings1, key = PC, value = zScore, -geneID )

ggplot(loadingsDF, mapping = aes(x = zScore, color = PC )) + geom_density()


```
 
 
 
```{r }

 
library(clusterProfiler)
loadingsDF = gather(loadings1, key = PC, value = zScore, -geneID )

universe <- bitr(unique(loadingsDF$geneID), fromType = "ENSEMBL",
        toType = c("ENTREZID", "SYMBOL"),
        OrgDb = org.Dm.eg.db)


colnames(loadingsDF) = c("ENSEMBL","PC","zScore")

loadingsDF = inner_join(loadingsDF, universe) 

kegg = bitr_kegg(geneID = unique(loadingsDF$ENTREZID), fromType = "ncbi-geneid",
        toType = c("kegg"),
        organism = 'dme')

colnames(kegg) = c("ENTREZID","KEGGID") 

loadingsDF = inner_join(loadingsDF,kegg)
head(loadingsDF)


# ggo <- groupGO(gene     = unique(gene.df$ENTREZID),
#               OrgDb    = org.Dm.eg.db,
#              ont      = "CC",
#               level    = 3,
#               readable = TRUE)



loadingsGD7 = loadingsDF %>% filter(PC == "PC2"  & zScore < -2)
loadingstoll10b = loadingsDF %>% filter(PC == "PC2"  & zScore > 2)
loadingstoll9 = loadingsDF %>% filter(PC == "PC3"  & zScore > 2)
loadings3h = loadingsDF %>% filter(PC == "PC1"  & zScore < -2)
loadings5h = loadingsDF %>% filter(PC == "PC1"  & zScore > 2)
loadingPC1 = loadingsDF %>% filter(PC == "PC1")
loadingPC2 = loadingsDF %>% filter(PC == "PC2")


DEGloadings = rbind(loadingsGD7,loadingstoll10b,loadingstoll9)

enrichGO2 <- function(geneNames, backgroundList, 
                      ontology = "BP", 
                      pvalueCutoff = 0.01,
                      qvalueCutoff = 0.05){
    enriched =  enrichGO(gene          = geneNames,
                universe      = backgroundList,
                OrgDb         = org.Dm.eg.db,
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = pvalueCutoff,
                qvalueCutoff  = qvalueCutoff,
        readable      = TRUE)
    return(enriched)
}


backgroundList = unique(loadingsDF$ENTREZID)

ego.gd7 = enrichGO2(unique(loadingsGD7$ENTREZID),backgroundList)
ego.toll10b = enrichGO2(unique(loadingstoll10b$ENTREZID),backgroundList)
ego.toll9 = enrichGO2(unique(loadingstoll9$ENTREZID),backgroundList)
ego.5h = enrichGO2(unique(loadings5h$ENTREZID),backgroundList)
ego.3h = enrichGO2(unique(loadings3h$ENTREZID),backgroundList)


ego.3h  = simplify(ego.3h)
ego.5h  = simplify(ego.5h)
ego.gd7  = simplify(ego.gd7)
ego.toll10b  = simplify(ego.toll10b)
ego.toll9  = simplify(ego.toll9)


library(enrichplot)
bp.gd7 = barplot(ego.gd7, showCategory=10)
bp.toll10b = barplot(ego.toll10b, showCategory=10)
bp.toll9 = barplot(ego.toll9, showCategory=10)
bp.3h = barplot(ego.3h, showCategory=10)
bp.5h = barplot(ego.5h, showCategory=10)

goterms.mutanst = cowplot::plot_grid(bp.gd7, bp.toll10b, bp.toll9, ncol=1, labels=c("gd7","toll10b", "toll9"))
cowplot::save_plot(filename = "GOterm.mutants.pdf", plot = goterms.mutanst , base_height = 7)
goterms.timePoints = cowplot::plot_grid(bp.3h, bp.5h, ncol=1, labels=c("3h","5h"))
cowplot::save_plot(filename = "GOterm.timePoints.pdf", plot = goterms.timePoints , base_height = 7)



cnet.5h <- cnetplot(ego.5h,circular = TRUE, colorEdge = TRUE)
cnet.3h <- cnetplot(ego.3h ,circular = TRUE, colorEdge = TRUE)
cnet.gd7 <- cnetplot(ego.gd7 ,circular = TRUE, colorEdge = TRUE)
cnet.toll9 <- cnetplot(ego.toll9 ,circular = TRUE, colorEdge = TRUE)
cnet.toll10b <- cnetplot(ego.toll10b ,circular = TRUE, colorEdge = TRUE)


goterms.gd7 = cowplot::plot_grid(bp.gd7, cnet.gd7, ncol=1,labels=LETTERS[1:2])
cowplot::save_plot(filename = "GOterm.gd7.pdf", plot = goterms.gd7 , base_height = 10)

goterms.toll9 = cowplot::plot_grid(bp.toll9, cnet.toll9, ncol=1,labels=LETTERS[1:2])
cowplot::save_plot(filename = "GOterm.toll9.pdf", plot = goterms.toll9 , base_height = 10)


goterms.toll10b = cowplot::plot_grid(bp.toll10b, cnet.toll10b, ncol=1,labels=LETTERS[1:2])
cowplot::save_plot(filename = "GOterm.toll10b.pdf", plot = goterms.toll10b , base_height = 10)


```



```{r DGE expression analysis for 3h}


metaInfo.3h  = metaInfo %>% filter(time == "3h")
colData.3h = exp.data[, metaInfo.3h$sampleName]

dds.3h <- DESeqDataSetFromMatrix(countData = colData.3h,
                               colData = metaInfo.3h,
                               design = ~mutant )


#countData["FBgn0003463", ]
#assay(rld)["FBgn0003463", ]
#loadingsDF %>% filter(ENSEMBL == "FBgn0003463")

#assay(rld)["FBgn0003463", ]

dds.3h <- DESeq(dds.3h)

rld <- rlog(dds.3h)


norm.count.3h.rlog = as.data.frame(assay(rld))
rownames(norm.count.3h.rlog) =  rownames(rld)

norm.count.3h.rlog$ENSEMBL = rownames(norm.count.3h.rlog)


#rds <- results(dds, contrast = list( c("mutant_toll10b_vs_gd7"),c("mutant_toll9_vs_gd7") ) )



toll9_vs_10.3h = getSigResult(dds.3h, contrast=c("mutant","toll9","toll10b"),
                           geneNameColumn = "ENSEMBL",
                           universe = universe, cutoff = 0.05)
toll9_vs_10.3h.lFC1 = toll9_vs_10 %>% filter(abs(log2FoldChange) > 1)  %>% dplyr::select(log2FoldChange,padj,ENSEMBL)
colnames(toll9_vs_10.3h.lFC1) = c("l2fc_toll9_vs_toll10b","padj_toll9_vs_toll10b","ENSEMBL")




# check specific gene
#toll9_vs_10.3h[toll9_vs_10.3h$ENSEMBL %in% c("FBgn0003463"), ]


toll9_vs_gd7.3h =getSigResult(dds.3h, contrast=c("mutant","toll9","gd7"),
                           geneNameColumn = "ENSEMBL",
                           universe = universe, cutoff = 0.05)
toll9_vs_gd7.3h.lFC1 = toll9_vs_gd7.3h %>% filter(abs(log2FoldChange) > 1)  %>% dplyr::select(log2FoldChange,padj,ENSEMBL)
colnames(toll9_vs_gd7.3h.lFC1) = c("l2fc_toll9_vs_gd7","padj_toll9_vs_gd7","ENSEMBL")


toll10_vs_gd7.3h =getSigResult(dds.3h, contrast=c("mutant","toll10b","gd7"),
                           geneNameColumn = "ENSEMBL",
                           universe = universe, cutoff = 0.05)
toll10_vs_gd7.3h.lFC1 = toll10_vs_gd7.3h %>% filter(abs(log2FoldChange) > 1) %>% dplyr::select(log2FoldChange,padj,ENSEMBL)
colnames(toll10_vs_gd7.3h.lFC1) = c("l2fc_toll10b_vs_gd7","padj_toll10b_vs_gd7","ENSEMBL")

DEGlist.3h= union(toll10_vs_gd7.3h.lFC1$ENSEMBL, toll9_vs_gd7.3h.lFC1$ENSEMBL )
DEGlist.3h = union(DEGlist.3h, toll9_vs_10.3h.lFC1$ENSEMBL)



norm.count.3h.rlog.DEG=  norm.count.3h.rlog[DEGlist.3h, ]
norm.count.3h.rlog.DEG = left_join(norm.count.3h.rlog.DEG, toll10_vs_gd7.3h.lFC1)
norm.count.3h.rlog.DEG = left_join(norm.count.3h.rlog.DEG, toll9_vs_gd7.3h.lFC1)
norm.count.3h.rlog.DEG = left_join(norm.count.3h.rlog.DEG, toll9_vs_10.3h.lFC1)
norm.count.3h.rlog.DEG = left_join(norm.count.3h.rlog.DEG, universe)


norm.count.3h.rlog.DEG = norm.count.3h.rlog.DEG[ , c(7,14,15,1,2,3,4,5,6,8,9,10,11,12,13)]

write.table(x = norm.count.3h.rlog.DEG, quote = FALSE, sep = "\t", col.names = TRUE, row.names = FALSE, 
            file =  paste(params$workingDir, params$dataDir,"normcounts.CDS.DE.3h.lfc1.txt", sep = "/"))
   



```


```{r DGE expression analysis for 5h}


metaInfo.5h  = metaInfo %>% filter(time == "5h")
colData.5h = exp.data[, metaInfo.5h$sampleName]

dim(colData.5h) 

dds.5h <- DESeqDataSetFromMatrix(countData = colData.5h,
                               colData = metaInfo.5h,
                               design = ~mutant )


#countData["FBgn0003463", ]
#assay(rld)["FBgn0003463", ]
#loadingsDF %>% filter(ENSEMBL == "FBgn0003463")

#assay(rld)["FBgn0003463", ]

dds.5h <- DESeq(dds.5h)

rld.5h <- rlog(dds.5h)


norm.count.5h.rlog = as.data.frame(assay(rld.5h))

norm.count.5h.rlog$ENSEMBL = rownames(norm.count.5h.rlog)


#rds <- results(dds, contrast = list( c("mutant_toll10b_vs_gd7"),c("mutant_toll9_vs_gd7") ) )




toll9_vs_10.5h = getSigResult(dds.5h, contrast=c("mutant","toll9","toll10b"),
                           geneNameColumn = "ENSEMBL",
                           universe = universe, cutoff = 0.05)
toll9_vs_10.5h.lFC1 = toll9_vs_10 %>% filter(abs(log2FoldChange) > 1)  %>% dplyr::select(log2FoldChange,padj,ENSEMBL)
colnames(toll9_vs_10.5h.lFC1) = c("l2fc_toll9_vs_toll10b","padj_toll9_vs_toll10b","ENSEMBL")



# check specific gene
#toll9_vs_10.3h[toll9_vs_10.3h$ENSEMBL %in% c("FBgn0003463"), ]


toll9_vs_gd7.5h =getSigResult(dds.5h, contrast=c("mutant","toll9","gd7"),
                           geneNameColumn = "ENSEMBL",
                           universe = universe, cutoff = 0.05)
toll9_vs_gd7.5h.lFC1 = toll9_vs_gd7.5h %>% filter(abs(log2FoldChange) > 1)  %>% dplyr::select(log2FoldChange,padj,ENSEMBL)
colnames(toll9_vs_gd7.5h.lFC1) = c("l2fc_toll9_vs_gd7","padj_toll9_vs_gd7","ENSEMBL")

toll10_vs_gd7.5h =getSigResult(dds.5h, contrast=c("mutant","toll10b","gd7"),
                           geneNameColumn = "ENSEMBL",
                           universe = universe, cutoff = 0.05)
toll10_vs_gd7.5h.lFC1 = toll10_vs_gd7.5h %>% filter(abs(log2FoldChange) > 1) %>% dplyr::select(log2FoldChange,padj,ENSEMBL)
colnames(toll10_vs_gd7.5h.lFC1) = c("l2fc_toll10b_vs_gd7","padj_toll10b_vs_gd7","ENSEMBL")



DEGlist.5h = union(toll10_vs_gd7.5h.lFC1$ENSEMBL, toll9_vs_gd7.5h.lFC1$ENSEMBL )
DEGlist.5h = union(DEGlist.5h, toll9_vs_10.5h.lFC1$ENSEMBL)


norm.count.5h.rlog.DEG=  norm.count.5h.rlog[DEGlist.5h, ]
norm.count.5h.rlog.DEG = left_join(norm.count.5h.rlog.DEG, toll10_vs_gd7.5h.lFC1)
norm.count.5h.rlog.DEG = left_join(norm.count.5h.rlog.DEG, toll9_vs_gd7.5h.lFC1)
norm.count.5h.rlog.DEG = left_join(norm.count.5h.rlog.DEG, toll9_vs_10.5h.lFC1)
norm.count.5h.rlog.DEG = left_join(norm.count.5h.rlog.DEG, universe)

colnames(norm.count.5h.rlog.DEG)[c(7,14,15,1,2,3,4,5,6,8,9,10,11,12,13)]
norm.count.5h.rlog.DEG = norm.count.5h.rlog.DEG[ , c(7,14,15,1,2,3,4,5,6,8,9,10,11,12,13)]

write.table(x = norm.count.5h.rlog.DEG, quote = FALSE, sep = "\t", col.names = TRUE, row.names = FALSE, 
            file =  paste(params$workingDir, params$dataDir,"normcounts.CDS.DE.5h.lfc1.txt", sep = "/"))
   
   
```



```{r compare the different timepoints}




intersect(DEGlist.5h,DEGlist.3h)




dim(toll9_vs_10.3h.lFC1)
dim(toll9_vs_gd7.3h.lFC1)
dim(toll10_vs_gd7.3h.lFC1)


dim(toll9_vs_gd7.5h.lFC1)
dim(toll9_vs_10.5h.lFC1)
dim(toll10_vs_gd7.5h.lFC1)


length(DEGlist.5h)
length(DEGlist.3h)
InCommonMutantGenes = intersect(DEGlist.5h,DEGlist.3h)

loadings = unique(DEGloadings$ENSEMBL)

intersect(InCommonMutantGenes,loadings )

genes = as.data.frame(table(c(DEGlist.5h,DEGlist.3h)))

intersect(genes$Var1[genes$Freq==2],loadings )
intersect(genes$Var1[genes$Freq==1],loadings )

intersect(knownGenes$`Current ID`, genes$Var1)
intersect(knownGenes$`Current ID`, loadings)

genes = as.data.frame(table(c(DEGlist.5h,DEGlist.3h)))



```

summary(dds.5h)

dds.3h

         







toll9 = intersect(toll9_vs_10.3h.lFC1$ENTREZID[toll9_vs_10.3h.lFC1$log2FoldChange>0],toll9_vs_gd7.3h.lFC1$ENTREZID[toll9_vs_gd7.3h.lFC1$log2FoldChange>0] )

toll10b = intersect(toll9_vs_10.3h.lFC1$ENTREZID[toll9_vs_10.3h.lFC1$log2FoldChange<0],
                    toll10_vs_gd7.3h.lFC1$ENTREZID[toll10_vs_gd7.3h.lFC1$log2FoldChange>0] )


toll10b.3h.lFC1.inersect = toll9_vs_10.3h.lFC1 %>% filter( ENTREZID %in% toll10b)
toll9_vs_10.3h.lFC1[toll9_vs_10.3h.lFC1$ENTREZID %in%  toll9, ]

ego.toll9.DE = enrichGO2(toll9_vs_10.3h.lFC1$ENTREZID[toll9_vs_10.3h.lFC1$log2FoldChange>1],backgroundList)
ego.toll9.DE.bp = barplot(ego.toll9.DE, showCategory=10)

ego.toll10b.DE = enrichGO2(toll10_vs_gd7$ENTREZID[toll10_vs_gd7$log2FoldChange>1],backgroundList)
ego.toll10b.DE.bp = barplot(ego.toll10b.DE, showCategory=10)

ego.gd7.DE = enrichGO2(toll10_vs_gd7$ENTREZID[toll10_vs_gd7$log2FoldChange<0],backgroundList)
ego.gd7.DE.bp = barplot(ego.gd7.DE, showCategory=10)




















```


```{r DGE expression analysis}
colData(dds)


design(dds) <- formula(~ time + mutant)
dds <- DESeq(dds)
#rds <- results(dds, contrast = list( c("mutant_toll10b_vs_gd7"),c("mutant_toll9_vs_gd7") ) )

getSigResult<- function(dds_function, contrast, geneNameColumn,universe, cutoff){
  DEresult = results(object = dds_function, 
                    contrast=contrast)

  DEresult[[geneNameColumn]] = rownames(DEresult)
  DEresult.Ordered <- DEresult[order(DEresult$pvalue),]
  DEresult.Ordered.sig <- subset(DEresult.Ordered, padj < cutoff)
  DEresult.Ordered.sig.df = inner_join(as.data.frame(DEresult.Ordered.sig),universe)
  
  
  
  return(DEresult.Ordered.sig.df)
}


toll9_vs_10 = getSigResult(dds, contrast=c("mutant","toll9","toll10b"),
                           geneNameColumn = "ENSEMBL",
                           universe = universe, cutoff = 0.05)

toll9_vs_10.lFC1 = toll9_vs_10 %>% filter(abs(log2FoldChange) > 1) 
toll9_vs_gd7 =getSigResult(dds, contrast=c("mutant","toll9","gd7"),
                           geneNameColumn = "ENSEMBL",
                           universe = universe, cutoff = 0.05)

toll10_vs_gd7 =getSigResult(dds, contrast=c("mutant","toll10b","gd7"),
                           geneNameColumn = "ENSEMBL",
                           universe = universe, cutoff = 0.05)


toll9 = intersect(toll9_vs_10$ENTREZID[toll9_vs_10$log2FoldChange>0],toll9_vs_gd7$ENTREZID[toll9_vs_gd7$log2FoldChange>0] )

ego.toll9.DE = enrichGO2(toll9_vs_gd7$ENTREZID[toll9_vs_gd7$log2FoldChange>1],backgroundList)
ego.toll9.DE.bp = barplot(ego.toll9.DE, showCategory=10)

ego.toll10b.DE = enrichGO2(toll10_vs_gd7$ENTREZID[toll10_vs_gd7$log2FoldChange>1],backgroundList)
ego.toll10b.DE.bp = barplot(ego.toll10b.DE, showCategory=10)

ego.gd7.DE = enrichGO2(toll10_vs_gd7$ENTREZID[toll10_vs_gd7$log2FoldChange<0],backgroundList)
ego.gd7.DE.bp = barplot(ego.gd7.DE, showCategory=10)




```

```{r enrichment analysis of diff exp genes. }

install.packages("enrichR")
library(enrichR)
dbs <- listEnrichrDbs()



enrichGO2 <- function(DEseqReult, DEcutoff = 0.05, 
                      ontology = "BP", 
                      pvalueCutoff = 0.01,
                      qvalueCutoff = 0.05){
    geneList = toll10_vs_gd7$log2FoldChange
    names(geneList) = toll10_vs_gd7$ENTREZID
    geneList = geneList[order(geneList, decreasing = TRUE)]
    
    ego3 <- gseGO(geneList     = geneList,
              OrgDb        = org.Dm.eg.db,
              ont          = "BP",
              nPerm        = 1000,
              minGSSize    = 20,
              maxGSSize    = 500,
              pvalueCutoff = 0.05,
              verbose      = FALSE)
    ego3 <- setReadable(ego3, OrgDb = org.Dm.eg.db)
 cnetplot(ego3, categorySize="pvalue", foldChange=geneList,  circular = TRUE, colorEdge = TRUE)   
    
    enriched =  enrichGO(gene          = geneNames,
                universe      = backgroundList,
                OrgDb         = org.Dm.eg.db,
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = pvalueCutoff,
                qvalueCutoff  = qvalueCutoff,
        readable      = TRUE)
    return(enriched)
}


```{shell analysis for  }


Files that will be analysed are 
./bam_dmel6/gd7_2-5_3h_rep1.dmel6.bowtie2.sorted.bam and ./bam_dmel6/gd7_2-5_3h_rep2.dmel6.bowtie2.sorted.bam 
vs 
./bam_dmel6/gd7_4-5_5h_rep1.dmel6.bowtie2.sorted.bam and ./bam_dmel6/gd7_4-5_5h_rep2.dmel6.bowtie2.sorted.bam 

in /proj/snic2019-30-14/nobackup/PRO_seq_dm6

Result files will be found in the /proj/snic2019-30-14/nobackup/PRO_seq_dm6/pause_out/gd7_time/ directory

Regions analysed are chr2R chr2L chr3L chr3R chr4 chrX chrY

Startup analysis
 ok   Cwd 
 ok   File::Basename 
 ok   File::Path 
 ok   FindBin 
 ok   Getopt::Long 
 ok   Statistics::Basic 
 ok   Statistics::R
 
 wdir
/proj/snic2019-30-14/nobackup/PRO_seq_dm6


File ./bam_dmel6/gd7_2-5_3h_rep1.dmel6.bowtie2.sorted.spikeRemoved.bam exists.
File ./bam_dmel6/gd7_2-5_3h_rep2.dmel6.bowtie2.sorted.spikeRemoved.bam exists.
File ./bam_dmel6/gd7_4-5_5h_rep1.dmel6.bowtie2.sorted.spikeRemoved.bam exists.
File ./bam_dmel6/gd7_4-5_5h_rep2.dmel6.bowtie2.sorted.spikeRemoved.bam exists.


Start the program

pause_PROseq.pl -o pause_out/gd7_time/ -in1 ./bam_dmel6/gd7_2-5_3h_rep1.dmel6.bowtie2.sorted.spikeRemoved.bam ./bam_dmel6/gd7_2-5_3h_rep2.dmel6.bowtie2.sorted.spikeRemoved.bam -in2 ./bam_dmel6/gd7_4-5_5h_rep1.dmel6.bowtie2.sorted.spikeRemoved.bam ./bam_dmel6/gd7_4-5_5h_rep2.dmel6.bowtie2.sorted.spikeRemoved.bam -m dm6

```

#### NRSA output


```{}
finished reading gtf!
pausing analysis ......
#now working on chrY-#!
working on geneFBtr0302336
.
.
138893 similair lines
.
.
working on geneFBtr0346248
working on geneFBtr0078795
working on geneFBtr0306101
working on geneFBtr0309663
working on geneFBtr0330121
working on geneFBtr0084151
Program exist with Error 

```

#### NRSA error output

```{ }
system R failed: 256 at /proj/snic2019-30-14/private/perl/NRSA-v2/bin/pause_PROseq.pl line 426.

```