---
title: "ChIPseq H3K27Ac analysis"
author: "Jenny Jakobsson"
date: "11/11/2020"
output: html_document

params:
  workingDir: ~/Applied_Bioinformatics_Group_4
  dataDir: data_johan/data/ChIPseq/H3K27ac
  sampleInfo: data_johan/information/cleaned/ChIP_ATAC_sampleInfo.tab.txt
  chipSeqData: K27ac.featureCount.count.tab.txt
  RscriptsDir: script/original_scripts
  resultsDir: results/original_method/ChIPseq
  Antibody: K27ac
  ChIPresultsDir: results/original_method/ChIPseq/K27ac
  ChIPSeq.data: ChIPseq.H3K27Ac.data.tsv
---
                             
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# Install required packages
#install.packages("kableExtra")
#install.packages("tidyverse")
#install.packages("RColorBrewer")
#BiocManager::install("DESeq2")
#BiocManager::install("edgeR")

# Include the functions from the expression analysis script
source(paste(params$workingDir, params$RscriptsDir, "ExpressionAnalysisFunctions.r",sep = "/"))

# Load the installed packages
library(tidyverse)
library(gplots)
library(knitr)
library(kableExtra)
library("viridis")
library(RColorBrewer)
library(DESeq2)
library(edgeR)

```

## H3K27ac

### Using feature count to get the counts

```{r Adding the first analysis of the data, include=FALSE}
#Load in the path to the H3K27ac feature count file
featureCountFile = paste(params$workingDir, params$dataDir, params$chipSeqData, sep = "/")

#Load the data into a table
FCdata = read.table( file = featureCountFile, header = TRUE, 
                     sep = "\t", quote = "",  
                     stringsAsFactors = FALSE)

#Divide into a countdata table and peakinfo table
peakInfoLocation  = FCdata [,1:6] %>% dplyr::rename(geneID = Geneid)
countData = FCdata[,-1:-6]
rownames(countData) = peakInfoLocation$geneID

#Load the ChIP ATAC sampleInfo file that contains info about the samples. 
sampleInfoFile =paste(params$workingDir, params$sampleInfo, sep = "/")
sampleInfo = read.table( file = sampleInfoFile, header = TRUE, 
                     sep = "\t", quote = "",  
                     stringsAsFactors = FALSE)

#Add a column with better names of the bam files
sampleInfo$bamFiles2  =  gsub(pattern = "/",replacement = ".",x = sampleInfo$bamFiles)
sampleInfo$bamFiles2  =  gsub(pattern = "-",replacement = ".",x = sampleInfo$bamFiles2)

#Modify names of the column Genotype
sampleInfo$Genotype = recode(sampleInfo$Genotype, Toll910 = "Toll9")
sampleInfo$Genotype = recode(sampleInfo$Genotype, gd7 = "Gd7")

#Pick out only the samples from K27ac
sampleInfo =sampleInfo %>% filter(Antibody == params$Antibody) 

#Create a dataframe with a row for each column in countData
sampleInfoCountTable = data.frame(bamFiles2 = colnames(countData))

#Join sampleInfo and sampleInfoCountTable
sampleInfo = inner_join(sampleInfo,sampleInfoCountTable)

#Pick the samples that are in sampleInfo
countData = countData[,sampleInfo$bamFiles2]
#Paste genotype and replicate name together
#Add as column name
colnames(countData) = paste(sampleInfo$Genotype, sampleInfo$Replicate, sep = "_")
#Add as a new column
sampleInfo$sampleID = paste(sampleInfo$Genotype, sampleInfo$Replicate, sep = "_")

```



## Differential expression analysis and filtering

The data is analyzed with deseq2 to normalize the counts and do a differential expression analysis. The differential expression is investigated through the log2 fold change. Genes with total counts less then 100 were filtered out. Then all peaks without at least two samples with rlog > 5. 

```{r differential gene expression analysis, message=FALSE}
#Create expression and meta data. 
exp.data = countData
metaInfo = sampleInfo

rownames(metaInfo) = sampleInfo$sampleID

#Create a factor of genotype and sampleID values
metaInfo$Genotype = as.factor(metaInfo$Genotype)
metaInfo$sampleID = as.factor(metaInfo$sampleID)

#Create a DESeq data set from exp.data and metaInfo
dds <- DESeqDataSetFromMatrix(countData = exp.data,
                               colData = metaInfo, ~1)

#Remove rows with low counts and normalize samples for visualization
dds <- dds[ rowSums(counts(dds)) > 100, ]

#Transform the count data in the DESeq data frame to the log2 scale which minimizes differences between samples for rows with small counts, and which normalizes with respect to library size
rld <- rlog(dds)

#Put this normalized data into a data frame
normExpression = as.data.frame(assay(rld))

#Create a normalized expression data frame
normExpression$geneID = rownames(normExpression)
normExpressionDF  = normExpression %>% gather( key = sampleID, value = rlog, -geneID)
normExpressionDF = inner_join(normExpressionDF, metaInfo)

#Plot normExpressionDF
ggplot(normExpressionDF, aes(x = rlog)) + geom_density()

peakInfo = normExpressionDF %>% dplyr::select(geneID, rlog, sampleID) %>% 
  dplyr::group_by (geneID) %>%
  dplyr::summarize ( min = min(rlog), max = max(rlog), mean = mean(rlog), sd = sd(rlog), overQC = length(which(rlog > 5)) ) 

#Filter out so that peakInfo.QC only has overQC values above 2. 
peakInfo.QC = peakInfo %>% filter(overQC > 2)
#Filter so that normExpressionDF.QC only contains elements with a geneID that also is in peakInfo.QC
normExpressionDF.QC = normExpressionDF %>% filter(geneID %in% peakInfo.QC$geneID)

#Plot the filtered normExpressionDF, but coloring sampleID now
ggplot(normExpressionDF.QC, aes(x = rlog, color = sampleID)) + geom_density()

#Create a table with the samples as columns and rlog as values
normExpression.QC = normExpressionDF.QC%>%
  dplyr::select(geneID, sampleID, rlog) %>% 
  spread(key = sampleID, value = rlog)

#The row names are the geneIDs
rownames(normExpression.QC) = normExpression.QC$geneID
normExpression.QC = normExpression.QC[, -1]
```


### Visualising distributions 


```{r  check the normalised data ,message=FALSE ,warning = FALSE }

#Plot a heat map 
plotSample2SampleDistance(normExpression.QC)
```

### Principal components analysis with normalized counts

```{r pca}

#Do a PCA analysis
mir.pca <- prcomp(t(normExpression.QC), center = TRUE, scale = FALSE) 
#Calculate variance from the PCA output
e.var = (mir.pca[['sdev']]^2 / sum(mir.pca[['sdev']]^2))
#Make e.var to a data frame
e.var = as.data.frame( e.var )
#Add column named PC 
e.var$PC = as.factor(1:nrow(e.var)) 
#Add column with the varience
e.var$Variance = e.var$e.var*100
#Add column with CumulativeVariance
e.var$CumulativeVariance = cumsum(e.var$Variance)
#Plot each PC component and its varience
qplot(PC, Variance, data=e.var, geom = c("point")
      , ylim = c(0,max(e.var$Variance)+5) 
      ,ylab = "Variance (%)")

#Make the first 5 PCA output to a data frame. 
pctable = as.data.frame(mir.pca$x)[, 1:5]
#Add column with sample name
pctable$sampleID = rownames(pctable)
#Add the metainfo
pctable = inner_join(pctable, metaInfo)
#Plot using a function in ExpressionAnalysisFucntions
pca  = plotPCAplot(PCAinfo = pctable, n.comp = 4,
            varianceInfo = e.var, 
            colorComponent = "Genotype",pchComponent = "Replicate"
              
)
# Needs to actually run the command to output plot
pca

# Calculate the mean of the PC to plot lines with respect to mutants
mean = pctable %>% dplyr::select(PC1,PC2,PC3,Genotype) %>%
  dplyr::group_by(Genotype) %>%
  dplyr::summarise(PC1 = mean(PC1) ,PC2 = mean(PC2), PC3 = mean(PC3) )
origo = data.frame(PC1=0, PC2=0, PC3 = 0)

# Create a vector plot with the values (-mean, origo, mean) for each mutant
line.gd7 = rbind(mean[1,2:4],origo,-mean[1,2:4])
line.gd7$factor = "Gd7"
line.toll10b = rbind(mean[2,2:4],origo,-mean[2,2:4])
line.toll10b$factor = "Toll10b"
line.toll9 = rbind(mean[3,2:4],origo,-mean[3,2:4])
line.toll9$factor = "Toll9"
line = rbind(line.gd7,line.toll10b,line.toll9)

```

### Plotting

```{r plotting}

ggplot(line, mapping = aes(x = PC1,y = PC2, color = factor))+ geom_line() + geom_point(data = pctable,  mapping = aes(x = PC1,y = PC2, color = Genotype))

ggplot(line, mapping = aes(x = PC2,y = PC3, color = factor))+ geom_line() + geom_point(data = pctable,  mapping = aes(x = PC2,y = PC3, color = Genotype))

ggplot(line, mapping = aes(x = PC1,y = PC3, color = factor))+ geom_line() + geom_point(data = pctable,  mapping = aes(x = PC1,y = PC3, color = Genotype))


```

## Gene scores
The sequencing data has a lot of dimension, time and the different mutants. We want to reduce the number of dimensions to a single score. This is done by the loading scores found by PCA. The loading scores are used as a distance. The distance will then tell us how much variance is found, i.e. the bigger the distance the more it is different. We can e.g. from the scores say if a sample is up-regulated with respect to gd7
 
```{r convert loadings to the factor Scores}
# Saves the loading scores for PC1-PC3
line = rbind(line.gd7, line.toll10b, line.toll9)
loadings = mir.pca$rotation[, 1:3]

# Converts the loading values by subtracting the column(PC) mean and dividing by the column(PC) standard deviation. The columns needs to be transposed to rows to calculate and are then transposed back to columns
loadings_norm = as.data.frame(t((t(loadings) - colMeans(loadings)) / colSds(loadings)))

loadings_norm$geneID = rownames(loadings)

# We calculate the distance of each sample and mutant in the 3D space of PC1-3. The mean is seen as the "normal" regulation and the distance is a measure of how large of a variance.
loadings_new = loadings_norm
for (j in unique(line$factor)) {
  loadings_new[[j]] = 0
  for (i in 1:nrow(loadings_norm)) {
    line1 = line %>% dplyr::filter(factor == j) %>% dplyr::select(PC1, PC2, PC3)
    loadings_new[i, j] = dist3d(
      b = as.numeric(line1[1,]),
      c = as.numeric(line1[3,]),
      point = as.numeric(loadings_new[i, 1:3])
    )
    
  }
  
}
loadingsDF = gather(loadings_new, key = PC, value = Loading, -geneID)

ggplot(loadingsDF, mapping = aes(x = Loading, color = PC)) + geom_density()

# Recreates the same data fram as loadings_new but in another column order
loadingsDF.all = loadingsDF %>%  spread(key = PC, value = Loading)

tmp = loadingsDF.all
tmp$distance =  ((tmp$PC1 ^ 2 + tmp$PC2 ^ 2 + tmp$PC3 ^ 2) ^ 0.5)
loadingsDF.all = tmp


loadingsDF.all.factorDistance = loadingsDF.all %>% gather(key = factor, value = distance2Factor, Gd7, Toll10b, Toll9)

loadingsDF.all.factorDistance$distanceFactor = sqrt(
  loadingsDF.all.factorDistance$distance ^ 2 -
    loadingsDF.all.factorDistance$distance2Factor ^
    2
)

loadingsDF.all.factorDistance$distanceFactor[is.na(loadingsDF.all.factorDistance$distanceFactor)] = 0

# Set the default value to down regulated
loadingsDF.all.factorDistance$direction = "Down"
# The mean is seen as up regulated, and down regulated as the negative value of the mean
line$Dir = rep(c("Up", "origo", "Down"), 3)

# Now we loop through all values and evaluate if it is up or down regulated with respect for each mutant
for (i in 1:nrow(loadingsDF.all.factorDistance)) {
  factor2 = loadingsDF.all.factorDistance[i,]$factor
  UpPoint = line %>% dplyr::filter(factor == factor2 &
                                     Dir == "Up") %>% dplyr::select(PC1, PC2, PC3)
  DownPoint = line %>% dplyr::filter(factor == factor2 &
                                       Dir == "Down") %>% dplyr::select(PC1, PC2, PC3)
  point = loadingsDF.all.factorDistance[i, c("PC1", "PC2", "PC3")]
  up = distance3d(as.numeric(UpPoint), as.numeric(point))
  down = distance3d(as.numeric(DownPoint), as.numeric(point))
  
  if (up < down) {
    loadingsDF.all.factorDistance[i, "direction"] = "Up"
  } else{
    loadingsDF.all.factorDistance[i, "distanceFactor"] =
      -loadingsDF.all.factorDistance[i, "distanceFactor"]
  }
}

# Collect all of the distances to the factors and have them in one column. Now there will be one distance per region
loadingsDF.all.factorDistance2 = loadingsDF.all.factorDistance %>% dplyr::select(geneID, PC1, PC2, PC3, distance, factor, distanceFactor) %>% spread(key = factor, value = distanceFactor)

# The max value is the biggest value of the distanceFactor for each geneID
loadingsDF.all.factorDistance.max = loadingsDF.all.factorDistance %>% group_by(geneID) %>% summarise(max = max(abs(distanceFactor)))
loadingsDF.all.factorDistance$max = abs(loadingsDF.all.factorDistance$distanceFactor)
# Add the max column to the rest of the data frame
loadingsDF.all.factorDistance.max = inner_join(loadingsDF.all.factorDistance,
                                               loadingsDF.all.factorDistance.max) %>% dplyr::select(geneID, factor, direction)

# Add the sample info
loadingsDF.all.factorDistance.max.peakInfo = inner_join (peakInfoLocation, loadingsDF.all.factorDistance.max)
# Add the loading scores
loadingsDF.all.factorDistance.max.peakInfo = inner_join(loadingsDF.all.factorDistance.max.peakInfo,
                                                        loadingsDF.all.factorDistance2)
```

### Print ChIPseq data to file

```{r save the files}
sampleInfoFile = paste(
  params$workingDir,
  params$resultsDir,
  params$Antibody,
  "H3K27ac.sampleInfo.QC.tab.txt",
  sep = "/"
)

sampleInfo.QC = sampleInfo %>%
  dplyr::select(sampleID, Genotype, Replicate)

write.table(
  x = sampleInfo.QC,
  file = sampleInfoFile,
  quote = F,
  sep =  "\t",
  row.names = F,
  col.names = T
)

peakTableFile = paste(
  params$workingDir,
  params$resultsDir,
  params$Antibody,
  "H3K27ac.peak.QC.tab.txt",
  sep = "/"
)
peakInfo.QC = inner_join(peakInfoLocation, peakInfo.QC)
write.table(
  x = peakInfo.QC,
  file = peakTableFile,
  quote = F,
  sep =  "\t",
  row.names = F,
  col.names = T
)

countData.QC = countData[peakInfo.QC$geneID, sampleInfo.QC$sampleID]
countTableFile = paste(
  params$workingDir,
  params$resultsDir,
  params$Antibody,
  "H3K27ac.counts.QC.tab.txt",
  sep = "/"
)
write.table(
  x = countData.QC,
  file = countTableFile,
  quote = F,
  sep =  "\t",
  row.names = T,
  col.names = T
)

rlogTableFile = paste(
  params$workingDir,
  params$resultsDir,
  params$Antibody,
  "H3K27ac.rlog.QC.tab.txt",
  sep = "/"
)


rlogExpression.QC = normExpression[peakInfo.QC$geneID, sampleInfo.QC$sampleID]
write.table(
  x = rlogExpression.QC,
  file = rlogTableFile,
  quote = F,
  sep =  "\t",
  row.names = T,
  col.names = T
)


# Remove PC scores and save as tsv file
loadingsDF.all.factorDistance.max.peakInfo2 =
  loadingsDF.all.factorDistance.max.peakInfo %>%
  dplyr::select(-PC1,-PC2,-PC3,-distance)


fileNameK27ac = paste(params$workingDir,
                      params$ChIPresultsDir,
                      params$ChIPSeq.data,
                      sep = "/")
write.table(
  x = loadingsDF.all.factorDistance.max.peakInfo2,
  file = fileNameK27ac,
  quote = F,
  sep = "\t",
  col.names = T,
  row.names = F
)
```







