---
title: "ChIPseq H3K27Ac analysis"
author: "Jenny Jakobsson"
date: "11/11/2020"
output: html_document

params:
  workingDir: ~/Applied_Bioinformatics_Group_4
  dataDir: data_johan/data/ChIPseq/H3K27ac
  sampleInfo: data_johan/information/cleaned/ChIP_ATAC_sampleInfo.tab.txt
  chipSeqData: K27ac.featureCount.count.tab.txt
  RscriptsDir: script/original_scripts
  resultsDir: results/original_method/ChIPseq
  Antibody: K27ac
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# Install required packages
#install.packages("kableExtra")
#install.packages("tidyverse")
#install.packages("RColorBrewer")
#BiocManager::install("DESeq2")
#BiocManager::install("edgeR")

# Include the functions from the expression analysis script
source(paste(params$workingDir, params$RscriptsDir, "ExpressionAnalysisFunctions.r",sep = "/"))

# Load the installed packages
library(tidyverse)
library(gplots)
library(knitr)
library(kableExtra)
library("viridis")
library(RColorBrewer)
library(DESeq2)
library(edgeR)

```

## H3K27ac

### Using feature count to get the counts

```{r Adding the first analysis of the data, include=FALSE}
#Load in the path to the H3K27ac feature count file
featureCountFile = paste(params$workingDir, params$dataDir, params$chipSeqData, sep = "/")

#Load the data into a table
FCdata = read.table( file = featureCountFile, header = TRUE, 
                     sep = "\t", quote = "",  
                     stringsAsFactors = FALSE)

#Divide into a countdata table and peakinfo table
peakInfoLocation  = FCdata [,1:6] %>% dplyr::rename(geneID = Geneid)
countData = FCdata[,-1:-6]
rownames(countData) = peakInfoLocation$geneID

#Load the ChIP ATAC sampleInfo file that contains info about the samples. 
sampleInfoFile =paste(params$workingDir, params$sampleInfo, sep = "/")
sampleInfo = read.table( file = sampleInfoFile, header = TRUE, 
                     sep = "\t", quote = "",  
                     stringsAsFactors = FALSE)

#Add a column with better names of the bam files
sampleInfo$bamFiles2  =  gsub(pattern = "/",replacement = ".",x = sampleInfo$bamFiles)
sampleInfo$bamFiles2  =  gsub(pattern = "-",replacement = ".",x = sampleInfo$bamFiles2)

#Modify names of the column Genotype
sampleInfo$Genotype = recode(sampleInfo$Genotype, Toll910 = "Toll9")
sampleInfo$Genotype = recode(sampleInfo$Genotype, gd7 = "Gd7")

#Pick out only the samples from K27ac
sampleInfo =sampleInfo %>% filter(Antibody == params$Antibody) 

#Create a dataframe with a row for each column in countData
sampleInfoCountTable = data.frame(bamFiles2 = colnames(countData))

#Join sampleInfo and sampleInfoCountTable
sampleInfo = inner_join(sampleInfo,sampleInfoCountTable)

#Pick the samples that are in sampleInfo
countData = countData[,sampleInfo$bamFiles2]
#Paste genotype and replicate name together
#Add as column name
colnames(countData) = paste(sampleInfo$Genotype, sampleInfo$Replicate, sep = "_")
#Add as a new column
sampleInfo$sampleID = paste(sampleInfo$Genotype, sampleInfo$Replicate, sep = "_")

```



## Filtering

1. removing all peaks with less than 100 counts in all samples
2. Removing all peaks where not at least two samples has more than 5 in normalized rlog values

```{r differential gene expression analysis, message=FALSE}
#Create expression and meta data. 
exp.data = countData
metaInfo = sampleInfo

rownames(metaInfo) = sampleInfo$sampleID

#Create a factor of genotype and sampleID values
metaInfo$Genotype = as.factor(metaInfo$Genotype)
metaInfo$sampleID = as.factor(metaInfo$sampleID)

#Create a DESeq data set from exp.data and metaInfo
dds <- DESeqDataSetFromMatrix(countData = exp.data,
                               colData = metaInfo, ~1)

#Remove rows with low counts and normalize samples for visualization
dds <- dds[ rowSums(counts(dds)) > 100, ]

dim(dds)

#Transform the count data in the DESeq data frame to the log2 scale which minimizes differences between samples for rows with small counts, and which normalizes with respect to library size
rld <- rlog(dds)

#Put this normalized data into a data frame
normExpression = as.data.frame(assay(rld))

#Create a normalized expression data frame
normExpression$geneID = rownames(normExpression)
normExpressionDF  = normExpression %>% gather( key = sampleID, value = rlog, -geneID)
normExpressionDF = inner_join(normExpressionDF, metaInfo)

#normExpressionDF = inner_join(normExpressionDF, sampleInfo)

#Plot normExpressionDF
ggplot(normExpressionDF, aes(x = rlog)) + geom_density()

peakInfo = normExpressionDF %>% dplyr::select(geneID, rlog, sampleID) %>% 
  dplyr::group_by (geneID) %>%
  dplyr::summarize ( min = min(rlog), max = max(rlog), mean = mean(rlog), sd = sd(rlog), overQC = length(which(rlog > 5)) ) 

#Filter out so that peakInfo.QC only has overQC values above 2. 
peakInfo.QC = peakInfo %>% filter(overQC > 2)
#Filter so that normExpressionDF.QC only contains elements with a geneID that also is in peakInfo.QC
normExpressionDF.QC = normExpressionDF %>% filter(geneID %in% peakInfo.QC$geneID)

#Plot the filtered normExpressionDF, but coloring sampleID now
ggplot(normExpressionDF.QC, aes(x = rlog, color = sampleID)) + geom_density()

#Create a table with the samples as columns and rlog as values
normExpression.QC = normExpressionDF.QC%>%
  dplyr::select(geneID, sampleID, rlog) %>% 
  spread(key = sampleID, value = rlog)

#The row names are the geneIDs
rownames(normExpression.QC) = normExpression.QC$geneID
normExpression.QC = normExpression.QC[, -1]



```


### Visualising distributions 


```{R  check the normalised data ,message=FALSE ,warning = FALSE }

#Plot a heat map 
plotSample2SampleDistance(normExpression.QC)

#Do a PCA analysis
mir.pca <- prcomp(t(normExpression.QC), center = TRUE, scale = FALSE) 
#Calculate evar from the PCA output
e.var = (mir.pca[['sdev']]^2 / sum(mir.pca[['sdev']]^2))
#Make e.var to a data frame
e.var = as.data.frame( e.var )
#Add column named PC 
e.var$PC = as.factor(1:nrow(e.var)) 
#Add column with the varience
e.var$Variance = e.var$e.var*100
#Add column with CumulativeVariance
e.var$CumulativeVariance = cumsum(e.var$Variance)
#Plot each PC component and its varience
qplot(PC, Variance, data=e.var, geom = c("point")
      , ylim = c(0,max(e.var$Variance)+5) 
      ,ylab = "Variance (%)")




#Make the first 5 PCA output to a data frame. 
pctable = as.data.frame(mir.pca$x)[, 1:5]
#Add column with sample name
pctable$sampleID = rownames(pctable)
#Add the metainfo
pctable = inner_join(pctable, metaInfo)

#Plot using a function in ExpressionAnalysisFucntions
test  = plotPCAplot(PCAinfo = pctable, n.comp = 4,
            varianceInfo = e.var, 
            colorComponent = "Genotype",pchComponent = "Replicate"
              
)

test





```


### Print ChIPseq data to file

```{r save the files}
sampleInfoFile = paste(params$workingDir, params$resultsDir, params$Antibody, "H3K27ac.sampleInfo.QC.tab.txt", sep = "/")

sampleInfo.QC = sampleInfo %>%
  dplyr::select( sampleID, Genotype, Replicate)

write.table(x = sampleInfo.QC, file = sampleInfoFile, quote = F, sep =  "\t", row.names = F,  col.names = T)

peakTableFile = paste(params$workingDir, params$resultsDir, params$Antibody, "H3K27ac.peak.QC.tab.txt", sep = "/")
peakInfo.QC = inner_join(peakInfoLocation,peakInfo.QC)
write.table(x = peakInfo.QC, file = peakTableFile, quote = F, sep =  "\t", row.names = F, col.names = T)

countData.QC = countData[peakInfo.QC$geneID,sampleInfo.QC$sampleID]
countTableFile = paste(params$workingDir, params$resultsDir, params$Antibody, "H3K27ac.counts.QC.tab.txt", sep = "/")
write.table(x = countData.QC, file = countTableFile, quote = F, sep =  "\t", row.names = T, col.names = T)

rlogTableFile = paste(params$workingDir, params$resultsDir, params$Antibody, "H3K27ac.rlog.QC.tab.txt", sep = "/")


rlogExpression.QC = normExpression[peakInfo.QC$geneID,sampleInfo.QC$sampleID]
write.table(x = rlogExpression.QC, file = rlogTableFile, quote = F, sep =  "\t", row.names = T, col.names = T)

```







