---
title: "ATACseq analysis"
author: "Jenny Jakobsson"
date: "11/11/2020"
output: html_document

params:
    workingDir: ~/Applied_Bioinformatics_Group_4
    dataDir: data_johan/data/ATACseq
    RscriptsDir: script/original_scripts
    proSeqDir: results/original_method/PRO-seq
    ATACresultsDir: results/altered_method/ATACseq
    sampleInfo: data_johan/information/cleaned/ChIP_ATAC_sampleInfo.tab.txt
    count_ATAC: ATAC.CBP.peak.featureCount.count.tab.txt
    count_ATAC_QC: ATAC.CBP.peak.featureCount.count.QC.tab.txt
    normalisedATACpeakData: ATAC.CBP.peak.featureCount.rlog.tab.txt
    ATACSeq.data: altered.ATACseq.data.tsv
    Method: ATAC-Seq
    PROseq.data: data_johan/results/PROseq/PROseq.data.tsv
    
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# Install required packages
#install.packages("kableExtra")
#install.packages("tidyverse")
#install.packages("RColorBrewer")

# Include the functions from the expression analysis script
source(paste(params$workingDir, params$RscriptsDir,"ExpressionAnalysisFunctions.r", sep = "/"))
library(tidyverse)
library(gplots)
library(kableExtra)
library(DESeq2)
library("viridis")
```

## Loading the data
### Loading the count data
Loads the count data and extracts the interesting information. Also changes column names to more understandable headers.

```{r Adding the first analysis of the data ,message=FALSE ,warning = FALSE}
# Read in the ATACseq count data
featureCountFile = paste(params$workingDir, params$dataDir,params$count_ATAC, sep = "/")
FCdata = read.table( file = featureCountFile, header = TRUE, sep = "\t", quote = "", stringsAsFactors = FALSE)
# Save the columns with info in peak info
peakInfo  = FCdata [,1:6]
# Save the count data info in table countData
countData = FCdata[,-1:-6]
# Change row names to geneID which is saved in peakInfo
rownames(countData) = peakInfo$Geneid

# Load the sample data
sampleInfoFile = paste(params$workingDir, params$sampleInfo, sep = "/")
sampleInfo = read.table(
  file = sampleInfoFile,
  header = TRUE,
  sep = "\t",
  quote = "",
  stringsAsFactors = FALSE
)

# Load the PROseq data
proseq_file = paste(params$workingDir, params$PROseq.data, sep ="/")
PROseq.data = read.table(file = proseq_file, header = TRUE, sep="\t", quote="", stringsAsFactors = FALSE)
```

### Loading the sample info
Load the sample info and merge the information with the count data

```{r handle the samples ,message=FALSE ,warning = FALSE}

#Change the names of bamFiles2 so that they make more sense
sampleInfo$bamFiles2  =  gsub(pattern = "/",
                              replacement = ".",
                              x = sampleInfo$bamFiles)
sampleInfo$bamFiles2  =  gsub(pattern = "-",
                              replacement = ".",
                              x = sampleInfo$bamFiles2)

# Rename the mutants in sampledata so that the names are universal for the data
sampleInfo$Genotype = recode(sampleInfo$Genotype, Toll910 = "Toll9")
sampleInfo$Genotype = recode(sampleInfo$Genotype, `Toll9/10` = "Toll9")
sampleInfo$Genotype = recode(sampleInfo$Genotype, gd7 = "Gd7")

# Filter out the sampleInfo that is not ATACseq
sampleInfo = sampleInfo %>% filter(Assaytype == params$Method)

#Make sure that the columns in countData is the same as the name of the bamfiles2 in sampleInfo
countData = countData[, sampleInfo$bamFiles2]

#Change NA in sampleinfo$uclearsample to Mutant
sampleInfo$Nuclearcycle[is.na(sampleInfo$Nuclearcycle)] = "Mutant"


#Rename the columns in countData
colnames(countData) = sampleInfo$SampleInfo

#Divide sampleInfo into fixed and spiked samples. 
fixedSamples = sampleInfo$SampleInfo[grep(pattern = "Fixed", x = sampleInfo$SampleInfo)]
spikedSamples = sampleInfo$SampleInfo[grep(pattern = "spiked", x = sampleInfo$SampleInfo)]

#Create sampleInfo_Mutants with only mutants and that are not in either fixed or spiked
sampleInfo_Mutants = sampleInfo %>% filter(Nuclearcycle == "Mutant") %>%
  filter(!SampleInfo %in% fixedSamples) %>%
  filter(!SampleInfo %in% spikedSamples)

#Create a count data table from these
countData_mutants = countData[, sampleInfo_Mutants$SampleInfo]
```


## Filtering

```{r differential gene expression analysis ,message=FALSE ,warning = FALSE}
#Cretae expression data and meta data and fix rownames
exp.data = countData_mutants
metaInfo = sampleInfo_Mutants
rownames(metaInfo) = sampleInfo_Mutants$SampleInfo

#Create factor values
metaInfo$Antibody = as.factor(metaInfo$Antibody)
metaInfo$Genotype = as.factor(metaInfo$Genotype)

#Create a DESeq data set from exp.data and metaInfo
dds <- DESeqDataSetFromMatrix(countData = exp.data,
                               colData = metaInfo,
                               design = ~Genotype )


#Remove rows with low counts and normalize samples for visualization
dds <- dds[ rowSums(counts(dds)) > 100, ]

dim(dds)

#Transform the count data in the DESeq data frame to the log2 scale which minimizes differences between samples for rows with small counts, and which normalizes with respect to library size
rld <- rlog(dds)

#Put this normalized data into a data frame
normExpression = as.data.frame(assay(rld))

#Create a normalized expression data frame
normExpression$geneID = rownames(normExpression)
normExpressionDF  = normExpression %>% gather( key = sampleName, value = rlog, -geneID)

#normExpressionDF = inner_join(normExpressionDF, sampleInfo)

#Plot and color by sampleName
ggplot(normExpressionDF, aes(x = rlog, color = sampleName)) + geom_density()

# What does 
geneInfo = normExpressionDF %>% 
  dplyr::select(geneID, rlog, sampleName) %>% 
  dplyr::group_by (geneID) %>%
  dplyr::summarize ( min = min(rlog), 
                     max = max(rlog), 
                     mean = mean(rlog), 
                     sd = sd(rlog), 
                     overQC = length(which(rlog > 5))
                     ) 


#Filter out overQC values under 2
geneInfo.QC2 = geneInfo %>% 
  filter(overQC > 2)
normExpressionDF.QC = normExpressionDF %>%
  filter(geneID %in% geneInfo.QC2$geneID)

#Plot again after filtering
ggplot(normExpressionDF.QC, aes(x = rlog, color = sampleName)) + geom_density()


#Create normExpression.QC with samples as columns and rows by geneID and values or rlog
normExpression.QC = normExpressionDF.QC %>%
  dplyr::select(geneID, sampleName, rlog) %>% 
  spread(key = sampleName, value = rlog)

rownames(normExpression.QC) = normExpression.QC$geneID
normExpression.QC = normExpression.QC[, -1]

```

### Print ATAC data to file

```{r save the files ,message=FALSE ,warning = FALSE}
## Saving sampleInfo File

#Load ATAC.sampleInfo.QC.tab.txt into a table named sampleInfo_Mutants
sampleInfoFile = paste(params$workingDir,
                       params$ATACresultsDir,
                       "ATAC.sampleInfo.QC.tab.txt",
                       sep = "/")
write.table(
  x = sampleInfo_Mutants,
  file = sampleInfoFile,
  quote = F,
  sep =  "\t",
  row.names = F,
  col.names = T
)

## Saving peakInfo File
#Create the table peakInfo_QC and save in the file ATAC.CBP.peaks.QC.tab.txt
peakInfo_QC = peakInfo %>%
  dplyr::rename(geneID = Geneid) %>%
  inner_join(geneInfo.QC2)

peakTableFile = paste(params$workingDir,
                      params$ATACresultsDir,
                      "ATAC.CBP.peaks.QC.tab.txt",
                      sep = "/")
write.table(
  x = peakInfo_QC,
  file = peakTableFile,
  quote = F,
  sep =  "\t",
  row.names = F,
  col.names = T
)

#Create the table countData.QC and save in the file ATAC.CBP.peak.featureCount.count.QC.tab.txt
count_ATAC_QC_File = paste(
  params$workingDir,
  params$ATACresultsDir,
  paste(params$count_ATAC_QC, sep = "."),
  sep = "/"
)
countData.QC = countData_mutants[peakInfo_QC$geneID,
                                 sampleInfo_Mutants$SampleInfo]
write.table(
  x = countData.QC,
  file = count_ATAC_QC_File,
  quote = F,
  sep =  "\t",
  row.names = T,
  col.names = T
)


## Saving normalised count file
#Create the table normExpression.QC and save in the file ATAC.CBP.peak.featureCount.rlog.tab.txt
NormalisedCountFile = paste(
  params$workingDir,
  params$ATACresultsDir,
  paste(params$normalisedATACpeakData, sep = "."),
  sep = "/"
)
normExpression.QC = normExpression[peakInfo_QC$geneID, sampleInfo_Mutants$SampleInfo]

write.table(
  x = normExpression.QC,
  file = NormalisedCountFile,
  quote = F,
  sep = "\t",
  col.names = T,
  row.names = T
)

```

```{r heatmap and PCA ,message=FALSE ,warning = FALSE}


plotSample2SampleDistance(normExpression.QC)

#'  __Figure 1 Plotting sample to sample distance__ .
#'  0 means that they are identical and 1 means that they are totally different.
#'  The darker the blue the more similair. Also dendogram shows how similair they are.
#'
#' Samples do not cluster according to pre an post op. Most likely more due to difference in mapping (technical problem)
#'
#' #### PCA analysis
#' Running PCA on the samples and plotting the different variables to see which of the parameterrs that fit the different components the best.
#'
#' First checking how much the different PC contribute.


mir.pca <-
  prcomp(t(normExpression.QC), center = TRUE, scale = FALSE)
e.var = (mir.pca[['sdev']] ^ 2 / sum(mir.pca[['sdev']] ^ 2))
e.var = as.data.frame(e.var)
e.var$PC = as.factor(1:nrow(e.var))
e.var$Variance = e.var$e.var * 100
e.var$CumulativeVariance = cumsum(e.var$Variance)
qplot(
  PC,
  Variance,
  data = e.var,
  geom = c("point"),
  ylim = c(0, max(e.var$Variance) + 5),
  ylab = "Variance (%)"
)

#'  __Figure 2 Plotting PCA variance__.
#'  Displays how much each of the PC contributes to the overall expression.
#'  This suggest that the two first PC explains most of the variation.
#'
#+ save2, include=FALSE
#'
#'

pctable = as.data.frame(mir.pca$x)[, 1:5]
pctable = cbind(pctable, sampleInfo_Mutants)

pca  = plotPCAplot(
  PCAinfo = pctable,
  n.comp = 4,
  varianceInfo = e.var,
  colorComponent = "Genotype",
  pchComponent = "Time"
)
pca

mean = pctable %>% dplyr::select(PC1, PC2, PC3, Genotype) %>%
  dplyr::group_by(Genotype) %>%
  dplyr::summarise(PC1 = mean(PC1) ,
                   PC2 = mean(PC2),
                   PC3 = mean(PC3))
meantime = pctable %>% dplyr::select(PC1, PC2, PC3, Time) %>%
  dplyr::group_by(Time) %>%
  dplyr::summarise(PC1 = mean(PC1) ,
                   PC2 = mean(PC2),
                   PC3 = mean(PC3))

# Extracting distances from PC1, PC2 and PC3
origo = data.frame(PC1 = 0, PC2 = 0, PC3 = 0)
line.gd7 = rbind(mean[1, 2:4], origo, -mean[1, 2:4])
line.gd7$factor = "Gd7"
line.toll10b = rbind(mean[2, 2:4], origo, -mean[2, 2:4])
line.toll10b$factor = "Toll10B"
line.toll9 = rbind(mean[3, 2:4], origo, -mean[3, 2:4])
line.toll9$factor = "Toll9"
line.time = rbind(-meantime[1, 2:4], origo, meantime[1, 2:4])
line.time$factor = "Time"

line = rbind(line.gd7, line.toll10b, line.toll9)

ggplot(line, mapping = aes(x = PC2, y = PC3, color = factor)) + geom_line() + geom_point(data = pctable,
                                                                                         mapping = aes(
                                                                                           x = PC2,
                                                                                           y = PC3,
                                                                                           color = Genotype,
                                                                                           shape = Time
                                                                                         ))

ggplot(line, mapping = aes(x = PC1, y = PC2, color = factor)) + geom_line() + geom_point(data = pctable,
                                                                                         mapping = aes(
                                                                                           x = PC1,
                                                                                           y = PC2,
                                                                                           color = Genotype,
                                                                                           shape = Time
                                                                                         ))

ggplot(line, mapping = aes(x = PC1, y = PC3, color = factor)) + geom_line() + geom_point(data = pctable,
                                                                                         mapping = aes(
                                                                                           x = PC1,
                                                                                           y = PC3,
                                                                                           color = Genotype,
                                                                                           shape = Time
                                                                                         ))
```
###  Gene scores

```{r convert loadings to the factor Scores}
# Saves the loading scores for PC1-PC3
line = rbind(line.gd7, line.toll10b, line.toll9, line.time)
loadings = mir.pca$rotation[, 1:3]

# Converts the loading values by subtracting the column(PC) mean and dividing by the column(PC) standard deviation. The columns needs to be transposed to rows to calculate and are then transposed back to columns
loadings_norm = as.data.frame(t((t(loadings) - colMeans(loadings)) / colSds(loadings)))
loadings_norm$Geneid = rownames(loadings)

loadings_new = loadings_norm
for (j in unique(line$factor)) {
  loadings_new[[j]] = 0
  for (i in 1:nrow(loadings_norm)) {
    line1 = line %>% dplyr::filter(factor == j) %>% dplyr::select(PC1, PC2, PC3)
    loadings_new[i, j] = dist3d(
      b = as.numeric(line1[1, ]),
      c = as.numeric(line1[3, ]),
      point = as.numeric(loadings_new[i, 1:3])
    )
  }
}


loadingsDF = gather(loadings_new, key = PC, value = Loading,-Geneid)

ggplot(loadingsDF, mapping = aes(x = Loading, color = PC)) + geom_density()

# Recreates the same data fram as loadings_new but in another column order
loadingsDF.all = loadingsDF %>%  spread(key = PC, value = Loading)

tmp = loadingsDF.all
tmp$distance =  ((tmp$PC1 ^ 2 + tmp$PC2 ^ 2 + tmp$PC3 ^ 2) ^ 0.5)
loadingsDF.all = tmp

loadingsDF.all.factorDistance = loadingsDF.all %>% gather(key = factor, value = distance2Factor, Time, Gd7, Toll10B, Toll9)
loadingsDF.all.factorDistance$distanceFactor = sqrt(
  loadingsDF.all.factorDistance$distance ^ 2 -
    loadingsDF.all.factorDistance$distance2Factor ^
    2
)
loadingsDF.all.factorDistance$distanceFactor[is.na(loadingsDF.all.factorDistance$distanceFactor)] = 0

loadingsDF.all.factorDistance$direction = "Down"
line$Dir = rep(c("Up", "origo", "Down"), 4)


for (i in 1:nrow(loadingsDF.all.factorDistance)) {
  factor2 = loadingsDF.all.factorDistance[i, ]$factor
  UpPoint = line %>% dplyr::filter(factor == factor2 &
                                     Dir == "Up") %>% dplyr::select(PC1, PC2, PC3)
  DownPoint = line %>% dplyr::filter(factor == factor2 &
                                       Dir == "Down") %>% dplyr::select(PC1, PC2, PC3)
  point = loadingsDF.all.factorDistance[i, c("PC1", "PC2", "PC3")]
  up = distance3d(as.numeric(UpPoint), as.numeric(point))
  down = distance3d(as.numeric(DownPoint), as.numeric(point))
  
  if (up < down) {
    loadingsDF.all.factorDistance[i, "direction"] = "Up"
  } else{
    loadingsDF.all.factorDistance[i, "distanceFactor"] =
      -loadingsDF.all.factorDistance[i, "distanceFactor"]
  }
}

loadingsDF.all.factorDistance2 = loadingsDF.all.factorDistance %>% dplyr::select(Geneid, PC1, PC2, PC3, distance, factor, distanceFactor) %>% spread(key = factor, value = distanceFactor)


loadingsDF.all.factorDistance.max = loadingsDF.all.factorDistance %>% group_by(Geneid) %>% summarise(max = max(abs(distanceFactor)))
loadingsDF.all.factorDistance$max = abs(loadingsDF.all.factorDistance$distanceFactor)
loadingsDF.all.factorDistance.max = inner_join(loadingsDF.all.factorDistance,
                                               loadingsDF.all.factorDistance.max) %>%
  dplyr::select(Geneid, factor, direction)
loadingsDF.all.factorDistance.max.peakInfo = inner_join (peakInfo, loadingsDF.all.factorDistance.max)
loadingsDF.all.factorDistance.max.peakInfo = inner_join(loadingsDF.all.factorDistance.max.peakInfo,
                                                        loadingsDF.all.factorDistance2)

loadingsDF.all.factorDistance.max.peakInfo2 =
  loadingsDF.all.factorDistance.max.peakInfo %>%
  dplyr::select(-PC1,-PC2, -PC3, -distance)

fileNameATACseq = paste(params$workingDir,
                        params$ATACresultsDir,
                        params$ATACSeq.data,
                        sep = "/")
write.table(
  x = loadingsDF.all.factorDistance.max.peakInfo2,
  file = fileNameATACseq,
  quote = F,
  sep = "\t",
  col.names = T,
  row.names = F
) 


```

```{r test, include=FALSE}
# # Parameter changeable??? Goes 9000 to 2000 genes
# loadingsDF.all.filtered = loadingsDF.all.factorDistance.max.peakInfo %>% filter(distance >2)  %>% dplyr::select(Geneid, factor, direction)
# 
# loadingsDF.all.filtered2  = inner_join(loadingsDF.all.filtered, loadingsDF.all.factorDistance2)
# 
# 
# loadingsDF.all.filtered = loadingsDF.all.filtered2 
# dim(loadingsDF.all.filtered)
# 
# loadingsDF.all.filtered.peakInfo = inner_join (peakInfo,loadingsDF.all.filtered )
# 
# PROseq.data.selected  = PROseq.data %>%filter (annotation2 == "DE") %>%dplyr::select(geneID,transcriptID, factor,direction) %>% distinct()
# 
# 
# write.table(x = loadingsDF.all.filtered.peakInfo, 
#             file = paste( params$workingDir, params$ATACresultsDir, 
#                           "selectedGenes.tsv" ,sep = "/"),
#             quote = F, sep = "\t", col.names = TRUE, row.names = F)  
```

