---
title: "Trying to connect gene and enhancer"
author: "Elin Antonsson"
output: html_document

params:
  workingDir: ~/Applied_Bioinformatics_Group_4/
  PROseq.gene.data: results/original_method/PROseq/PROseq.data.tsv
  enhancer.data: results/altered_method/enhancers/enhancers.tsv

  
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r from ATAC-CBP_seq_analysis}

#This is just the output from PROseq_analysis
our_enhancer_file = paste(params$workingDir, params$enhancer.data, sep = "")

our_enhancer_data = read.table(
  file = our_enhancer_file ,
  quote = "",
  sep = "\t",
  header = TRUE,
  stringsAsFactors = F
)

our_enhancer_data = tibble::rowid_to_column(our_enhancer_data, "enhancer_row")

our_enhancer_data= our_enhancer_data %>% dplyr::rename(
  "EnhancerID" = geneID)

#This is the output from joint_peak_with_UMAP
ourEnhancerData = makeGRangesFromDataFrame(our_enhancer_data,
                                      seqnames.field = "Chr", 
                                      start.field = "start", 
                                      end.field = "end")



PROseq_gene_file = paste(params$workingDir, params$PROseq.gene.data, sep = "")

PROseq.gene.data = read.table(
  file = PROseq_gene_file ,
  quote = "",
  sep = "\t",
  header = TRUE,
  stringsAsFactors = F
)

PROseqGeneInfo = PROseq.gene.data %>%  dplyr::select(geneID, Chr, Start, End, Strand, direction, factor) %>% dplyr::rename(
  "Gene_geneID" = geneID,
  "Gene_Chr" = Chr,
  "Gene_Start" = Start,
  "Gene_End" = End,
  "Gene_Strand" = Strand,
  "Gene_Direction" = direction,
  "Gene_Factor" = factor
)

PROseqGeneInfo = PROseqGeneInfo %>% distinct()

PROseqGeneInfo$Gene_Chr = paste("chr",PROseqGeneInfo$Gene_Chr, sep = "")
PROseqGeneInfo = tibble::rowid_to_column(PROseqGeneInfo, "gene_row")

PROData = makeGRangesFromDataFrame(PROseqGeneInfo,
                                      seqnames.field = "Gene_Chr", 
                                      start.field = "Gene_Start", 
                                      end.field = "Gene_End")



test_pairs = data.frame(distanceToNearest(ourEnhancerData, PROData, ignore.strand=TRUE))
colnames(test_pairs) = c("enhancer_row","gene_row","EnhancerDistance")

#test_pairs$K27ac_geneID = names(PROseqGeneInfo)[test_pairs$subjectHits]
#test_pairs$geneID = names(our_enhancer_data)[test_pairs$queryHits]
#test_pairs = test_pairs %>% dplyr::select(K27ac_geneID,geneID)

joined = inner_join(PROseqGeneInfo, test_pairs)
joined = inner_join(our_enhancer_data, joined)

joined$PRO_direction = gsub(pattern = "None", replacement = "Up", joined$Gene_Direction)
joined$PROseq_Classification = paste(joined$Gene_Factor,joined$PRO_direction, sep = " " )
joined$ATACseq_Classification = paste(joined$ATAC_factor,joined$ATAC_direction, sep = " " )

enhancer_gene_pairs = joined %>% dplyr::select(EnhancerID, Chr, start, end, Gene_geneID, Gene_Chr, Gene_Start, Gene_End, EnhancerDistance, PROseq_Classification, ATACseq_Classification)

enhancer_gene_pairs_unique = unique(enhancer_gene_pairs)

```

```{r Filtering}
enhancer_gene_pairs_unique_0kb = enhancer_gene_pairs_unique %>% filter(EnhancerDistance == 0 )  

enhancer_gene_pairs_unique_ambiguous = enhancer_gene_pairs_unique %>% filter(EnhancerDistance == 0 & EnhancerDistance < 1500 )  
enhancer_gene_pairs_unique_over_15kb = enhancer_gene_pairs_unique %>% filter(EnhancerDistance > 1500 )  
```

```{r Filtering visualization}


confusion_matrix <- as.data.frame(table(enhancer_gene_pairs_over_15kb$PROseq_Classification, enhancer_gene_pairs_over_15kb$ATACseq_Classification))
colnames(confusion_matrix) = c("PROseq classification", "ATACseq classification", "Freq")
ggplot(data = confusion_matrix,
       mapping = aes(x = `PROseq classification`,
                     y = `ATACseq classification`)) + 
  geom_tile(aes(fill = Freq)) +
  geom_text(aes(label = sprintf("%1.0f", Freq)), vjust = 1) +
  scale_fill_gradient(low = "blue",
                      high = "red",
                      trans = "log")+ 
  labs(title = "ATACseq peak and PROseq Gene correlation ",
              subtitle = "15 - 100 kb up- or downstream of gene",
              caption = "Gene is subset of genes that where identified using PROseq data
Genes are divided based on PROseq analysis annotation
         Peak is 15 - 100 kb up- or downstream of gene") 
#ggsave(filename = paste(params$workingDir,params$ATACSeqDir, "ATAC_PROseqCorrelation_100kb.pdf", sep="/"), width = 8)

ggplot(joined, mapping = aes(x =ATAC_factor , fill = ATAC_direction ))+ 
  geom_bar() + 
  facet_grid( cols = vars(Gene_Factor,Gene_Direction))+ 
  labs(title = "ATACseq peak and PROseq Gene correlation ",
              subtitle = "15 -100 kb up- or downstream of gene",
              caption = "Gene is subset of genes that where identified using PROseq data
Genes are divided based on PROseq analysis annotation
         Peak is 15 -100 kb up- or downstream of gene") + theme(axis.text.x = element_text(angle = 90, hjust = 1))
#ggsave(filename = paste(params$workingDir,params$ATACSeqDir, "ATAC_PROseqCorrelation_100kb.barplot.pdf", sep="/"), width = 8)

```


```{r Visualising correlation  background}

Both_data  = joined %>% filter(ATAC_factor != "Time")


ggplot(Both_data, aes(x= log10(EnhancerDistance+1 )))+ geom_histogram()
ggsave(filename = paste(params$workingDir,params$ATACSeqDir, "ATAC_PROseqCorrelation_distance.pdf", sep="/"), width = 8)

confusion_matrix <- as.data.frame(table(Both_data$PROseq_Classification, Both_data$ATACseq_Classification))
colnames(confusion_matrix) = c("PROseq classification", "ATACseq classification", "Freq")
ggplot(data = confusion_matrix,
       mapping = aes(x = `PROseq classification`,
                     y = `ATACseq classification`)) + 
  geom_tile(aes(fill = Freq)) +
  geom_text(aes(label = sprintf("%1.0f", Freq)), vjust = 1) +
  scale_fill_gradient(low = "blue",
                      high = "red",
                      trans = "log")+ 
  labs(title = "ATACseq peak and PROseq Gene correlation ",
              subtitle = "All closest peak and gene",
              caption = "Gene is subset of genes that where identified using PROseq data
Genes are divided based on PROseq analysis annotation
         Peak is all distances  up- or downstream of gene") 



ggplot(Both_data, mapping = aes(x =ATAC_factor , fill = ATAC_direction ))+ 
  geom_bar() + 
  facet_grid( cols = vars(Gene_Factor,Gene_Direction))+ 
  labs(title = "ATACseq peak and PROseq Gene correlation ",
              subtitle = "All closest peak and gene",
              caption = "Gene is subset of genes that where identified using PROseq data
Genes are divided based on PROseq analysis annotation
         Peak is all distances  up- or downstream of gene")+
  theme(axis.text.x = element_text(angle = 90, hjust = 1))



```



