---
title: "Joint Peak Analysis with UMAP"
author: "Elin Antonsson"
output: html_document

params:
  workingDir: ~/Applied_Bioinformatics_Group_4/
  RscriptsDir: script/original_scripts/
  inSitu.data: data_johan/results/cleaned/inSitu/inSitu.Stark.dm6.tsv
  PROseq.data: results/altered_method/PROseq/PROseq.CBP.data.tsv #this was not
  PROseq.gene.data: results/original_method/PROseq/PROseq.data.tsv
  ChIPseqDir: results/original_method/ChIPseq/
  CBP.data: CBP/ChIPseq.CBP.data.tsv
  K27ac.data: K27ac/ChIPseq.H3K27Ac.data.tsv
  ATACseq.data: results/original_method/ATACseq/ATACseq.data.tsv
  GD7enhancer: data_johan/results/cleaned/enhancers/h3k27ac_gd7_csaw_intersect.bed
  Toll10Benhancer: data_johan/results/cleaned/enhancers/h3k27ac_Toll10B_csaw_intersect.bed
  Toll9enhancer: data_johan/results/cleaned/enhancers/h3k27ac_Tollrm910_csaw_intersect.bed
  enhancerResult: results/altered_method/enhancers/enhancers.tsv
  
  results_file: results/altered_method/from_joint_peak/UMAP_enhancers.tsv
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
source(paste(params$workingDir, params$RscriptsDir,"ExpressionAnalysisFunctions.r", sep = ""))

#install.packages("tidyverse")
#install.packages("ROCit")
#install.packages("patchwork")
#BiocManager::install("heatmaps")
#devtools::install_github("ropenscilabs/umapr")
library("heatmaps")
library(patchwork)
library(ROCit)
library(tidyverse)
library(GenomicRanges)
library(umapr)
```
# Combining the scores to predict enhancer gene pairs

## Load the required files
```{r Loading the ATACseq data files}
# Load the ATAC seq loading scores
ATACseq.data = read.table(
  file = paste(params$workingDir, params$ATACseq.data, sep = ""),
  quote = "",
  sep = "\t",
  header = TRUE,
  stringsAsFactors = F
)
# Rename the columns
ATACseqInfo = ATACseq.data %>% dplyr::rename(
  "ATAC_Gd7" = Gd7,
  "ATAC_Toll10b" = Toll10b,
  "ATAC_Toll9" = Toll9,
  "ATAC_Time" = Time,
  "ATAC_Start" = Start,
  "ATAC_End" = End ,
  "ATAC_factor" = factor,
  "ATAC_direction" = direction, 
  
)

```

```{r laoding the ChIPseq data files}
#Load the ChIPseq data for CBP
CBPInfo =  read.table(
  file = paste(params$workingDir, params$ChIPseqDir, params$CBP.data, sep = ""), 
  quote = "",
  sep = "\t",
  header = TRUE,
  stringsAsFactors = F
)
# Rename the mutant headers
CBPInfo = CBPInfo %>%
  dplyr::select(geneID, Gd7, Toll10b, Toll9) %>%
  dplyr::rename("CBP_Gd7" = Gd7,
                "CBP_Toll10b" = Toll10b,
                "CBP_Toll9" = Toll9)

# Load the ChIPseq data for H3K27ac
K27acInfo =  read.table(
  file = paste(params$workingDir,params$ChIPseqDir, params$K27ac.data, sep = "/"),
  quote = "",
  sep = "\t",
  header = TRUE,
  stringsAsFactors = F
)
# Rename the columns
K27acInfo = K27acInfo %>% dplyr::rename(
  "K27ac_geneID" = geneID,
  "K27ac_Start" = Start,
  "K27ac_End" = End ,
  "K27ac_factor" = factor,
  "K27ac_direction" = direction,
  "K27ac_Gd7" = Gd7,
  "K27ac_Toll10b" = Toll10b,
  "K27ac_Toll9" = Toll9,
  "K27ac_Chr" = Chr,
  "K27ac_Strand" = Strand,
  "K27ac_Length" = Length
)
```

```{r Load the PROseq data files}
# Load the PROseq data
PROseq_file = paste(params$workingDir, params$PROseq.data, sep = "")

PROseq.data = read.table(
  file = PROseq_file ,
  quote = "",
  sep = "\t",
  header = TRUE,
  stringsAsFactors = F
)

# Rename headers.
PROseqInfo = PROseq.data %>%  dplyr::select(geneID, Gd7, Toll10b, Toll9, Time, factor, direction) %>% dplyr::rename(
  "PRO_Gd7" = Gd7,
  "PRO_Toll10b" = Toll10b,
  "PRO_Toll9" = Toll9,
  "PRO_Time" = Time, 
  "Factor" = factor, 
  "Direction" = direction
)

PROseq_gene_file = paste(params$workingDir, params$PROseq.gene.data, sep = "")

PROseq.gene.data = read.table(
  file = PROseq_gene_file ,
  quote = "",
  sep = "\t",
  header = TRUE,
  stringsAsFactors = F
)


PROseqGeneInfo = PROseq.gene.data %>%  dplyr::select(geneID, Chr, Start, End, Strand) %>% dplyr::rename(
  "Gene_geneID" = geneID,
  "Gene_Chr" = Chr,
  "Gene_Start" = Start,
  "Gene_End" = End,
  "Gene_Strand" = Strand
)
PROseqGeneInfo = PROseqGeneInfo %>% distinct()

```

```{r Load the Insitu data files}
# Load in the inSitu data
inSituFile = paste(params$workingDir, params$inSitu.data, sep = "")
inSitu.Data = read.table(
  file = inSituFile,
  header = T,
  sep = "\t",
  quote = "",
  stringsAsFactors = F
)

# Need to transform the columns to match our data
inSitu.Data = inSitu.Data %>% dplyr::select(seqnames, start, end, stg4_6) %>% separate_rows(stg4_6, sep = "\\|") %>%
  separate(stg4_6, into = c("Location", "score"), sep = ";")
inSitu.Data = inSitu.Data %>% replace_na(list(score = 0, Location = "None"))
inSitu.Data$insituID = paste("region", 1:nrow(inSitu.Data))

# Rename the columns
inSituInfo = inSitu.Data %>% dplyr::select(insituID , Location, score) %>% dplyr::rename(inSituLocation = Location, inSituScore = score)
```

```{r Load the enhancer data files}
# Load the enhancer data for all mutants to compare the data to
GD7enhancersFile = paste(params$workingDir, params$GD7enhancer, sep = "")
GD7enhancers = read.table(
  file = GD7enhancersFile,
  header = F,
  sep = "\t",
  quote = "",
  stringsAsFactors = F
)
GD7enhancers = GD7enhancers[, 1:3]

Toll10BenhancersFile = paste(params$workingDir, params$Toll10Benhancer, sep = "/")
Toll10Benhancers = read.table(
  file = Toll10BenhancersFile,
  header = F,
  sep = "\t",
  quote = "",
  stringsAsFactors = F
)
Toll10Benhancers = Toll10Benhancers[, 1:3]

Toll9enhancersFile = paste(params$workingDir, params$Toll9enhancer, sep = "/")
Toll9enhancers = read.table(
  file = Toll9enhancersFile,
  header = F,
  sep = "\t",
  quote = "",
  stringsAsFactors = F
)
Toll9enhancers = Toll9enhancers[, 1:3]

# Rename columns
colnames(GD7enhancers) = c("Chr", "Start", "End")
GD7enhancers$Genotype = "Gd7"
colnames(Toll10Benhancers) = c("Chr", "Start", "End")
Toll10Benhancers$Genotype = "Toll10b"
colnames(Toll9enhancers) = c("Chr", "Start", "End")
Toll9enhancers$Genotype = "Toll9"

# Combine all enhancer data into one dataframe
Enhancers = rbind(GD7enhancers,Toll10Benhancers,Toll9enhancers)
Enhancers$Strand = "."
Enhancers$Chr = paste("chr",Enhancers$Chr, sep = "")
Enhancers$enhancerID = paste("region",1:nrow(Enhancers))
rownames(Enhancers) = Enhancers$enhancerID
```

```{r convert into genomic ranges}
# Convert dataframe to a genomic range
K27ac_peakData = makeGRangesFromDataFrame(
  K27acInfo,
  seqnames.field = "K27ac_Chr",
  start.field = "K27ac_Start",
  end.field = "K27ac_End",
  strand.field = "K27ac_Strand",
)
names(K27ac_peakData) = K27acInfo$K27ac_geneID

PRO_genes_peakData = makeGRangesFromDataFrame(PROseqGeneInfo,
                                              seqnames.field = "Gene_Chr",
                                              start.field = "Gene_Start",
                                              end.field = "Gene_End",
                                              strand.field = "Gene_Strand"
                                              )
names(PRO_genes_peakData) = PROseqGeneInfo$geneID

# Kanske kan anvÃ¤nda detta till att joina PROoch ATAC seq???
#test = data.frame(distanceToNearest(ATAC_peakData,PRO_genes_peakData,  ignore.strand=TRUE))

inSitu_peakData = makeGRangesFromDataFrame(inSitu.Data,
                                      seqnames.field = c("Chr","seqnames"), 
                                      start.field = c("Start","start"), 
                                      end.field = c("End","end"), 
                                      strand.field = "Strand", 
                                      keep.extra.columns = TRUE)
names(inSitu_peakData) = inSitu.Data$insituID


ATAC_peakData = makeGRangesFromDataFrame(ATACseqInfo,
                                      seqnames.field = "Chr", 
                                      start.field = "ATAC_Start", 
                                      end.field = "ATAC_End", 
                                      strand.field = "Strand")
names(ATAC_peakData) = ATACseqInfo$geneID


EnhancersInfo = Enhancers %>% 
  dplyr::select(Genotype,enhancerID) %>%  
  dplyr::rename(enhancerGenotype = Genotype)
 

EnhancersPeakData = makeGRangesFromDataFrame(Enhancers,
                                      seqnames.field = "Chr", 
                                      start.field = "Start", 
                                      end.field = "End", 
                                      strand.field = "Strand", 
                                      keep.extra.columns = TRUE)
names(EnhancersPeakData) = EnhancersPeakData$enhancerID

```

```{r join the data}
CBPpeak.ATAC.PRO  = inner_join(ATACseqInfo, PROseqInfo, by = "geneID")
CBPpeak.ATAC.PRO.CBP  = inner_join(CBPpeak.ATAC.PRO, CBPInfo, by = "geneID")

```


```{r find the overlap and adding annotation}
# Parameter???
overlap_ATAC_K27ac = data.frame(findOverlaps(ATAC_peakData,K27ac_peakData, maxgap = 0))
overlap_ATAC_K27ac$K27ac_geneID = names(K27ac_peakData)[overlap_ATAC_K27ac$subjectHits]
overlap_ATAC_K27ac$geneID = names(ATAC_peakData)[overlap_ATAC_K27ac$queryHits]
overlap_ATAC_K27ac = overlap_ATAC_K27ac %>% dplyr::select(K27ac_geneID,geneID )

CBPpeak.ATAC.PRO.CBP  = inner_join(CBPpeak.ATAC.PRO.CBP,  overlap_ATAC_K27ac)
CBPpeak.ATAC.PRO.CBP.K27ac =  inner_join(CBPpeak.ATAC.PRO.CBP,  K27acInfo, by="K27ac_geneID")


overlap_CBP_enhancers = data.frame(findOverlaps(ATAC_peakData,EnhancersPeakData, maxgap = 0))
overlap_CBP_enhancers$enhancerID = EnhancersPeakData$enhancerID[overlap_CBP_enhancers$subjectHits]
overlap_CBP_enhancers$geneID = names(ATAC_peakData)[overlap_CBP_enhancers$queryHits]
overlap_CBP_enhancers = overlap_CBP_enhancers %>% dplyr::select(enhancerID,geneID )


CBPpeak.ATAC.PRO.CBP.K27ac.enhancer = left_join(CBPpeak.ATAC.PRO.CBP.K27ac, overlap_CBP_enhancers)

CBPpeak.ATAC.PRO.CBP.K27ac.enhancer = left_join(CBPpeak.ATAC.PRO.CBP.K27ac.enhancer, EnhancersInfo)



overlap_CBP_inSitu = data.frame(findOverlaps(ATAC_peakData,inSitu_peakData, maxgap = 0))
overlap_CBP_inSitu$insituID = inSitu_peakData$insituID[overlap_CBP_inSitu$subjectHits]
overlap_CBP_inSitu$geneID = names(ATAC_peakData)[overlap_CBP_inSitu$queryHits]
overlap_CBP_inSitu = overlap_CBP_inSitu %>% dplyr::select(insituID,geneID )


CBPpeak.ATAC.PRO.CBP.K27ac.enhancer.inSitu = left_join(CBPpeak.ATAC.PRO.CBP.K27ac.enhancer, overlap_CBP_inSitu)

CBPpeak.ATAC.PRO.CBP.K27ac.enhancer.inSitu = left_join(CBPpeak.ATAC.PRO.CBP.K27ac.enhancer.inSitu, inSituInfo)

```

```{r adding annotation as 0 and 1, include=FALSE}


EnhancerInfo = CBPpeak.ATAC.PRO.CBP.K27ac.enhancer.inSitu
EnhancerInfo$geneID_unique = make.names(EnhancerInfo$geneID, unique = TRUE)



EnhancerInfo$enhancer=1
EnhancerInfo$enhancer[is.na(EnhancerInfo$enhancerGenotype)]  = 0 

EnhancerInfo$inSitu=0
EnhancerInfo$inSitu[EnhancerInfo$inSituLocation %in% c("dorsal_ectoderm_AISN_broad", "mesoderm_AISN", "ventral_ectoderm_AISN")] = 1

PE =  EnhancerInfo%>% filter(enhancer == 1) %>% ggplot(aes(PRO_Toll9, PRO_Toll10b, color = enhancerGenotype)) + geom_point()+ geom_density_2d(EnhancerInfo,mapping = aes(PRO_Toll9, PRO_Toll10b), color= "grey")
PI =  EnhancerInfo %>% filter(inSitu == 1) %>% ggplot(aes(PRO_Toll9, PRO_Toll10b, color = inSituLocation )) + geom_point()+ geom_density_2d(EnhancerInfo,mapping = aes(PRO_Toll9, PRO_Toll10b), color= "grey")
AE =  EnhancerInfo%>% filter(enhancer == 1) %>% ggplot(aes(ATAC_Toll9, ATAC_Toll10b, color = enhancerGenotype)) + geom_point()+ geom_density_2d(EnhancerInfo,mapping = aes(ATAC_Toll9, ATAC_Toll10b), color= "grey")
AI =  EnhancerInfo %>% filter(inSitu == 1) %>% ggplot(aes(ATAC_Toll9, ATAC_Toll10b, color = inSituLocation )) + geom_point()+ geom_density_2d(EnhancerInfo,mapping = aes(ATAC_Toll9, ATAC_Toll10b), color= "grey")
CE =  EnhancerInfo%>% filter(enhancer == 1) %>% ggplot(aes(CBP_Toll9, CBP_Toll10b, color = enhancerGenotype)) + geom_point()+ geom_density_2d(EnhancerInfo,mapping = aes(CBP_Toll9, CBP_Toll10b), color= "grey") 
CI =  EnhancerInfo %>% filter(inSitu == 1) %>% ggplot(aes(CBP_Toll9, CBP_Toll10b, color = inSituLocation )) + geom_point()+ geom_density_2d(EnhancerInfo,mapping = aes(CBP_Toll9, CBP_Toll10b), color= "grey")
KE =  EnhancerInfo%>% filter(enhancer == 1) %>% ggplot(aes(K27ac_Toll9, K27ac_Toll10b, color = enhancerGenotype)) + geom_point()+ geom_density_2d(EnhancerInfo,mapping = aes(K27ac_Toll9, K27ac_Toll10b), color= "grey")
KI =  EnhancerInfo %>% filter(inSitu == 1) %>% ggplot(aes(K27ac_Toll9, K27ac_Toll10b, color = inSituLocation )) + geom_point()+ geom_density_2d(EnhancerInfo,mapping = aes(K27ac_Toll9, K27ac_Toll10b), color= "grey")

```


```{r Trying UMAPR, filter after, better}
allEnhancerInfo2 = EnhancerInfo %>% dplyr::select(-geneID,-enhancerID,-enhancerGenotype,-insituID,-inSituLocation,-inSituScore,-enhancer,-inSitu,-K27ac_geneID, -Chr, -Strand, -Length, -ATAC_direction, -ATAC_factor, -ATAC_Start, -ATAC_End, -K27ac_Strand, -K27ac_Start, -K27ac_End, -K27ac_Length, -K27ac_direction, -K27ac_factor, -K27ac_Chr, -geneID_unique)

all_enhancer.data = allEnhancerInfo2[, grep("ATAC|PRO|CBP|K27ac", colnames(allEnhancerInfo2))]
embedding <- umap(all_enhancer.data)
new_embedding = left_join(EnhancerInfo,embedding)

#I think it is best not to filter
#new_embedding = new_embedding %>% filter(!is.na(enhancerGenotype))

new_embedding %>% 
  mutate(Species = new_embedding$enhancerGenotype) %>%
  ggplot(aes(UMAP1, UMAP2, color = Species)) + geom_point()


AllData = new_embedding %>% mutate(distance1 = abs(UMAP1))
AllData = AllData %>% mutate(distance2 = (UMAP1^2+UMAP2^2)^0.5)

# Plot density plot
FE =  AllData %>% filter(enhancer == 1) %>% ggplot(aes(UMAP1, UMAP2, color = enhancerGenotype )) + geom_point()+ geom_density_2d(AllData,mapping = aes(UMAP1, UMAP2), color= "grey")
FI =  AllData %>% filter(inSitu == 1) %>% ggplot(aes(UMAP1, UMAP2, color = inSituLocation )) + geom_point()+ geom_density_2d(AllData,mapping = aes(UMAP1, UMAP2), color= "grey")

FE + FI+ plot_layout(ncol = 1)

```

```{r plot this steps }

PE+PI+ plot_layout(ncol = 1)
AE+AI+ plot_layout(ncol = 1)
CE+CI+ plot_layout(ncol = 1)
KE+KI+ plot_layout(ncol = 1)
FE + FI+ plot_layout(ncol = 1)

PE+PI+AE+AI+CE+CI+KE+KI+FE+FI +plot_layout(ncol = 2)


```

### Assesment and selection

```{r roc analysis  , include=FALSE}
enhancerData = AllData

d1ROCi = getROCinfo(AllData,scoreColumn = "distance1", classColumn = "inSitu")
d2ROCi = getROCinfo(AllData,scoreColumn = "distance2", classColumn = "inSitu")
d1ROCe = getROCinfo(AllData,scoreColumn = "distance1", classColumn = "enhancer")
d2ROCe = getROCinfo(AllData,scoreColumn = "distance2", classColumn = "enhancer")

ROCplotInfo = rbind (d1ROCi$ROCplotInfo, d2ROCi$ROCplotInfo,
                     d1ROCe$ROCplotInfo, d2ROCe$ROCplotInfo)
ROCplotSummary = rbind(d1ROCi$Youden, d2ROCi$Youden,
                     d1ROCe$Youden, d2ROCe$Youden)

ggplot(ROCplotInfo, mapping =  aes(x = FPR, y = TPR, color = score))+
  geom_line()+
  geom_abline()+
  geom_point(data = ROCplotSummary, mapping = aes(y = TPR, x = FPR, color = score ), size = 4) +
    facet_grid(.~classifier)

```


# Enhancer selection


Based on ROC statistics  it is best if we include both UMAP1 and UMAP2 and that we look at the enhancer data 

```{r plot classification }
scores = ROCplotSummary %>% dplyr::filter(classifier == "enhancer" & score == "distance2")

AllData.enhancers = AllData %>% filter(distance2 > scores$YoudenCutoff)

AllData.enhancers = AllData.enhancers %>% replace_na(list(enhancerID = "None", enhancerGenotype = "None",insituID = "None",inSituLocation = "None", inSituScore = -1 ))


peakInfo = data.frame(ATAC_peakData)
peakInfo$geneID = names(ATAC_peakData)

AllData.enhancers = AllData.enhancers %>% dplyr::select(-geneID_unique)
AllData.enhancers = inner_join(peakInfo, AllData.enhancers, by = "geneID")
AllData.enhancers_cleaned = AllData.enhancers %>% dplyr::select(geneID, -seqnames, Chr, start, end, -width, -strand, -ATAC_Start, -ATAC_End, -Strand, Length, ATAC_factor, ATAC_direction, ATAC_Gd7,  ATAC_Toll10b, ATAC_Toll9, ATAC_Time, PRO_Gd7, PRO_Toll10b, PRO_Toll9, PRO_Time, CBP_Gd7,  CBP_Toll10b, CBP_Toll9, K27ac_geneID, -K27ac_Chr, -K27ac_Start, -K27ac_End, -K27ac_Strand, -K27ac_Length,  -K27ac_factor, -K27ac_direction, K27ac_Gd7, K27ac_Toll10b, K27ac_Toll9, enhancerID, enhancerGenotype, insituID, inSituLocation, inSituScore, enhancer, inSitu, UMAP1, UMAP2, distance1, distance2)


```


## Taken from combining_all_data

```{r merge all the data }
AllData_test = AllData.enhancers_cleaned %>% mutate(Enhancer_Gd7 =  ATAC_Gd7 + CBP_Gd7 + K27ac_Gd7)
AllData_test = AllData_test %>% mutate(Enhancer_Toll10b =  ATAC_Toll10b + CBP_Toll10b + K27ac_Toll10b)
AllData_test = AllData_test %>% mutate(Enhancer_Toll9 =  ATAC_Toll9 + CBP_Toll9 + K27ac_Toll9)

PRO_ATAC_CHiP_data = AllData_test %>%
  group_by(ATAC_factor, ATAC_direction) %>%
  summarise(
    PRO_Time = mean(PRO_Time),
    PRO_Gd7 = mean(PRO_Gd7),
    PRO_Toll10b = mean(PRO_Toll10b),
    PRO_Toll9 = mean(PRO_Toll9),
    ATAC_Time = mean(ATAC_Time),
    ATAC_Gd7 = mean(ATAC_Gd7),
    ATAC_Toll10b = mean(ATAC_Toll10b),
    ATAC_Toll9 = mean(ATAC_Toll9),
    K27ac_Gd7 = mean(K27ac_Gd7),
    K27ac_Toll10b = mean(K27ac_Toll10b),
    K27ac_Toll9 = mean(K27ac_Toll9),
    CBP_Gd7 = mean(CBP_Gd7),
    CBP_Toll10b = mean(CBP_Toll10b),
    CBP_Toll9 = mean(CBP_Toll9)
  )

enhancers.filtering = AllData.enhancers %>% dplyr::select(
  geneID,
  Chr,
  ATAC_Start, 
  ATAC_End,
  ATAC_factor,
  ATAC_direction,
  ATAC_Time,
  ATAC_Gd7,
  ATAC_Toll10b,
  ATAC_Toll9,
  PRO_Time,
  PRO_Gd7,
  PRO_Toll10b,
  PRO_Toll9,
  CBP_Gd7,
  CBP_Toll10b,
  CBP_Toll9,
  K27ac_geneID,
  K27ac_Gd7,
  K27ac_Toll10b,
  K27ac_Toll9,
  enhancerID,
  insituID
)

enhancers.filtering.unique = unique(enhancers.filtering)

AllData.saved = AllData_test


```


```{r save the data}
write.table(
  x = enhancers.filtering.unique,
  file = paste(params$workingDir, params$results_file , sep = "/"),
  quote = F,
  sep = "\t",
  row.names = F,
  col.names = T
)

```






