---
title: "Identifying the enhancer gene pairs"
author: "Jenny Jakobsson"
date: "11/19/2020"
output: html_document

params:
  workingDir: ~/Applied_Bioinformatics_Group_4/
  RscriptsDir: script/original_scripts/
  inSitu.data: data_johan/results/cleaned/inSitu/inSitu.Stark.dm6.tsv
  PROseq.data: results/altered_method/PROseq/PROseq.CBP.data.tsv
  ChIPseqDir: results/original_method/ChIPseq/
  CBP.data: CBP/ChIPseq.CBP.data.tsv
  K27ac.data: K27ac/ChIPseq.H3K27Ac.data.tsv
  ATACseq.data: results/original_method/ATACseq/ATACseq.data.tsv
  GD7enhancer: data_johan/results/cleaned/enhancers/h3k27ac_gd7_csaw_intersect.bed
  Toll10Benhancer: data_johan/results/cleaned/enhancers/h3k27ac_Toll10B_csaw_intersect.bed
  Toll9enhancer: data_johan/results/cleaned/enhancers/h3k27ac_Tollrm910_csaw_intersect.bed
  enhancerResult: results/altered_method/enhancers/enhancers.tsv
  Regions: results/altered_method/enhancers/regions_id.tsv
  results_file: results/altered_method/from_joint_peak/enhancers.tsv
---

```{r setup, include=FALSE}
# Importing the functions from the script
source(
  paste(
    params$workingDir,
    params$RscriptsDir,
    "ExpressionAnalysisFunctions.r",
    sep = ""
  )
)

# Install the needed packages (only needs to be run once)
#install.packages("tidyverse")
#install.packages("ROCit")
#install.packages("patchwork")
#BiocManager::install("heatmaps")

# Load the needed packages
library("heatmaps")
library(patchwork)
library(ROCit)
library(tidyverse)
library(GenomicRanges)
# Needed to knit files
knitr::opts_chunk$set(echo = TRUE)
```
# Combining the scores to predict enhancer gene pairs
The last step of the analysis now needs to be done. We want to combine our predicted enhancers with the differential expressed genes. 

## Load the required files
We load the ATAC-seq, ChIP-seq for CBP and H3K27ac and PRO-seq(per CBP peak) data. The also rename the columns to distinguish them. We then also load the external data, in Situ and enhancer data. This data is used to find true enhancers in our data.
```{r Loading the ATACseq data files}
# Load the ATAC seq loading scores
ATACseq.data = read.table(
  file = paste(params$workingDir, params$ATACseq.data, sep = ""),
  quote = "",
  sep = "\t",
  header = TRUE,
  stringsAsFactors = F
)
# Rename the columns
ATACseqInfo = ATACseq.data %>% dplyr::rename(
  "ATAC_Gd7" = Gd7,
  "ATAC_Toll10b" = Toll10b,
  "ATAC_Toll9" = Toll9,
  "ATAC_Time" = Time,
  "ATAC_Start" = Start,
  "ATAC_End" = End ,
  "ATAC_factor" = factor,
  "ATAC_direction" = direction,
)
```

```{r loading the ChIPseq data files}
#Load the ChIPseq data for CBP
CBPInfo =  read.table(
  file = paste(params$workingDir, params$ChIPseqDir, params$CBP.data, sep = ""),
  quote = "",
  sep = "\t",
  header = TRUE,
  stringsAsFactors = F
)
# Rename the columns
CBPInfo = CBPInfo %>%
  dplyr::select(geneID, Gd7, Toll10b, Toll9) %>%
  dplyr::rename("CBP_Gd7" = Gd7,
                "CBP_Toll10b" = Toll10b,
                "CBP_Toll9" = Toll9)

# Load the ChIPseq data for H3K27ac
K27acInfo =  read.table(
  file = paste(params$workingDir, params$ChIPseqDir, params$K27ac.data, sep = "/"),
  quote = "",
  sep = "\t",
  header = TRUE,
  stringsAsFactors = F
)
# Rename the columns
K27acInfo = K27acInfo %>% dplyr::rename(
  "K27ac_geneID" = geneID,
  "K27ac_Start" = Start,
  "K27ac_End" = End ,
  "K27ac_factor" = factor,
  "K27ac_direction" = direction,
  "K27ac_Gd7" = Gd7,
  "K27ac_Toll10b" = Toll10b,
  "K27ac_Toll9" = Toll9,
  "K27ac_Chr" = Chr,
  "K27ac_Strand" = Strand,
  "K27ac_Length" = Length
)
```

```{r Load the PROseq data files}
# Load the PROseq data per CBP peak
PROseq_file = paste(params$workingDir, params$PROseq.data, sep = "")

PROseq.data = read.table(
  file = PROseq_file ,
  quote = "",
  sep = "\t",
  header = TRUE,
  stringsAsFactors = F
)

# Rename headers.
PROseqInfo = PROseq.data %>%  dplyr::select(geneID, Gd7, Toll10b, Toll9, Time, factor, direction) %>% dplyr::rename(
  "PRO_Gd7" = Gd7,
  "PRO_Toll10b" = Toll10b,
  "PRO_Toll9" = Toll9,
  "PRO_Time" = Time,
  "Factor" = factor,
  "Direction" = direction
)
```

```{r Load the Insitu data files}
# Load in the inSitu data
inSituFile = paste(params$workingDir, params$inSitu.data, sep = "")
inSitu.Data = read.table(
  file = inSituFile,
  header = T,
  sep = "\t",
  quote = "",
  stringsAsFactors = F
)

# Need to transform the columns to match our data
inSitu.Data = inSitu.Data %>% dplyr::select(seqnames, start, end, stg4_6) %>% separate_rows(stg4_6, sep = "\\|") %>%
  separate(stg4_6, into = c("Location", "score"), sep = ";")
inSitu.Data = inSitu.Data %>% replace_na(list(score = 0, Location = "None"))
inSitu.Data$insituID = paste("region", 1:nrow(inSitu.Data))

# Rename the columns
inSituInfo = inSitu.Data %>% dplyr::select(insituID , Location, score) %>% dplyr::rename(inSituLocation = Location, inSituScore = score)
```

```{r Load the enhancer data files}
# Load the enhancer data for all mutants to compare the data to
GD7enhancersFile = paste(params$workingDir, params$GD7enhancer, sep = "")
GD7enhancers = read.table(
  file = GD7enhancersFile,
  header = F,
  sep = "\t",
  quote = "",
  stringsAsFactors = F
)
Toll10BenhancersFile = paste(params$workingDir, params$Toll10Benhancer, sep = "/")
Toll10Benhancers = read.table(
  file = Toll10BenhancersFile,
  header = F,
  sep = "\t",
  quote = "",
  stringsAsFactors = F
)
Toll9enhancersFile = paste(params$workingDir, params$Toll9enhancer, sep = "/")
Toll9enhancers = read.table(
  file = Toll9enhancersFile,
  header = F,
  sep = "\t",
  quote = "",
  stringsAsFactors = F
)
# Extract the relevant information
GD7enhancers = GD7enhancers[, 1:3]
Toll10Benhancers = Toll10Benhancers[, 1:3]
Toll9enhancers = Toll9enhancers[, 1:3]

# Rename columns
colnames(GD7enhancers) = c("Chr", "Start", "End")
GD7enhancers$Genotype = "Gd7"
colnames(Toll10Benhancers) = c("Chr", "Start", "End")
Toll10Benhancers$Genotype = "Toll10b"
colnames(Toll9enhancers) = c("Chr", "Start", "End")
Toll9enhancers$Genotype = "Toll9"

# Combine enhancer data into one dataframe
Enhancers = rbind(GD7enhancers, Toll10Benhancers, Toll9enhancers)
Enhancers$Strand = "."
Enhancers$Chr = paste("chr", Enhancers$Chr, sep = "")
Enhancers$enhancerID = paste("region", 1:nrow(Enhancers))
rownames(Enhancers) = Enhancers$enhancerID
```

## Finding the overlap between the data
Our ambition is to combine data from different sources. Since the data has been generated differently the regions will differ and we need to find the overlap between the different dataset. This can be done with the functionality of genomic ranges. 

### Converting into genomic ranges
```{r convert into genomic ranges}
# Convert dataframe to a genomic range
# ChIP-seq for H3K27ac
K27ac_peakData = makeGRangesFromDataFrame(
  K27acInfo,
  seqnames.field = "K27ac_Chr",
  start.field = "K27ac_Start",
  end.field = "K27ac_End",
  strand.field = "K27ac_Strand",
)
names(K27ac_peakData) = K27acInfo$K27ac_geneID

# In situ data
inSitu_peakData = makeGRangesFromDataFrame(
  inSitu.Data,
  seqnames.field = c("Chr", "seqnames"),
  start.field = c("Start", "start"),
  end.field = c("End", "end"),
  strand.field = "Strand",
  keep.extra.columns = TRUE
)
names(inSitu_peakData) = inSitu.Data$insituID

# ATAC-seq data
ATAC_peakData = makeGRangesFromDataFrame(
  ATACseqInfo,
  seqnames.field = "Chr",
  start.field = "ATAC_Start",
  end.field = "ATAC_End",
  strand.field = "Strand"
)
names(ATAC_peakData) = ATACseqInfo$geneID

# Enhancer data
EnhancersInfo = Enhancers %>%
  dplyr::select(Genotype, enhancerID) %>%
  dplyr::rename(enhancerGenotype = Genotype)
EnhancersPeakData = makeGRangesFromDataFrame(
  Enhancers,
  seqnames.field = "Chr",
  start.field = "Start",
  end.field = "End",
  strand.field = "Strand",
  keep.extra.columns = TRUE
)
names(EnhancersPeakData) = EnhancersPeakData$enhancerID
```

### Joining the incommon peak data
The ATAC-seq, PRO-seq and ChIP-seq for CBP are all based on the same regions, CBP peaks. Since they have the same regions, it can be joined directly.

```{r join the data}
CBPpeak.ATAC.PRO  = inner_join(ATACseqInfo, PROseqInfo, by = "geneID")
CBPpeak.ATAC.PRO.CBP  = inner_join(CBPpeak.ATAC.PRO, CBPInfo, by = "geneID")
```
### Finding overlap and joining
Here we actually find the overlap between the different data sets and join all of the data

```{r find the overlap and adding annotation}
# Overlap ATAC-seq and ChIP-seq H3K27ac data with genomic ranges
overlap_ATAC_K27ac = data.frame(findOverlaps(ATAC_peakData, K27ac_peakData, maxgap = 0))
# The output from findOverlaps needs to be reformatted to include the geneIDs we need to join
overlap_ATAC_K27ac$K27ac_geneID = names(K27ac_peakData)[overlap_ATAC_K27ac$subjectHits]
overlap_ATAC_K27ac$geneID = names(ATAC_peakData)[overlap_ATAC_K27ac$queryHits]
overlap_ATAC_K27ac = overlap_ATAC_K27ac %>% dplyr::select(K27ac_geneID, geneID)
# Join the IDs from the genomic range and then all ChIP-seq H3K27ac data
CBPpeak.ATAC.PRO.CBP  = inner_join(CBPpeak.ATAC.PRO.CBP,  overlap_ATAC_K27ac, by =
                                     "geneID")
CBPpeak.ATAC.PRO.CBP.K27ac =  inner_join(CBPpeak.ATAC.PRO.CBP,  K27acInfo, by =
                                           "K27ac_geneID")

#Add enhancer data the same way
overlap_CBP_enhancers = data.frame(findOverlaps(ATAC_peakData, EnhancersPeakData, maxgap = 0))
overlap_CBP_enhancers$enhancerID = EnhancersPeakData$enhancerID[overlap_CBP_enhancers$subjectHits]
overlap_CBP_enhancers$geneID = names(ATAC_peakData)[overlap_CBP_enhancers$queryHits]
overlap_CBP_enhancers = overlap_CBP_enhancers %>% dplyr::select(enhancerID, geneID)
CBPpeak.ATAC.PRO.CBP.K27ac.enhancer = left_join(CBPpeak.ATAC.PRO.CBP.K27ac, overlap_CBP_enhancers, by =
                                                  "geneID")
CBPpeak.ATAC.PRO.CBP.K27ac.enhancer = left_join(CBPpeak.ATAC.PRO.CBP.K27ac.enhancer, EnhancersInfo, by =
                                                  "enhancerID")

# Add inSitu data the same way
overlap_CBP_inSitu = data.frame(findOverlaps(ATAC_peakData, inSitu_peakData, maxgap = 0))
overlap_CBP_inSitu$insituID = inSitu_peakData$insituID[overlap_CBP_inSitu$subjectHits]
overlap_CBP_inSitu$geneID = names(ATAC_peakData)[overlap_CBP_inSitu$queryHits]
overlap_CBP_inSitu = overlap_CBP_inSitu %>% dplyr::select(insituID, geneID)
CBPpeak.ATAC.PRO.CBP.K27ac.enhancer.inSitu = left_join(CBPpeak.ATAC.PRO.CBP.K27ac.enhancer, overlap_CBP_inSitu, by =
                                                         "geneID")
CBPpeak.ATAC.PRO.CBP.K27ac.enhancer.inSitu = left_join(CBPpeak.ATAC.PRO.CBP.K27ac.enhancer.inSitu, inSituInfo, by =
                                                         "insituID")
```

## Finding the accuracy of our method by ROC curve
We now have our predicted enhancers. We want to see how accurate our scoring system works. This is done by using the external data, in situ and enhancer data, in a ROC curve.

### Finding true enhancers
We have joined the external data now we just need to convert it into a format we can input into the ROC curve function

```{r adding annotation as 0 and 1, include=FALSE}
# Convert a column into 1 and 0, 1 if annotation exists of that enhancers and 0 if not.
EnhancerInfo = CBPpeak.ATAC.PRO.CBP.K27ac.enhancer.inSitu
EnhancerInfo$geneID_unique = make.names(EnhancerInfo$geneID, unique = TRUE)
EnhancerInfo$enhancer = 1
EnhancerInfo$enhancer[is.na(EnhancerInfo$enhancerGenotype)]  = 0

# Same for the in Situ data. Only here we are only interested if it exists in one of the three mutants(tissues)
EnhancerInfo$inSitu = 0
EnhancerInfo$inSitu[EnhancerInfo$inSituLocation %in% c("dorsal_ectoderm_AISN_broad",
                                                       "mesoderm_AISN",
                                                       "ventral_ectoderm_AISN")] = 1
```

```{r plot the annotations}
# Here we plot the true positves dependent on which annotation and which data set it is.
PE =  EnhancerInfo %>% filter(enhancer == 1) %>% ggplot(aes(PRO_Toll9, PRO_Toll10b, color = enhancerGenotype)) +
  ggtitle("PRO-seq data for enhancer data") + geom_point() + geom_density_2d(EnhancerInfo,
                                                                             mapping = aes(PRO_Toll9, PRO_Toll10b),
                                                                             color = "grey")

PI =  EnhancerInfo %>% filter(inSitu == 1) %>% ggplot(aes(PRO_Toll9, PRO_Toll10b, color = inSituLocation)) +
  ggtitle("PRO-seq data for in Situ data") + geom_point() + geom_density_2d(EnhancerInfo,
                                                                            mapping = aes(PRO_Toll9, PRO_Toll10b),
                                                                            color = "grey")
AE =  EnhancerInfo %>% filter(enhancer == 1) %>% ggplot(aes(ATAC_Toll9, ATAC_Toll10b, color = enhancerGenotype)) + geom_point() + geom_density_2d(EnhancerInfo,
                                                                                                                                                  mapping = aes(ATAC_Toll9, ATAC_Toll10b),
                                                                                                                                                  color = "grey")
AI =  EnhancerInfo %>% filter(inSitu == 1) %>% ggplot(aes(ATAC_Toll9, ATAC_Toll10b, color = inSituLocation)) + geom_point() + geom_density_2d(EnhancerInfo,
                                                                                                                                              mapping = aes(ATAC_Toll9, ATAC_Toll10b),
                                                                                                                                              color = "grey")
CE =  EnhancerInfo %>% filter(enhancer == 1) %>% ggplot(aes(CBP_Toll9, CBP_Toll10b, color = enhancerGenotype)) + geom_point() + geom_density_2d(EnhancerInfo,
                                                                                                                                                mapping = aes(CBP_Toll9, CBP_Toll10b),
                                                                                                                                                color = "grey")
CI =  EnhancerInfo %>% filter(inSitu == 1) %>% ggplot(aes(CBP_Toll9, CBP_Toll10b, color = inSituLocation)) + geom_point() + geom_density_2d(EnhancerInfo,
                                                                                                                                            mapping = aes(CBP_Toll9, CBP_Toll10b),
                                                                                                                                            color = "grey")
KE =  EnhancerInfo %>% filter(enhancer == 1) %>% ggplot(aes(K27ac_Toll9, K27ac_Toll10b, color = enhancerGenotype)) + geom_point() + geom_density_2d(EnhancerInfo,
                                                                                                                                                    mapping = aes(K27ac_Toll9, K27ac_Toll10b),
                                                                                                                                                    color = "grey") + ggtitle("Enhancer data for H3K27ac")
KI =  EnhancerInfo %>% filter(inSitu == 1) %>% ggplot(aes(K27ac_Toll9, K27ac_Toll10b, color = inSituLocation)) + geom_point() + geom_density_2d(EnhancerInfo,
                                                                                                                                                mapping = aes(K27ac_Toll9, K27ac_Toll10b),
                                                                                                                                                color = "grey") + ggtitle("In Situ data for H3K27ac")
# Execute to output the plots
PE + PI + plot_layout(ncol = 1)
AE + AI + plot_layout(ncol = 1)
CE + CI + plot_layout(ncol = 1)
KE + KI + plot_layout(ncol = 1)
```
## Principal component analysis
We now combine the data to create a distance for all data. Principal component analysis once again reduces the dimensions and creates a loading score. These loading score are then used as a distance. The five first PCs are seen as the most important. Five distances are created. Distance one is just the absolute value of PC1. Distance 2 is the distance between PC1 and PC2 and so forth until distance five is the distance in the 5D space of PC1-5.   

```{r PCA analysis, include=FALSE}
# Select only the columns with all scores
allEnhancerInfo2 = EnhancerInfo %>% dplyr::select(-geneID,-enhancerID,-enhancerGenotype,-insituID,-inSituLocation,-inSituScore,-enhancer,-inSitu,-K27ac_geneID, -Chr, -Strand, -Length, -ATAC_direction, -ATAC_factor, -ATAC_Start, -ATAC_End, -K27ac_Strand, -K27ac_Start, -K27ac_End, -K27ac_Length, -K27ac_direction, -K27ac_factor, -K27ac_Chr, -geneID_unique, -Factor, -Direction)
rownames(allEnhancerInfo2) = EnhancerInfo$geneID_unique

# PCA to create a new score
mir.pca <- prcomp(allEnhancerInfo2, center = TRUE, scale = FALSE) 
e.var = (mir.pca[['sdev']]^2 / sum(mir.pca[['sdev']]^2))
e.var = as.data.frame( e.var )
e.var$PC = as.factor(1:nrow(e.var)) 
e.var$Variance = e.var$e.var*100
e.var$CumulativeVariance = cumsum(e.var$Variance)
qplot(PC, Variance, data=e.var, geom = c("point")
      , ylim = c(0,max(e.var$Variance)+5) 
      ,ylab = "Variance (%)")
# Extract the loadings of PC1-5
pctable = as.data.frame(mir.pca$x)[, 1:5]
pctable$geneID_unique = rownames(pctable)

#Join the loadings with the original data frame
AllData = left_join(EnhancerInfo,pctable, by="geneID_unique")
# Create the 5 different distances
AllData = AllData %>% mutate(distance5 = (PC1^2+PC2^2+PC3^2+PC4^2++PC5^2)^0.5)
AllData = AllData %>% mutate(distance4 = (PC1^2+PC2^2+PC3^2+PC4^2)^0.5)
AllData = AllData %>% mutate(distance3 = (PC1^2+PC2^2+PC3^2)^0.5)
AllData = AllData %>% mutate(distance2 = (PC1^2+PC2^2)^0.5)
AllData = AllData %>% mutate(distance1 = abs(PC1))
```

```{r plot this steps }
# Plots to visualize the first two PCs which contain the most variance
FE =  AllData %>% filter(enhancer == 1) %>% ggplot(aes(PC1, PC2, color = enhancerGenotype )) + geom_point()+ geom_density_2d(AllData,mapping = aes(PC1, PC2), color= "grey")
FI =  AllData %>% filter(inSitu == 1) %>% ggplot(aes(PC1, PC2, color = inSituLocation )) + geom_point()+ geom_density_2d(AllData,mapping = aes(PC1, PC2), color= "grey")
# Need to execute to output the plots
FE + FI+ plot_layout(ncol = 1)
```

## Assesment with ROC curves
We now have identified true positives in our data, by in situ and enhancer data respectively. 

```{r roc analysis  , include=FALSE}
# Use a function which extracts ROC curve info
enhancerData = AllData
d1ROCi = getROCinfo(AllData, scoreColumn = "distance1", classColumn = "inSitu")
d2ROCi = getROCinfo(AllData, scoreColumn = "distance2", classColumn = "inSitu")
d3ROCi = getROCinfo(AllData, scoreColumn = "distance3", classColumn = "inSitu")
d4ROCi = getROCinfo(AllData, scoreColumn = "distance4", classColumn = "inSitu")
d5ROCi = getROCinfo(AllData, scoreColumn = "distance5", classColumn = "inSitu")
d1ROCe = getROCinfo(AllData, scoreColumn = "distance1", classColumn = "enhancer")
d2ROCe = getROCinfo(AllData, scoreColumn = "distance2", classColumn = "enhancer")
d3ROCe = getROCinfo(AllData, scoreColumn = "distance3", classColumn = "enhancer")
d4ROCe = getROCinfo(AllData, scoreColumn = "distance4", classColumn = "enhancer")
d5ROCe = getROCinfo(AllData, scoreColumn = "distance5", classColumn = "enhancer")

ROCplotInfo = rbind (
  d1ROCi$ROCplotInfo,
  d2ROCi$ROCplotInfo,
  d3ROCi$ROCplotInfo,
  d4ROCi$ROCplotInfo,
  d5ROCi$ROCplotInfo,
  d1ROCe$ROCplotInfo,
  d2ROCe$ROCplotInfo,
  d3ROCe$ROCplotInfo,
  d4ROCe$ROCplotInfo,
  d5ROCe$ROCplotInfo
)
ROCplotSummary = rbind(
  d1ROCi$Youden,
  d2ROCi$Youden,
  d3ROCi$Youden,
  d4ROCi$Youden,
  d5ROCi$Youden,
  d1ROCe$Youden,
  d2ROCe$Youden,
  d3ROCe$Youden,
  d4ROCe$Youden,
  d5ROCe$Youden
)
```

```{r plot ROC}
qplot(
  PC,
  Variance,
  data = e.var,
  geom = c("point")
  ,
  ylim = c(0, max(e.var$Variance) + 5)
  ,
  ylab = "Variance (%)"
)

ggplot(ROCplotInfo, mapping =  aes(x = FPR, y = TPR, color = score)) +
  geom_line() +
  geom_abline() +
  geom_point(
    data = ROCplotSummary,
    mapping = aes(y = TPR, x = FPR, color = score),
    size = 4
  ) +
  facet_grid(. ~ classifier)
```
## Enhancer selection
Based on ROC statistics  it is best if we include all five PC. Therefore the distance5 is the score that is selected. The data is selected based on a Youden score. The enhancers below the cutoff will be classified as background enhancers.

```{r plot classification }
scores = ROCplotSummary %>% dplyr::filter(classifier == "enhancer" &
                                            score == "distance5")
AllData.enhancers = AllData
AllData.enhancers$Background = "Yes"
AllData.enhancers$Background[AllData$distance5 > scores$YoudenCutoff] = "No"
AllData.enhancers = AllData.enhancers %>% replace_na(
  list(
    enhancerID = "None",
    enhancerGenotype = "None",
    insituID = "None",
    inSituLocation = "None",
    inSituScore = -1
  )
)
peakInfo = data.frame(ATAC_peakData)
peakInfo$geneID = names(ATAC_peakData)
AllData.enhancers = AllData.enhancers %>% dplyr::select(-geneID_unique)
AllData.enhancers = inner_join(peakInfo, AllData.enhancers, by = "geneID")
```


## Combine and polish the results
### Combine the scores to a single enhanecers core
Combine the score of the ATAC-seq, ChIP-seq and PRO-seq data to a single score for each mutant. The bigger the score, the more important it is for that mutant.

```{r combine scores}
AllData_test = AllData.enhancers %>% mutate(Enhancer_Gd7 =  ATAC_Gd7 + CBP_Gd7 + K27ac_Gd7 +
                                              PRO_Gd7)
AllData_test = AllData_test %>% mutate(Enhancer_Toll10b =  ATAC_Toll10b + CBP_Toll10b + K27ac_Toll10b +
                                         PRO_Toll10b)
AllData_test = AllData_test %>% mutate(Enhancer_Toll9 =  ATAC_Toll9 + CBP_Toll9 + K27ac_Toll9 +
                                         PRO_Toll9)
AllData_test = AllData_test %>% mutate(Enhancer_Time =  ATAC_Time + PRO_Time)
```

### Save the data

```{r save the data}
enhancers.filtering = AllData_test %>% dplyr::select(
  geneID,
  Chr,
  ATAC_Start,
  ATAC_End,
  ATAC_factor,
  ATAC_direction,
  enhancerID,
  insituID,
  Enhancer_Gd7,
  Enhancer_Toll10b,
  Enhancer_Toll9,
  Enhancer_Time,
  Background
) %>% unique()

write.table(
  x = enhancers.filtering,
  file = paste(params$workingDir, params$results_file , sep = "/"),
  quote = F,
  sep = "\t",
  row.names = F,
  col.names = T
)

```






