---
title: "Identifying the enhancer gene pairs"
author: "Jenny Jakobsson"
date: "11/19/2020"
output: html_document

params:
  workingDir: ~/Applied_Bioinformatics_Group_4/
  RscriptsDir: script/original_scripts/
  inSitu.data: data_johan/results/cleaned/inSitu/inSitu.Stark.dm6.tsv
  #PROseqDir: results/original_method/PROseq
  PROseq.data: data_johan/results/cleaned/PROseq/PROseq.CBP.peak.tsv
  #PROseq.data: results/original_method/PROseq/PROseq.data.tsv
  ChIPseqDir: results/original_method/ChIPseq/
  CBP.data: CBP/ChIPseq.CBP.data.tsv
  K27ac.data: K27ac/ChIPseq.H3K27Ac.data.tsv
  ATACseq.data: results/original_method/ATACseq/ATACseq.data.tsv
  GD7enhancer: data_johan/results/cleaned/enhancers/h3k27ac_gd7_csaw_intersect.bed
  Toll10Benhancer: data_johan/results/cleaned/enhancers/h3k27ac_Toll10B_csaw_intersect.bed
  Toll9enhancer: data_johan/results/cleaned/enhancers/h3k27ac_Tollrm910_csaw_intersect.bed
  enhancerResult: results/altered_method/enhancers/enhancers.tsv
  Regions: results/altered_method/enhancers/regions_id.tsv
  
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
source(paste(params$workingDir, params$RscriptsDir,"ExpressionAnalysisFunctions.r", sep = ""))

#install.packages("tidyverse")
#install.packages("ROCit")
#install.packages("patchwork")
#BiocManager::install("heatmaps")
library("heatmaps")
library(patchwork)
library(ROCit)
library(tidyverse)
library(GenomicRanges)

```
# Combining the scores to predict enhancer gene pairs

## Load the required files
```{r Loading the ATACseq data files}
# Load the ATAC seq loading scores
ATACseq.data = read.table(
  file = paste(params$workingDir, params$ATACseq.data, sep = ""),
  quote = "",
  sep = "\t",
  header = TRUE,
  stringsAsFactors = F
)
# Rename the columns
ATACseqInfo = ATACseq.data %>% dplyr::rename(
  "ATAC_Gd7" = Gd7,
  "ATAC_Toll10b" = Toll10b,
  "ATAC_Toll9" = Toll9,
  "ATAC_Time" = Time,
  "ATAC_Start" = Start,
  "ATAC_End" = End ,
  "ATAC_factor" = factor,
  "ATAC_direction" = direction
)

```

```{r laoding the ChIPseq data files}
#Load the ChIPseq data for CBP
CBPInfo =  read.table(
  file = paste(params$workingDir, params$ChIPseqDir, params$CBP.data, sep = ""), 
  quote = "",
  sep = "\t",
  header = TRUE,
  stringsAsFactors = F
)
# Rename the mutant headers
CBPInfo = CBPInfo %>%
  dplyr::select(geneID, Gd7, Toll10b, Toll9) %>%
  dplyr::rename("CBP_Gd7" = Gd7,
                "CBP_Toll10b" = Toll10b,
                "CBP_Toll9" = Toll9)

# Load the ChIPseq data for H3K27ac
K27acInfo =  read.table(
  file = paste(params$workingDir,params$ChIPseqDir, params$K27ac.data, sep = "/"),
  quote = "",
  sep = "\t",
  header = TRUE,
  stringsAsFactors = F
)
# Rename the columns
K27acInfo = K27acInfo %>% dplyr::rename(
  "K27ac_geneID" = geneID,
  "K27ac_Start" = Start,
  "K27ac_End" = End ,
  "K27ac_factor" = factor,
  "K27ac_direction" = direction,
  "K27ac_Gd7" = Gd7,
  "K27ac_Toll10b" = Toll10b,
  "K27ac_Toll9" = Toll9,
  "K27ac_Chr" = Chr,
  "K27ac_Strand" = Strand,
  "K27ac_Length" = Length
)
```

```{r Load the PROseq data files}
# Load the PROseq data
PROseq_file = paste(params$workingDir, params$PROseq.data, sep = "")

PROseq.data = read.table(
  file = PROseq_file ,
  quote = "",
  sep = "\t",
  header = TRUE,
  stringsAsFactors = F
)

# Rename headers.
PROseqInfo = PROseq.data %>%  dplyr::select(geneID, Gd7, Toll10b, Toll9, Time) %>% dplyr::rename(
  "PRO_Gd7" = Gd7,
  "PRO_Toll10b" = Toll10b,
  "PRO_Toll9" = Toll9,
  "PRO_Time" = Time
)

PROseq.data %>% dplyr::select(geneID,Chr,Start,End,Strand,factor, direction, annotation,annotation2,PC1,PC2,PC3) %>% distinct()  %>%  filter( factor != "Time") %>% ggplot(mapping = aes(x = PC2, y = PC3, color = factor, shape = direction )) + geom_point() 


annotation2 =  PROseqInfo %>% filter(factor %in% c("Toll10b","Toll9","Gd7") & annotation2 == "DE") %>%
  dplyr::select(geneID, annotation, distance, factor, direction ) %>%
  distinct()


PROseqData = PROseqInfo2 %>% dplyr::select(geneID,Chr,Start,End,Strand,factor, direction, annotation,annotation2, Gd7,Toll10b,Toll9) %>% distinct()

PROseqData.PI = PROseqInfo2 %>% group_by(geneID) %>% summarise(PI = max(PI))
colnames(PROseqData)[6] = paste("PROseq", colnames(PROseqData)[6], sep = "_")
colnames(PROseqData)[7] = paste("PROseq", colnames(PROseqData)[7], sep = "_")
colnames(PROseqData)[8] = paste("PROseq", colnames(PROseqData)[8], sep = "_")
colnames(PROseqData)[9] = paste("PROseq", colnames(PROseqData)[9], sep = "_")

PROseqData.DE  = PROseqData %>% filter(PROseq_annotation2 == "DE" & PROseq_factor != "Time") 

PROseqData.DE = inner_join(PROseqData.DE, PROseqData.PI)
PROseqData.DE = PROseqData.DE %>% mutate(Chr = paste("chr",Chr, sep = ""))

rownames(PROseqData.DE) = PROseqData.DE$geneID
```

```{r Load the Insitu data files}
# Load in the inSitu data
inSituFile = paste(params$workingDir, params$inSitu.data, sep = "")
inSitu.Data = read.table(
  file = inSituFile,
  header = T,
  sep = "\t",
  quote = "",
  stringsAsFactors = F
)

# Need to transform the columns to match our data
inSitu.Data = inSitu.Data %>% dplyr::select(seqnames, start, end, stg4_6) %>% separate_rows(stg4_6, sep = "\\|") %>%
  separate(stg4_6, into = c("Location", "score"), sep = ";")
inSitu.Data = inSitu.Data %>% replace_na(list(score = 0, Location = "None"))
inSitu.Data$insituID = paste("region", 1:nrow(inSitu.Data))

# Rename the columns
inSituInfo = inSitu.Data %>% dplyr::select(insituID , Location, score) %>% dplyr::rename(inSituLocation = Location, inSituScore = score)
```

```{r Load the enhancer data files}
# Load the enhancer data for all mutants to compare the data to
GD7enhancersFile = paste(params$workingDir, params$GD7enhancer, sep = "")
GD7enhancers = read.table(
  file = GD7enhancersFile,
  header = F,
  sep = "\t",
  quote = "",
  stringsAsFactors = F
)
GD7enhancers = GD7enhancers[, 1:3]

Toll10BenhancersFile = paste(params$workingDir, params$Toll10Benhancer, sep = "/")
Toll10Benhancers = read.table(
  file = Toll10BenhancersFile,
  header = F,
  sep = "\t",
  quote = "",
  stringsAsFactors = F
)
Toll10Benhancers = Toll10Benhancers[, 1:3]

Toll9enhancersFile = paste(params$workingDir, params$Toll9enhancer, sep = "/")
Toll9enhancers = read.table(
  file = Toll9enhancersFile,
  header = F,
  sep = "\t",
  quote = "",
  stringsAsFactors = F
)
Toll9enhancers = Toll9enhancers[, 1:3]

# Rename columns
colnames(GD7enhancers) = c("Chr", "Start", "End")
GD7enhancers$Genotype = "Gd7"
colnames(Toll10Benhancers) = c("Chr", "Start", "End")
Toll10Benhancers$Genotype = "Toll10b"
colnames(Toll9enhancers) = c("Chr", "Start", "End")
Toll9enhancers$Genotype = "Toll9"

# Combine all enhancer data into one dataframe
Enhancers = rbind(GD7enhancers,Toll10Benhancers,Toll9enhancers)
Enhancers$Strand = "."
Enhancers$Chr = paste("chr",Enhancers$Chr, sep = "")
Enhancers$enhancerID = paste("region",1:nrow(Enhancers))
rownames(Enhancers) = Enhancers$enhancerID
```

```{r convert into genomic ranges}
# Convert dataframe to a genomic range
K27ac_peakData = makeGRangesFromDataFrame(
  K27acInfo,
  seqnames.field = "K27ac_Chr",
  start.field = "K27ac_Start",
  end.field = "K27ac_End",
  strand.field = "K27ac_Strand",
)
names(K27ac_peakData) = K27acInfo$K27ac_geneID

inSitu_peakData = makeGRangesFromDataFrame(inSitu.Data,
                                      seqnames.field = c("Chr","seqnames"), 
                                      start.field = c("Start","start"), 
                                      end.field = c("End","end"), 
                                      strand.field = "Strand", 
                                      keep.extra.columns = TRUE)
names(inSitu_peakData) = inSitu.Data$insituID


ATAC_peakData = makeGRangesFromDataFrame(ATACseqInfo,
                                      seqnames.field = "Chr", 
                                      start.field = "ATAC_Start", 
                                      end.field = "ATAC_End", 
                                      strand.field = "Strand")
names(ATAC_peakData) = ATACseqInfo$geneID


EnhancersInfo = Enhancers %>% 
  dplyr::select(Genotype,enhancerID) %>%  
  dplyr::rename(enhancerGenotype = Genotype)
 

EnhancersPeakData = makeGRangesFromDataFrame(Enhancers,
                                      seqnames.field = "Chr", 
                                      start.field = "Start", 
                                      end.field = "End", 
                                      strand.field = "Strand", 
                                      keep.extra.columns = TRUE)
names(EnhancersPeakData) = EnhancersPeakData$enhancerID

```

```{r join the data}
CBPpeak.ATAC.PRO  = inner_join(ATACseqInfo, PROseqInfo, by = "geneID")
CBPpeak.ATAC.PRO.CBP  = inner_join(CBPpeak.ATAC.PRO, CBPInfo, by = "geneID")

```


```{r find the overlap and adding annotation}
# Parameter???
overlap_ATAC_K27ac = data.frame(findOverlaps(ATAC_peakData,K27ac_peakData, maxgap = 0))
overlap_ATAC_K27ac$K27ac_geneID = names(K27ac_peakData)[overlap_ATAC_K27ac$subjectHits]
overlap_ATAC_K27ac$geneID = names(ATAC_peakData)[overlap_ATAC_K27ac$queryHits]
overlap_ATAC_K27ac = overlap_ATAC_K27ac %>% dplyr::select(K27ac_geneID,geneID )

CBPpeak.ATAC.PRO.CBP  = inner_join(CBPpeak.ATAC.PRO.CBP,  overlap_ATAC_K27ac)
CBPpeak.ATAC.PRO.CBP.K27ac =  inner_join(CBPpeak.ATAC.PRO.CBP,  K27acInfo, by="K27ac_geneID")


overlap_CBP_enhancers = data.frame(findOverlaps(ATAC_peakData,EnhancersPeakData, maxgap = 0))
overlap_CBP_enhancers$enhancerID = EnhancersPeakData$enhancerID[overlap_CBP_enhancers$subjectHits]
overlap_CBP_enhancers$geneID = names(ATAC_peakData)[overlap_CBP_enhancers$queryHits]
overlap_CBP_enhancers = overlap_CBP_enhancers %>% dplyr::select(enhancerID,geneID )


CBPpeak.ATAC.PRO.CBP.K27ac.enhancer = left_join(CBPpeak.ATAC.PRO.CBP.K27ac, overlap_CBP_enhancers)

CBPpeak.ATAC.PRO.CBP.K27ac.enhancer = left_join(CBPpeak.ATAC.PRO.CBP.K27ac.enhancer, EnhancersInfo)



overlap_CBP_inSitu = data.frame(findOverlaps(ATAC_peakData,inSitu_peakData, maxgap = 0))
overlap_CBP_inSitu$insituID = inSitu_peakData$insituID[overlap_CBP_inSitu$subjectHits]
overlap_CBP_inSitu$geneID = names(ATAC_peakData)[overlap_CBP_inSitu$queryHits]
overlap_CBP_inSitu = overlap_CBP_inSitu %>% dplyr::select(insituID,geneID )


CBPpeak.ATAC.PRO.CBP.K27ac.enhancer.inSitu = left_join(CBPpeak.ATAC.PRO.CBP.K27ac.enhancer, overlap_CBP_inSitu)

CBPpeak.ATAC.PRO.CBP.K27ac.enhancer.inSitu = left_join(CBPpeak.ATAC.PRO.CBP.K27ac.enhancer.inSitu, inSituInfo)

```

```{r adding annotation as 0 and 1, include=FALSE}


EnhancerInfo = CBPpeak.ATAC.PRO.CBP.K27ac.enhancer.inSitu
EnhancerInfo$geneID_unique = make.names(EnhancerInfo$geneID, unique = TRUE)



EnhancerInfo$enhancer=1
EnhancerInfo$enhancer[is.na(EnhancerInfo$enhancerGenotype)]  = 0 

EnhancerInfo$inSitu=0
EnhancerInfo$inSitu[EnhancerInfo$inSituLocation %in% c("dorsal_ectoderm_AISN_broad", "mesoderm_AISN", "ventral_ectoderm_AISN")] = 1

PE =  EnhancerInfo%>% filter(enhancer == 1) %>% ggplot(aes(PRO_Toll9, PRO_Toll10b, color = enhancerGenotype)) + geom_point()+ geom_density_2d(EnhancerInfo,mapping = aes(PRO_Toll9, PRO_Toll10b), color= "grey")
PI =  EnhancerInfo %>% filter(inSitu == 1) %>% ggplot(aes(PRO_Toll9, PRO_Toll10b, color = inSituLocation )) + geom_point()+ geom_density_2d(EnhancerInfo,mapping = aes(PRO_Toll9, PRO_Toll10b), color= "grey")
AE =  EnhancerInfo%>% filter(enhancer == 1) %>% ggplot(aes(ATAC_Toll9, ATAC_Toll10b, color = enhancerGenotype)) + geom_point()+ geom_density_2d(EnhancerInfo,mapping = aes(ATAC_Toll9, ATAC_Toll10b), color= "grey")
AI =  EnhancerInfo %>% filter(inSitu == 1) %>% ggplot(aes(ATAC_Toll9, ATAC_Toll10b, color = inSituLocation )) + geom_point()+ geom_density_2d(EnhancerInfo,mapping = aes(ATAC_Toll9, ATAC_Toll10b), color= "grey")
CE =  EnhancerInfo%>% filter(enhancer == 1) %>% ggplot(aes(CBP_Toll9, CBP_Toll10b, color = enhancerGenotype)) + geom_point()+ geom_density_2d(EnhancerInfo,mapping = aes(CBP_Toll9, CBP_Toll10b), color= "grey") 
CI =  EnhancerInfo %>% filter(inSitu == 1) %>% ggplot(aes(CBP_Toll9, CBP_Toll10b, color = inSituLocation )) + geom_point()+ geom_density_2d(EnhancerInfo,mapping = aes(CBP_Toll9, CBP_Toll10b), color= "grey")
KE =  EnhancerInfo%>% filter(enhancer == 1) %>% ggplot(aes(K27ac_Toll9, K27ac_Toll10b, color = enhancerGenotype)) + geom_point()+ geom_density_2d(EnhancerInfo,mapping = aes(K27ac_Toll9, K27ac_Toll10b), color= "grey")
KI =  EnhancerInfo %>% filter(inSitu == 1) %>% ggplot(aes(K27ac_Toll9, K27ac_Toll10b, color = inSituLocation )) + geom_point()+ geom_density_2d(EnhancerInfo,mapping = aes(K27ac_Toll9, K27ac_Toll10b), color= "grey")

```


```{r PCA analysis, include=FALSE}

allEnhancerInfo2 = EnhancerInfo %>% dplyr::select(-geneID,-enhancerID,-enhancerGenotype,-insituID,-inSituLocation,-inSituScore,-enhancer,-inSitu,-K27ac_geneID, -Chr, -Strand, -Length, -ATAC_direction, -ATAC_factor, -ATAC_Start, -ATAC_End, -K27ac_Strand, -K27ac_Start, -K27ac_End, -K27ac_Length, -K27ac_direction, -K27ac_factor, -K27ac_Chr, -geneID_unique)

rownames(allEnhancerInfo2) = EnhancerInfo$geneID_unique

mir.pca <- prcomp(allEnhancerInfo2, center = TRUE, scale = FALSE) 
e.var = (mir.pca[['sdev']]^2 / sum(mir.pca[['sdev']]^2))
e.var = as.data.frame( e.var )
e.var$PC = as.factor(1:nrow(e.var)) 
e.var$Variance = e.var$e.var*100
e.var$CumulativeVariance = cumsum(e.var$Variance)
qplot(PC, Variance, data=e.var, geom = c("point")
      , ylim = c(0,max(e.var$Variance)+5) 
      ,ylab = "Variance (%)")

pctable = as.data.frame(mir.pca$x)[, 1:5]

pctable$geneID_unique = rownames(pctable)

AllData = left_join(EnhancerInfo,pctable)

AllData = AllData %>% mutate(distance5 = (PC1^2+PC2^2+PC3^2+PC4^2++PC5^2)^0.5)
AllData = AllData %>% mutate(distance4 = (PC1^2+PC2^2+PC3^2+PC4^2)^0.5)
AllData = AllData %>% mutate(distance3 = (PC1^2+PC2^2+PC3^2)^0.5)
AllData = AllData %>% mutate(distance2 = (PC1^2+PC2^2)^0.5)
AllData = AllData %>% mutate(distance1 = abs(PC1))

FE =  AllData %>% filter(enhancer == 1) %>% ggplot(aes(PC1, PC2, color = enhancerGenotype )) + geom_point()+ geom_density_2d(AllData,mapping = aes(PC1, PC2), color= "grey")
FI =  AllData %>% filter(inSitu == 1) %>% ggplot(aes(PC1, PC2, color = inSituLocation )) + geom_point()+ geom_density_2d(AllData,mapping = aes(PC1, PC2), color= "grey")
```

```{r plot this steps }

PE+PI+ plot_layout(ncol = 1)
AE+AI+ plot_layout(ncol = 1)
CE+CI+ plot_layout(ncol = 1)
KE+KI+ plot_layout(ncol = 1)
FE + FI+ plot_layout(ncol = 1)

PE+PI+AE+AI+CE+CI+KE+KI+FE+FI +plot_layout(ncol = 2)


```


### Assesment and selection

```{r roc analysis  , include=FALSE}
enhancerData = AllData

d1ROCi = getROCinfo(AllData,scoreColumn = "distance1", classColumn = "inSitu")
d2ROCi = getROCinfo(AllData,scoreColumn = "distance2", classColumn = "inSitu")
d3ROCi = getROCinfo(AllData,scoreColumn = "distance3", classColumn = "inSitu")
d4ROCi = getROCinfo(AllData,scoreColumn = "distance4", classColumn = "inSitu")
d5ROCi = getROCinfo(AllData,scoreColumn = "distance5", classColumn = "inSitu")
d1ROCe = getROCinfo(AllData,scoreColumn = "distance1", classColumn = "enhancer")
d2ROCe = getROCinfo(AllData,scoreColumn = "distance2", classColumn = "enhancer")
d3ROCe = getROCinfo(AllData,scoreColumn = "distance3", classColumn = "enhancer")
d4ROCe = getROCinfo(AllData,scoreColumn = "distance4", classColumn = "enhancer")
d5ROCe = getROCinfo(AllData,scoreColumn = "distance5", classColumn = "enhancer")



ROCplotInfo = rbind (d1ROCi$ROCplotInfo, d2ROCi$ROCplotInfo, d3ROCi$ROCplotInfo,d4ROCi$ROCplotInfo, d5ROCi$ROCplotInfo,
                     d1ROCe$ROCplotInfo, d2ROCe$ROCplotInfo, d3ROCe$ROCplotInfo,d4ROCe$ROCplotInfo, d5ROCe$ROCplotInfo)
ROCplotSummary = rbind(d1ROCi$Youden, d2ROCi$Youden, d3ROCi$Youden,d4ROCi$Youden, d5ROCi$Youden,
                     d1ROCe$Youden, d2ROCe$Youden, d3ROCe$Youden,d4ROCe$Youden, d5ROCe$Youden)


```

```{r plot ROC}

ROCplotSummary

qplot(PC, Variance, data=e.var, geom = c("point")
      , ylim = c(0,max(e.var$Variance)+5)
      ,ylab = "Variance (%)")


ggplot(ROCplotInfo, mapping =  aes(x = FPR, y = TPR, color = score))+
  geom_line()+
  geom_abline()+
  geom_point(data = ROCplotSummary, mapping = aes(y = TPR, x = FPR, color = score ), size = 4) +
    facet_grid(.~classifier)

```
# Enhancer selection


Based on ROC statistics  it is best if we include all five PC  

```{r plot classification }


scores = ROCplotSummary %>% dplyr::filter(classifier == "enhancer" & score == "distance5")

AllData.enhancers = AllData %>% filter(distance5 > scores$YoudenCutoff)

AllData.enhancers = AllData.enhancers %>% replace_na(list(enhancerID = "None", enhancerGenotype = "None",insituID = "None",inSituLocation = "None", inSituScore = -1 ))


peakInfo = data.frame(ATAC_peakData)
peakInfo$geneID = names(ATAC_peakData)

AllData.enhancers = AllData.enhancers %>% dplyr::select(-geneID_unique)
AllData.enhancers = inner_join(peakInfo, AllData.enhancers, by = "geneID")
AllData.enhancers_cleaned = AllData.enhancers %>% dplyr::select(geneID, -seqnames, Chr, start, end, -width, -strand, -ATAC_Start, -ATAC_End, -Strand, Length, ATAC_factor, ATAC_direction, ATAC_Gd7,  ATAC_Toll10b, ATAC_Toll9, ATAC_Time, PRO_Gd7, PRO_Toll10b, PRO_Toll9, PRO_Time, CBP_Gd7,  CBP_Toll10b, CBP_Toll9, K27ac_geneID, -K27ac_Chr, -K27ac_Start, -K27ac_End, -K27ac_Strand, -K27ac_Length,  -K27ac_factor, -K27ac_direction, K27ac_Gd7, K27ac_Toll10b, K27ac_Toll9, enhancerID, enhancerGenotype, insituID, inSituLocation, inSituScore, enhancer, inSitu, PC1, PC2, PC3, PC4, PC5, distance1, distance2, distance3, distance4, distance5)

fileName = paste(params$workingDir, params$enhancerResult, sep = "")

#write.table(x = AllData.enhancers_cleaned, file = fileName, sep = "\t", quote = F, row.names = F, col.names = T )

regions_id = AllData.enhancer %>% dplyr::select(geneID)
regionsFile = paste(params$workingDir, params$Regions, sep="")
#write.table(x = regions_id, file = fileName, sep = "\t", quote = F, row.names = F, col.names = T )

```











