---
title: "R Notebook"
output: html_notebook

params:
  workingDir: /Users/johanreimegard/git/flyEmbryogenesis
  #workingDir: /Users/roshan/projects/tollTranscription/m_mannervik_19
  synonymsFileIn: information/fb_synonym_fb_2019_06.tsv  
  knownTargetsFile: information/pipe_Toll_known_Targets.txt
  EisenInfoFile: Eisen-ChIP_data-info.txt
  ZetlingerInfoFile: Zetlinger-ChIP-DATA-info.txt
  WieshausInfoFile: Weishaus-ATAC_data-info.txt
  ATACseqFile: ATAC_seq_mapped_dm6.updated.tab.txt
  dataDir: data/PROseq

---

# Fixing the bamFiles

```{r}
library(tidyverse)

sampleInfoFile  = paste(params$workingDir,"information", params$ATACseqFile , sep = "/") 

sampleInfo = read.table(file = sampleInfoFile, header = T, sep = "\t", quote = "", stringsAsFactors = F)

```


##Create the correct genrichFile

```{r generich ATAC seq }





Genotypes = unique(sampleInfo$Genotype)

blackListFile = "/proj/snic2019-30-14/private/genomes/dm6/dm6-blacklist.v2.bed"



GenerichCalls = data.frame(test = 1:20, call = "test" ,stringsAsFactors = F)
counts = 1
for (j in 1:length(Genotypes)){
  experiment = sampleInfo %>%  filter( Genotype ==Genotypes[j] )
    if(nrow(experiment) > 0){
      experiment$bamFiles = gsub(pattern = "sorted.bam","sorted.name.bam", experiment$bamFiles)

        GenerichCall = paste( "Genrich",
                            "-r -y -g 20",
                            paste( "-f",paste("Genrich",paste(Genotypes[j],"genrich.encode.narrowPeak.log",sep = "_"), sep  ="/")),
                            paste( "-k",paste("Genrich",paste(Genotypes[j],"genrich.encode.narrowPeak.pileups",sep = "_"), sep  ="/")),
                            paste( "-b",paste("Genrich",paste(Genotypes[j],"genrich.encode.narrowPeak.intervals",sep = "_"), sep  ="/")),
                            paste( "-R",paste("Genrich",paste(Genotypes[j],"genrich.encode.narrowPeak.pcrDuplicates",sep = "_"), sep  ="/")),
                            paste( "-E",blackListFile, sep = " "),
                            paste( "-t",paste(experiment$bamFiles, collapse = ","), sep  =" "),
                            paste( "-o",paste("Genrich",paste(Genotypes[j],"genrich.encode.narrowPeak",sep = "_"), sep  ="/")),
                            sep = " ")
      GenerichCall
      GenerichCalls[counts ,2 ] = GenerichCall
      counts = counts + 1
    }
    
  }

GenerichCalls = GenerichCalls[1:11,2]

write.table(file = "ATACseq.genrich.sh", x = GenerichCalls, quote = F, row.names = F, col.names = F)
info = data.frame(calls = GenerichCalls)

dim(info)


as.data.frame(test =GenerichCalls)

fileConn<-file("Genrich.ATACseq1.sh")
writeLines(c("cd /proj/snic2019-30-14/nobackup"), fileConn)
for(i in 1:length(GenerichCalls)){
  
writeLines(GenerichCalls[[i]], fileConn)
 
  
}

close(fileConn)



    experiment = ChIPseq %>%  filter(Antibody == Antibodies[5] & Genotype ==Genotypes[2] )

    nrow(experiment)







```

```{r run featureCount on chipseq data}


#      featureCounts -F SAF -a Genrich/K27me3.genrich.encode.narrowPeak.merged.SAF -o ChIPseq/K27me3.counts.regions.txt ChIP_seq_mapped_dm6/bam_dmel/gd7_K27me3_2_R1.dmel.bowtie2.sorted.bam  ChIP_seq_mapped_dm6/bam_dmel/gd7_K27me3_1_R1.dmel.bowtie2.sorted.bam

featureCountCalls = data.frame(test = 1:20, call = "test" ,stringsAsFactors = F)
counts = 1
i = 1
for (i in 1:length(Antibodies)){
  experiment = ChIPseq %>%  filter(Antibody == Antibodies[i])
  input = ChIPseq %>%  filter(Antibody == "Input")
  
  if(nrow(experiment) > 0){
    featureCountCall = paste( "featureCounts  -F SAF -T 5",
                              paste( "-a",paste("Genrich",paste(Antibodies[i],"genrich.encode.narrowPeak.merged.SAF",sep = "."), sep  ="/"),  sep  =" "),
                              paste( "-o",paste("ChIPseq",paste(Antibodies[i],"featureCount.count.tab.txt",sep = "."), sep  ="/"), sep  =" "),
                              paste(experiment$bamFiles, collapse = " "),
                              paste(input$bamFiles, collapse = " "), sep = " ") 
    featureCountCalls[counts ,2 ] = featureCountCall
    counts = counts + 1
  }
  
}

featureCountCalls = featureCountCalls[c(1, 3:5),2]

write.table(file = "ChIPseq.featureCount.sh", x = featureCountCalls, quote = F, row.names = F, col.names = F)














```


Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Cmd+Option+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Cmd+Shift+K* to preview the HTML file).
