---
title: "Joint seq analysis"
author: "Johan"
date: "10/25/2019"
html_document:
    toc: true
    toc_depth: 2

df_print: paged

params:
  workingDir: /Users/johanreimegard/git/flyEmbryogenesis
  proSeqDir: results/PROseq
  proSeq.data: PROseq.data.tsv
  proSeq.data.PPC: PROseq.data.PPC.tsv
  ATACSeqDir: results/ATACseq
  ATACSeq.data: ATACseq.data.raw.tsv
  k27ac.data: K27ac.ChIPseq.peak.pattern.raw.tsv
  CBP.data: ATAC-CBP_dm6.ChIPseq.peak.pattern.raw.tsv
  chipSeqDir: results/chipSeq
  RscriptsDir: R_Analysis/Rscripts  
  dataDir: data/ATACseq
  sampleInfo: information/ChIP_ATAC_sampleInfo.tab.txt
  count_ATAC: ATAC.peak.featureCount.count.tab.txt 
  Method: ATAC-Seq
  Antibody: K27ac
  resultDir: results/multi
  bestTargets: information/TopTargets.txt


  RfunctionsDirectory: /Users/johanreimegard/GoogleDrive/git/RNAmappingPipeline/R
  
  
---






```{r setup, include=FALSE}

source(paste(params$RfunctionsDirectory,"ExpressionAnalysisFunctions.R",sep = "/"))
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)

library(gplots)

library(knitr)
#install.packages("kableExtra")
library(kableExtra)
library("viridis")     

#install.packages("heatmap3")

library(heatmap3)

#install.packages("ggrepel")
library(ggrepel)

library(tsne)


gtfFile = paste(params$workingDir, "annotations/Drosophila_melanogaster.BDGP6.28.99.gtf", sep = "/")



```










## Loading and merging data 

```{r PROseq  CDS info}





PROseqInfo = read.table( file = paste( params$workingDir, params$proSeqDir, params$proSeq.data,
                          sep = "/"),
                         quote = "", sep = "\t", 
                         header = TRUE, stringsAsFactors = F) 



PROseqInfo2 = PROseqInfo %>% dplyr::select(geneID, PI,GBClog,mutant, time,replicate)

AllData2 = read.table(file = paste(params$workingDir,"results/combined.AllData.withMetaData.ATACnaive.txt" ,sep = "/")
                     ,header = T, sep = "\t")

ATACpeakInfo2 = peakInfo %>% dplyr::select(Geneid, Start, End) %>% 
  rename(Geneid = "ATACseq_Geneid",
         Start = "ATACseq_Start",
         End = "ATACseq_End" )



geneInfo = PROseqInfo %>% dplyr::select(geneID, Chr,Start, End) %>% distinct()
AllData3 = inner_join(ATACpeakInfo2, AllData2)
AllData3 = inner_join(geneInfo, AllData3)


write.table(file = paste(params$workingDir,"results/combined.AllData.withMetaData.ATACnaive.txt" ,sep = "/"),
            x = AllData3, quote = FALSE, col.names = TRUE,row.names = FALSE, sep  = "\t") 

PROseq = left_join(PROseqInfo2,AllData2 )


PROseq$annotation2 = "background"
PROseq$annotation2[!is.na(PROseq$annotation)] = "DE"
PROseq$annotation = as.character(PROseq$annotation)
PROseq$annotation[is.na(PROseq$annotation)] = "background"

PROseq$kmeans_UMAP[is.na(PROseq$kmeans_UMAP)]  = 0
PROseq$factor = as.character(PROseq$factor)
PROseq$factor[is.na(PROseq$factor)]  = "None"
PROseq$direction = as.character(PROseq$direction)
PROseq$direction[is.na(PROseq$direction)]  = "None"
PROseq = PROseq %>% mutate(class = paste(factor, direction, sep = "\n"))



```



## Plot according to PROseq data classification
Plotting the data with the classifications that we did using the PROseq gene body covereage data


```{r PROseq classificiation }

ggplot(data = AllData2, mapping = aes(x = umap1, y = umap2 , color = factor(factor) , shape = direction))+ geom_point() + scale_color_viridis(discrete = TRUE, option = "D")
ggplot(data = PROseq, mapping = aes(x = class, y = GBClog,fill = mutant ))+ geom_boxplot()+
  scale_fill_viridis(discrete = TRUE, option = "D") 
ggplot(data = PROseq, mapping = aes(x = as.factor(annotation2), y = PI,fill = mutant ))+ geom_boxplot()+
  scale_fill_viridis(discrete = TRUE, option = "D") 
ggplot(data = PROseq, mapping = aes(x = class, y = PI,fill = mutant))+ geom_boxplot()+
  scale_fill_viridis(discrete = TRUE, option = "D") 
ggplot(data = PROseq, mapping = aes(x = class,
                                      y = ATAC_NC_PC2,fill =factor(factor) )) +
  geom_boxplot()+
  scale_fill_viridis(discrete = TRUE, option = "D")

```



## Plot according to UMAP classification


```{r UMAPclassification}
ggplot(data = AllData2, mapping = aes(x = umap1, y = umap2 , color = factor(kmeans_UMAP) , shape = direction))+ geom_point() + scale_color_brewer(palette="Dark2")
ggplot(data = PROseq, mapping = aes(x = factor(kmeans_UMAP), y = GBClog,fill = mutant ))+ geom_boxplot()+
  scale_fill_viridis(discrete = TRUE, option = "D") 
ggplot(data = PROseq, mapping = aes(x = as.factor(kmeans_UMAP), y = PI,fill = mutant ))+ geom_boxplot()+
  scale_fill_viridis(discrete = TRUE, option = "D") 



ggplot(data = AllData2, mapping = aes(x = factor(kmeans_UMAP),
                                      y = ATAC_NC_PC2,fill =factor(kmeans_UMAP) )) +
  geom_boxplot()+
  scale_color_brewer(palette="Dark2")


```

## Plot according to hclust classification

```{r hclust classification}
ggplot(data = AllData2, mapping = aes(x = umap1, y = umap2 , color = factor(hclust) , shape = direction))+ geom_point() + scale_color_brewer(palette="Dark2")
ggplot(data = PROseq, mapping = aes(x = factor(hclust), y = GBClog,fill = mutant ))+ geom_boxplot()+
  scale_fill_viridis(discrete = TRUE, option = "D") 
ggplot(data = PROseq, mapping = aes(x = as.factor(hclust), y = PI,fill = mutant ))+ geom_boxplot()+
  scale_fill_viridis(discrete = TRUE, option = "D") 



ggplot(data = AllData2, mapping = aes(x = factor(hclust),
                                      y = ATAC_NC_PC2,fill =factor(hclust) )) +
  geom_boxplot()+
  scale_color_brewer(palette="Dark2")




```
