---
title: "ATACseq analysis"
author: "Jenny Jakobsson"
date: "11/11/2020"
output: html_document

params:
    workingDir: /Users/jj/Applied_Bioinformatics_Group_4
    dataDir: data_johan/data/ATACseq
    RscriptsDir: script/original_method
    proSeqDir: results/original_method/PRO-seq
    resultsDir: results/original_method/ATACseq
    sampleInfo: data_johan/information/cleaned/ChIP_ATAC_sampleInfo.tab.txt
    count_ATAC: ATAC.CBP.peak.featureCount.count.tab.txt
    count_ATAC_QC: ATAC.CBP.peak.featureCount.count.QC.tab.txt
    normalisedATACpeakData: ATAC.CBP.peak.featureCount.rlog.tab.txt
    gtffile: data_johan/annotations/Drosophila_melanogaster.BDGP6.28.99.PROseq.gtf
    Method: ATAC-Seq
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# Install required packages
#install.packages("kableExtra")
#install.packages("tidyverse")
#install.packages("RColorBrewer")

# Include the functions from the expression analysis script
source(paste(params$workingDir, params$RscriptsDir,"ExpressionAnalysisFunctions.r", sep = "/"))
library(tidyverse)
library(gplots)
library(kableExtra)
library(DESeq2)
library("viridis")
```

## Reads mapping


## Comparing mapping pattern between the samples

### Using feature count to get the difference between the samples

```{r Adding the first analysis of the data ,message=FALSE ,warning = FALSE}
featureCountFile = paste(params$workingDir, params$dataDir,params$count_ATAC, sep = "/")

FCdata = read.table( file = featureCountFile, header = TRUE, sep = "\t", quote = "", stringsAsFactors = FALSE)

peakInfo  = FCdata [,1:6]
countData = FCdata[,-1:-6]
rownames(countData) = peakInfo$Geneid

```



```{r handle the samples ,message=FALSE ,warning = FALSE}

sampleInfoFile =paste(params$workingDir, params$sampleInfo, sep = "/")
sampleInfo = read.table( file = sampleInfoFile, header = TRUE, 
                     sep = "\t", quote = "",  
                     stringsAsFactors = FALSE)


sampleInfo$bamFiles2  =  gsub(pattern = "/",replacement = ".",x = sampleInfo$bamFiles)
sampleInfo$bamFiles2  =  gsub(pattern = "-",replacement = ".",x = sampleInfo$bamFiles2)


sampleInfo$Genotype = recode(sampleInfo$Genotype, Toll910 = "Toll9")
sampleInfo$Genotype = recode(sampleInfo$Genotype, `Toll9/10` = "Toll9")
sampleInfo$Genotype = recode(sampleInfo$Genotype, gd7 = "Gd7")

sampleInfo =sampleInfo %>% filter(Assaytype == params$Method) 

sampleInfoCountTable = data.frame(bamFiles2 = colnames(countData))


sampleInfo = inner_join(sampleInfo,sampleInfoCountTable)

countData = countData[,sampleInfo$bamFiles2]
sampleInfo$Nuclearcycle[is.na(sampleInfo$Nuclearcycle)]="Mutant"



colnames(countData) = sampleInfo$SampleInfo


fixedSamples = sampleInfo$SampleInfo[grep(pattern = "Fixed", x = sampleInfo$SampleInfo)]
spikedSamples = sampleInfo$SampleInfo[grep(pattern = "spiked", x = sampleInfo$SampleInfo)]



sampleInfo_Mutants = sampleInfo %>% filter( Nuclearcycle == "Mutant" ) %>%
  filter(!SampleInfo %in% fixedSamples) %>%
  filter(!SampleInfo %in% spikedSamples)

countData_mutants = countData[,sampleInfo_Mutants$SampleInfo  ]




```






## Filtering

```{r differential gene expression analysis ,message=FALSE ,warning = FALSE}

exp.data = countData_mutants
metaInfo = sampleInfo_Mutants

rownames(metaInfo) = sampleInfo_Mutants$SampleInfo

metaInfo$Antibody = as.factor(metaInfo$Antibody)
metaInfo$Genotype = as.factor(metaInfo$Genotype)

dds <- DESeqDataSetFromMatrix(countData = exp.data,
                               colData = metaInfo,
                               design = ~Genotype )


#Remove rows with low counts and normalise samples for visualisation
dds <- dds[ rowSums(counts(dds)) > 100, ]

dim(dds)
#Normalizing and change to 2 fold log changes. 
rld <- rlog(dds)


normExpression = as.data.frame(assay(rld))
head(normExpression)

normExpression$geneID = rownames(normExpression)
normExpressionDF  = normExpression %>% gather( key = sampleName, value = rlog, -geneID)

#normExpressionDF = inner_join(normExpressionDF, sampleInfo)


ggplot(normExpressionDF, aes(x = rlog, color = sampleName)) + geom_density()

geneInfo = normExpressionDF %>% 
  dplyr::select(geneID, rlog, sampleName) %>% 
  dplyr::group_by (geneID) %>%
  dplyr::summarize ( min = min(rlog), 
                     max = max(rlog), 
                     mean = mean(rlog), 
                     sd = sd(rlog), 
                     overQC = length(which(rlog > 5))
                     ) 



geneInfo.QC2 = geneInfo %>% 
  filter(overQC > 2)

normExpressionDF.QC = normExpressionDF %>%
  filter(geneID %in% geneInfo.QC2$geneID)

ggplot(normExpressionDF.QC, aes(x = rlog, color = sampleName)) + geom_density()



normExpression.QC = normExpressionDF.QC %>%
  dplyr::select(geneID, sampleName, rlog) %>% 
  spread(key = sampleName, value = rlog)

rownames(normExpression.QC) = normExpression.QC$geneID
normExpression.QC = normExpression.QC[, -1]


```



```{r PCA on the sampleData ,message=FALSE ,warning = FALSE}

plotSample2SampleDistance(normExpression.QC)



#'  __Figure 1 Plotting sample to sample distance__ . 
#'  0 means that they are identical and 1 means that they are totally different.  
#'  The darker the blue the more similair. Also dendogram shows how similair they are. 
#' 
#' Samples do not cluster according to pre an post op. Most likely more due to difference in mapping (technical problem)
#' 
#' #### PCA analysis
#' Running PCA on the samples and plotting the different variables to see which of the parameterrs that fit the different components the best.
#' 
#' First checking how much the different PC contribute. 


mir.pca <- prcomp(t(normExpression.QC), center = TRUE, scale = FALSE) 
e.var = (mir.pca[['sdev']]^2 / sum(mir.pca[['sdev']]^2))
e.var = as.data.frame( e.var )
e.var$PC = as.factor(1:nrow(e.var)) 
e.var$Variance = e.var$e.var*100
e.var$CumulativeVariance = cumsum(e.var$Variance)
qplot(PC, Variance, data=e.var, geom = c("point")
      , ylim = c(0,max(e.var$Variance)+5) 
      ,ylab = "Variance (%)")

#'  __Figure 2 Plotting PCA variance__.  
#'  Displays how much each of the PC contributes to the overall expression.    
#'  This suggest that the two first PC explains most of the variation.
#'
#+ save2, include=FALSE
#ggsave(paste(params$workingDirectory,params$resultsDirectory, "PCAvariance.png", sep = "/"))
#'
#'


pctable = as.data.frame(mir.pca$x)[, 1:5]
pctable = cbind(pctable, sampleInfo_Mutants)


test  = plotPCAplot(PCAinfo = pctable, n.comp = 4,
            varianceInfo = e.var, 
            colorComponent = "Genotype",pchComponent = "Time"
              
)

test

mean = pctable %>% dplyr::select(PC1,PC2,PC3,Genotype) %>%
  dplyr::group_by(Genotype) %>%
  dplyr::summarise(PC1 = mean(PC1) ,PC2 = mean(PC2), PC3 = mean(PC3) )



origo = data.frame(PC2=0, PC3 = 0)


line.gd7 = rbind(mean[1,3:4],origo,-mean[1,3:4])
line.gd7$factor = "Gd7"
line.toll10b = rbind(mean[2,3:4],origo,-mean[2,3:4])
line.toll10b$factor = "Toll10b"
line.toll9 = rbind(mean[3,3:4],origo,-mean[3,3:4])
line.toll9$factor = "Toll9"


line = rbind(line.gd7,line.toll10b,line.toll9)

ggplot(line, mapping = aes(x = PC2,y = PC3, color = factor))+ geom_line() + geom_point(data = pctable,  mapping = aes(x = PC2,y = PC3, color = Genotype, shape = Time))




```

### Print ATAC data to file 

```{r save the files ,message=FALSE ,warning = FALSE}



## Saving sampleInfo File

sampleInfoFile = paste(params$workingDir, params$dataDir,"ATAC.sampleInfo.QC.tab.txt", sep = "/")

write.table(x = sampleInfo_Mutants, 
            file = sampleInfoFile,
            quote = F,
            sep =  "\t",
            row.names = F,
            col.names = T
            )





## Saving peakInfo File

peakInfo_QC = peakInfo %>% 
  dplyr::rename(geneID = Geneid) %>% 
  inner_join( geneInfo.QC2)

peakTableFile = paste(params$workingDir, params$dataDir,"ATAC.CBP.peaks.QC.tab.txt", sep = "/")


write.table(x = peakInfo_QC, 
            file = peakTableFile,
            quote = F,
            sep =  "\t",
            row.names = F,
            col.names = T
)




count_ATAC_QC_File = paste(params$workingDir, params$dataDir,
                         paste(params$count_ATAC_QC, sep = "."), sep = "/")



countData.QC = countData_mutants[peakInfo_QC$geneID,
                                 sampleInfo_Mutants$SampleInfo
                                 ]



write.table(x = countData.QC, 
            file = count_ATAC_QC_File,
            quote = F,
            sep =  "\t",
            row.names = T,
            col.names = T
            )




## Saving normalised count file
NormalisedCountFile = paste(params$workingDir, params$dataDir,
                         paste(params$normalisedATACpeakData, sep = "."), sep = "/")

normExpression.QC = normExpression[peakInfo_QC$geneID,
                                 sampleInfo_Mutants$SampleInfo
                                 ]

write.table(x = normExpression.QC, file = NormalisedCountFile,
                  quote = F, sep = "\t", col.names = T, row.names = T) 


```