---
title: "ChIPseq analysis H3K4me1"
author: "Johan"
date: "10/25/2019"
output: html_document

params:
  workingDir: /Users/johanreimegard/git/flyEmbryogenesis
  RscriptsDir: R_Analysis/Rscripts  
  dataDir: data/ChIPseq
  sampleInfo: information/ChIP_ATAC_sampleInfo.tab.txt
  count_K27ac: K27ac.featureCount.count.tab.txt 
  count_K27me3: K27me3.featureCount.count.tab.txt
  count_H3K4me1: H3K4me1.featureCount.count.tab.txt 
  count_zld: zld.featureCount.count.tab.txt
  Antibody: H3K4me1

  RfunctionsDirectory: /Users/johanreimegard/GoogleDrive/git/RNAmappingPipeline/R
  
  
---

```{r setup, include=FALSE}

source(paste(params$RfunctionsDirectory,"ExpressionAnalysisFunctions.R",sep = "/"))
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)

library(gplots)

library(knitr)
#install.packages("kableExtra")
library(kableExtra)



```

# ChIPseq analysis


## Reads mapping


## Comparing mapping pattern between the samples

### Using feature count to get the difference between the samples

```{r Adding the first analysis of the data}
# install RColorBrewer if missing

library(RColorBrewer)
library(gplots)
featureCountFile = paste(params$workingDir, params$dataDir,
                         paste(params$Antibody , "featureCount.count.tab.txt", sep = "."), sep = "/")

FCdata = read.table( file = featureCountFile, header = TRUE, 
                     sep = "\t", quote = "",  
                     stringsAsFactors = FALSE)

peakInfo  = FCdata [,1:6]
countData = FCdata[,-1:-6]
rownames(countData) = peakInfo$Geneid

```


```{r handle the samples}

sampleInfoFile =paste(params$workingDir, params$sampleInfo, sep = "/")
sampleInfo = read.table( file = sampleInfoFile, header = TRUE, 
                     sep = "\t", quote = "",  
                     stringsAsFactors = FALSE)


sampleInfo$bamFiles2  =  gsub(pattern = "/",replacement = ".",x = sampleInfo$bamFiles)
sampleInfo$bamFiles2  =  gsub(pattern = "-",replacement = ".",x = sampleInfo$bamFiles2)

sampleInfo =sampleInfo %>% filter(Antibody == params$Antibody) 

sampleInfoCountTable = data.frame(bamFiles2 = colnames(countData))


sampleInfo = inner_join(sampleInfo,sampleInfoCountTable)

countData = countData[,sampleInfo$bamFiles2]
colnames(countData) = paste(sampleInfo$Genotype, sampleInfo$Replicate, sep = "_")

```


## Filtering

```{r first qc analysis}

peakInfo$expression  =log(rowSums(countData))

peakInfo.QC = peakInfo %>% dplyr::filter(expression > 1) 
countData.QC =  countData[peakInfo.QC$Geneid,]


```




```{r differential gene expression analysis}
library(DESeq2)

exp.data = countData.QC
metaInfo = sampleInfo

rownames(metaInfo) = paste(sampleInfo$Genotype, sampleInfo$Replicate, sep = "_")

metaInfo$Antibody = as.factor(metaInfo$Antibody)
metaInfo$Genotype = as.factor(metaInfo$Genotype)

dds <- DESeqDataSetFromMatrix(countData = exp.data,
                               colData = metaInfo,
                               design = ~Genotype )


#Remove rows with low counts and normalise samples for visualisation
dds <- dds[ rowSums(counts(dds)) > 100, ]

dim(dds)
#Normalizing and change to 2 fold log changes. 
rld <- rlog(dds)


#' ### Visualising the data


#+ save1, include=FALSE
#png(paste(params$workingDirectory, params$resultsDirectory,"SampleDistance.png", sep = "/"))
#plotSample2SampleDistance(assay(rld))
#dev.off()

#' 
#'
plotSample2SampleDistance(assay(rld))
#'  __Figure 1 Plotting sample to sample distance__ . 
#'  0 means that they are identical and 1 means that they are totally different.  
#'  The darker the blue the more similair. Also dendogram shows how similair they are. 
#' 
#' Samples do not cluster according to pre an post op. Most likely more due to difference in mapping (technical problem)
#' 
#' #### PCA analysis
#' Running PCA on the samples and plotting the different variables to see which of the parameterrs that fit the different components the best.
#' 
#' First checking how much the different PC contribute. 


mir.pca <- prcomp(t(assay(rld)), center = TRUE, scale = FALSE) 
e.var = (mir.pca[['sdev']]^2 / sum(mir.pca[['sdev']]^2))
e.var = as.data.frame( e.var )
e.var$PC = as.factor(1:nrow(e.var)) 
e.var$Variance = e.var$e.var*100
e.var$CumulativeVariance = cumsum(e.var$Variance)
qplot(PC, Variance, data=e.var, geom = c("point")
      , ylim = c(0,max(e.var$Variance)+5) 
      ,ylab = "Variance (%)")

#'  __Figure 2 Plotting PCA variance__.  
#'  Displays how much each of the PC contributes to the overall expression.    
#'  This suggest that the two first PC explains most of the variation.
#'
#+ save2, include=FALSE
#ggsave(paste(params$workingDirectory,params$resultsDirectory, "PCAvariance.png", sep = "/"))
#'
#'
pctable = as.data.frame(mir.pca$x)[, 1:3]
pctable = cbind(pctable, metaInfo)


test  = plotPCAplot(PCAinfo = pctable, n.comp = 3,
            varianceInfo = e.var, 
            colorComponent = "Genotype"
)

test


```

