---
title: "Enhancer In Situ analysis"
author: "Johan"
date: "6/12/2020"
output: html_document

html_document:
    toc: true
    toc_depth: 2

df_print: paged

params:
  workingDir: /Users/johanreimegard/git/flyEmbryogenesis
  proSeqDir: results/PROseq
  proSeq.data: PROseq.data.tsv
  proSeq.data.PPC: PROseq.data.PPC.tsv
  ATACSeqDir: results/ATACseq
  InSituDir:  results/inSitu
  InSituFile: inSitu.Stark.dm6.tsv
  ATACSeq.data: ATACseq.data.raw.tsv
  k27ac.data: K27ac.ChIPseq.peak.pattern.raw.tsv
  CBP.data: ATAC-CBP_dm6.ChIPseq.peak.pattern.raw.tsv
  chipSeqDir: results/chipSeq
  RscriptsDir: R_Analysis/Rscripts  
  dataDir: data/ATACseq
  sampleInfo: information/ChIP_ATAC_sampleInfo.tab.txt
  count_ATAC: ATAC.peak.featureCount.count.tab.txt 
  Method: ATAC-Seq
  Antibody: K27ac
  resultDir: results/multi
  bestTargets: information/TopTargets.txt
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
#install.packages ("tm")
#install.packages ("wordcloud")

#if (!requireNamespace("BiocManager", quietly = TRUE))
#install.packages("BiocManager")
#BiocManager::install("rtracklayer", version = "3.8")

library(rtracklayer)



#Loading Packages
library(tm)
library(wordcloud)
library(RColorBrewer)


library(knitr)
library(kableExtra)


```

## Enhancer specificity 


This document cross reference the enhancer we have identified as enhancers with the enhancers analysis carried out by the stark lab (https://enhancers.starklab.org/search)


We use fisher exact test to get an exact p-value to cross reference what kind of enhancer groups are enriched in the different classes. 


```{r load peak data ,include=FALSE }

featureCountFile = paste(params$workingDir, params$dataDir,params$count_ATAC, sep = "/")
FCdata = read.table( file = featureCountFile, header = TRUE, 
                     sep = "\t", quote = "",  
                     stringsAsFactors = FALSE)

peakInfo  = FCdata [,1:6]

peakInfo.GRanges = makeGRangesFromDataFrame(peakInfo,
                         keep.extra.columns=TRUE,
                         ignore.strand=TRUE,
                         seqinfo=NULL,
                         seqnames.field=c("seqnames", "seqname",
                                          "Chrosome", "chrom",
                                          "chr", "chromosome_name",
                                          "seqid","Chr"),
                         start.field=c("Start","start"),
                         end.field=c("End", "stop"),
                         starts.in.df.are.0based=FALSE)


```



```{r Load in situ data,include=FALSE}

inSituFiles = list.files("data/inSitu/")

enhancers = list()
for(i in 1:length(inSituFiles)){
  fileName = paste("data/inSitu", inSituFiles[i], sep = "/")
  enhancers[[i]] = read.csv(file = fileName)  %>% distinct() 
}

AllEnhancers=  data.frame(enhancers %>% purrr:: reduce(full_join)  %>% distinct())

rownames(AllEnhancers) = AllEnhancers$VTID

AllEnhancers.GRanges = makeGRangesFromDataFrame(AllEnhancers,
                         keep.extra.columns=TRUE,
                         ignore.strand=TRUE,
                         seqinfo=NULL,
                         seqnames.field=c("seqnames", "seqname",
                                          "Chrosome", "chrom",
                                          "chr", "chromosome_name",
                                          "seqid"),
                         start.field="Start",
                         end.field=c("End", "stop"),
                         starts.in.df.are.0based=FALSE)

# import the chain file
chainObject <- import.chain("annotations/dm3ToDm6.over.chain")



AllEnhancers.GRanges.dm6 = unlist(liftOver(AllEnhancers.GRanges, chainObject))

InSituDF = data.frame(AllEnhancers.GRanges.dm6)


inSituFile = paste(params$workingDir, 
                            params$InSituDir,
                            params$InSituFile,
                            sep = "/")

write.table(x = InSituDF,
            file = inSituFile,
            quote = F, 
            sep = "\t", 
            col.names = T, 
            row.names = F) 



```


```{r add new table,include=FALSE}
TollGenes = read.table(file = paste(params$workingDir,"results/combined.AllData.withMetaData.ATACnaive.txt" ,sep = "/"),
             quote = "", header = T, sep  = "\t", stringsAsFactors = F) 



```



## Merge the data
```{r cars,include=FALSE}



overLap = findOverlaps(peakInfo.GRanges,AllEnhancers.GRanges.dm6)

peakInfo = data.frame(peakInfo.GRanges)
peakInfo$queryHits = as.integer(rownames(peakInfo))

InSitu = data.frame(AllEnhancers.GRanges.dm6)
InSitu$subjectHits = as.integer(rownames(InSitu))
InSitu = InSitu %>% dplyr::select(-seqnames,-start, -end, -strand, -Length,-width)


merged = inner_join(peakInfo, data.frame(overLap))
merged = inner_join(merged,InSitu) %>% gather(key = stage, value = Annotation, stg4_6, stg7_8, stg9_10, stg11_12, stg13_14, stg15_16) %>%
  separate_rows(Annotation,sep = "\\|") %>% 
  separate(Annotation, into = c("Annotation", "Intensity"), sep = ";",remove = T) %>%
  rename(Geneid = "ATACseq_Geneid" )

mergedToll = left_join(merged, TollGenes)
mergedToll$kmeans_UMAP[is.na(mergedToll$kmeans_UMAP)] = 0
mergedToll$class = paste(mergedToll$factor,mergedToll$direction)
mergedToll$class[mergedToll$class == "NA NA"] = "Background"

```


## Example of stark data
```{r stark data,include=TRUE}

kable(data.frame(AllEnhancers.GRanges.dm6)[100:200,] ) %>%
  kable_styling() %>%
  scroll_box(width = "100%", height = "200px")
```

## Fisher exact test on the overlap between our identified classes and annotation from stark labs enhancer analysis.

Only keeping the annotation where there is is at least on of the classes that has a adjusted p-value less than 0.1 in the heatmap figure
## Using annotation from PROseq analysis 
#### stg4_6 
```{r Fisher exact test for enrichment of the enhancers 1 }


mergedToll2 = mergedToll %>% 
  filter(stage %in% c( "stg4_6"))   

test = mergedToll2 %>% dplyr::select(Annotation, class)
#test = mergedToll2 %>% dplyr::select(Annotation,kmeans_UMAP)

countTable = data.frame(table(test), stringsAsFactors = FALSE) %>% 
  group_by(Annotation) %>%
  mutate(Total = sum(Freq)) %>%
  ungroup() %>% spread(key = class, value = Freq)


kable(data.frame(countTable)) %>%
  kable_styling() %>%
  scroll_box(width = "100%", height = "200px")

countTableM = as.matrix(countTable %>% dplyr::select(-Annotation))
totalCounts = colSums(countTableM)

fisherTestMatrix = matrix(ncol = ncol(countTableM) -1, nrow = nrow(countTableM), data = 0 )
for(i in 2:ncol(countTableM)){
  for(j in 1:nrow(countTableM)){
    dat = t(matrix(data = c(countTableM[j,i],
                            countTableM[j,1] - countTableM[j,i],
                            totalCounts[i]-countTableM[j,i] , 
                            totalCounts[1] - totalCounts[i] ), 
                   ncol = 2))
    fisherTest<- fisher.test(dat,alternative = "greater")
    fisherTestMatrix[j, i-1] = fisherTest$p.value
  }
}

fisherTestMatrix = fisherTestMatrix * (nrow(fisherTestMatrix)*ncol(fisherTestMatrix))

fisherTestMatrix[fisherTestMatrix>1] = 1 
significant = which(Biobase::rowMin(fisherTestMatrix) < 0.1)
fisherTestMatrix  =data.frame(fisherTestMatrix)
colnames(fisherTestMatrix) = colnames(countTable)[c(-1,-2)]


fisherTestMatrix$Annotation = countTable$Annotation
fisherTestMatrix =fisherTestMatrix[significant,]


fisherTestMatrix %>% 
  gather(key = class, value = adj_p_value, -Annotation)%>% 
  ggplot( aes(class, Annotation, fill= -log(adj_p_value))) + 
  geom_tile() +
  scale_fill_distiller(palette = "RdGy", direction = -1) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))




```


### stg7_8 

```{r Fisher exact test for enrichment of the enhancers 2}


mergedToll2 = mergedToll %>% 
  filter(stage %in% c( "stg7_8"))   

test = mergedToll2 %>% dplyr::select(Annotation, class)
#test = mergedToll2 %>% dplyr::select(Annotation,kmeans_UMAP)

countTable = data.frame(table(test), stringsAsFactors = FALSE) %>% 
  group_by(Annotation) %>%
  mutate(Total = sum(Freq)) %>%
  ungroup() %>% spread(key = class, value = Freq)


countTableM = as.matrix(countTable %>% dplyr::select(-Annotation))
totalCounts = colSums(countTableM)

fisherTestMatrix = matrix(ncol = ncol(countTableM) -1, nrow = nrow(countTableM), data = 0 )
for(i in 2:ncol(countTableM)){
  for(j in 1:nrow(countTableM)){
    dat = t(matrix(data = c(countTableM[j,i],
                            countTableM[j,1] - countTableM[j,i],
                            totalCounts[i]-countTableM[j,i] , 
                            totalCounts[1] - totalCounts[i] ), 
                   ncol = 2))
    fisherTest<- fisher.test(dat,alternative = "greater")
    fisherTestMatrix[j, i-1] = fisherTest$p.value
  }
}

fisherTestMatrix = fisherTestMatrix * (nrow(fisherTestMatrix)*ncol(fisherTestMatrix))

fisherTestMatrix[fisherTestMatrix>1] = 1 
significant = which(Biobase::rowMin(fisherTestMatrix) < 0.1)
fisherTestMatrix  =data.frame(fisherTestMatrix)
colnames(fisherTestMatrix) = colnames(countTable)[c(-1,-2)]


fisherTestMatrix$Annotation = countTable$Annotation
fisherTestMatrix =fisherTestMatrix[significant,]


fisherTestMatrix %>% 
  gather(key = class, value = adj_p_value, -Annotation)%>% 
  ggplot( aes(class, Annotation, fill= -log(adj_p_value))) + 
  geom_tile() +
  scale_fill_distiller(palette = "RdGy", direction = -1) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))



```

## Using annotation from kmean UMAP analysis

#### stg4_6 


```{r Fisher exact test for enrichment of the enhancers 3 }


mergedToll2 = mergedToll %>% 
  filter(stage %in% c( "stg4_6"))   
#test = mergedToll2 %>% dplyr::select(Annotation, class)
test = mergedToll2 %>% dplyr::select(Annotation,kmeans_UMAP)

countTable = data.frame(table(test), stringsAsFactors = FALSE) %>% 
  group_by(Annotation) %>%
  mutate(Total = sum(Freq)) %>%
  ungroup() %>% spread(key = kmeans_UMAP, value = Freq)


countTableM = as.matrix(countTable %>% dplyr::select(-Annotation))
totalCounts = colSums(countTableM)

fisherTestMatrix = matrix(ncol = ncol(countTableM) -1, nrow = nrow(countTableM), data = 0 )
for(i in 2:ncol(countTableM)){
  for(j in 1:nrow(countTableM)){
    dat = t(matrix(data = c(countTableM[j,i],
                            countTableM[j,1] - countTableM[j,i],
                            totalCounts[i]-countTableM[j,i] , 
                            totalCounts[1] - totalCounts[i] ), 
                   ncol = 2))
    fisherTest<- fisher.test(dat,alternative = "greater")
    fisherTestMatrix[j, i-1] = fisherTest$p.value
  }
}

fisherTestMatrix = fisherTestMatrix * (nrow(fisherTestMatrix)*ncol(fisherTestMatrix))

fisherTestMatrix[fisherTestMatrix>1] = 1 
significant = which(Biobase::rowMin(fisherTestMatrix) < 0.1)
fisherTestMatrix  =data.frame(fisherTestMatrix)
colnames(fisherTestMatrix) = colnames(countTable)[c(-1,-2)]


fisherTestMatrix$Annotation = countTable$Annotation
fisherTestMatrix =fisherTestMatrix[significant,]


fisherTestMatrix %>% 
  gather(key = class, value = adj_p_value, -Annotation)%>% 
  ggplot( aes(class, Annotation, fill= -log(adj_p_value))) + 
  geom_tile() +
  scale_fill_distiller(palette = "RdGy", direction = -1) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

fisherTestMatrix %>% 
  gather(key = class, value = adj_p_value, -Annotation)%>% 
  filter(class %in% (c(0, 1,2,4,5,6)))%>%
  mutate(newClass = "None")%>% 
  mutate(newClass = replace( x = newClass , list = class %in% c(1,2), "Neural ectoderm" ))  %>% 
  mutate(newClass = replace( x = newClass , list = class %in% c(5), "Dorsal ectoderm" ))  %>% 
  mutate(newClass = replace( x = newClass , list = class %in% c(4,6), "Mesoderm" ))


fisherTestMatrix %>% 
  gather(key = class, value = adj_p_value, -Annotation)%>% 
  ggplot( aes(class, Annotation, fill= -log(adj_p_value))) + 
  geom_tile() +
  scale_fill_distiller(palette = "RdGy", direction = -1) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))




```

#### stg7_8 


```{r Fisher exact test for enrichment of the enhancers  4 }


mergedToll2 = mergedToll %>% 
  filter(stage %in% c( "stg7_8"))   

#test = mergedToll2 %>% dplyr::select(Annotation, class)
test = mergedToll2 %>% dplyr::select(Annotation,kmeans_UMAP)

countTable = data.frame(table(test), stringsAsFactors = FALSE) %>% 
  group_by(Annotation) %>%
  mutate(Total = sum(Freq)) %>%
  ungroup() %>% spread(key = kmeans_UMAP, value = Freq)


countTableM = as.matrix(countTable %>% dplyr::select(-Annotation))
totalCounts = colSums(countTableM)

fisherTestMatrix = matrix(ncol = ncol(countTableM) -1, nrow = nrow(countTableM), data = 0 )
for(i in 2:ncol(countTableM)){
  for(j in 1:nrow(countTableM)){
    dat = t(matrix(data = c(countTableM[j,i],
                            countTableM[j,1] - countTableM[j,i],
                            totalCounts[i]-countTableM[j,i] , 
                            totalCounts[1] - totalCounts[i] ), 
                   ncol = 2))
    fisherTest<- fisher.test(dat,alternative = "greater")
    fisherTestMatrix[j, i-1] = fisherTest$p.value
  }
}

fisherTestMatrix = fisherTestMatrix * (nrow(fisherTestMatrix)*ncol(fisherTestMatrix))

fisherTestMatrix[fisherTestMatrix>1] = 1 
significant = which(Biobase::rowMin(fisherTestMatrix) < 0.1)
fisherTestMatrix  =data.frame(fisherTestMatrix)
colnames(fisherTestMatrix) = colnames(countTable)[c(-1,-2)]


fisherTestMatrix$Annotation = countTable$Annotation
fisherTestMatrix =fisherTestMatrix[significant,]


fisherTestMatrix %>% 
  gather(key = class, value = adj_p_value, -Annotation)%>% 
  ggplot( aes(class, Annotation, fill= -log(adj_p_value))) + 
  geom_tile() +
  scale_fill_distiller(palette = "RdGy", direction = -1) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))



```


```{r Fisher exact test for enrichment of the enhancers 3 }


mergedToll2 = mergedToll %>% 
  filter(stage %in% c( "stg4_6"))   

mergedToll2 = mergedToll2 %>% 
  filter(kmeans_UMAP %in% (c(0, 1,2,4,5,6)))%>%
  mutate(newClass = "None")%>% 
  mutate(newClass = replace( x = newClass , list = kmeans_UMAP %in% c(1,2), "Neural ectoderm" ))  %>% 
  mutate(newClass = replace( x = newClass , list = kmeans_UMAP %in% c(5), "Dorsal ectoderm" ))  %>% 
  mutate(newClass = replace( x = newClass , list = kmeans_UMAP %in% c(4,6), "Mesoderm" ))

#test = mergedToll2 %>% dplyr::select(Annotation, class)
test = mergedToll2 %>% dplyr::select(Annotation,newClass)

countTable = data.frame(table(test), stringsAsFactors = FALSE) %>% 
  group_by(Annotation) %>%
  mutate(Total = sum(Freq)) %>%
  ungroup() %>% spread(key = newClass, value = Freq)


countTableM = as.matrix(countTable %>% dplyr::select(-Annotation))
totalCounts = colSums(countTableM)

fisherTestMatrix = matrix(ncol = ncol(countTableM) -1, nrow = nrow(countTableM), data = 0 )
for(i in 2:ncol(countTableM)){
  for(j in 1:nrow(countTableM)){
    dat = t(matrix(data = c(countTableM[j,i],
                            countTableM[j,1] - countTableM[j,i],
                            totalCounts[i]-countTableM[j,i] , 
                            totalCounts[1] - totalCounts[i] ), 
                   ncol = 2))
    fisherTest<- fisher.test(dat,alternative = "greater")
    fisherTestMatrix[j, i-1] = fisherTest$p.value
  }
}

fisherTestMatrix = fisherTestMatrix * (nrow(fisherTestMatrix)*ncol(fisherTestMatrix))

fisherTestMatrix[fisherTestMatrix>1] = 1 
significant = which(Biobase::rowMin(fisherTestMatrix) < 0.005)
fisherTestMatrix  =data.frame(fisherTestMatrix)
colnames(fisherTestMatrix) = colnames(countTable)[c(-1,-2)]


fisherTestMatrix$Annotation = countTable$Annotation
fisherTestMatrix =fisherTestMatrix[significant,]


fisherTestMatrix %>% 
  gather(key = class, value = adj_p_value, -Annotation)%>% 
  ggplot( aes(class, Annotation, fill= -log(adj_p_value))) + 
  geom_tile() +
  scale_fill_distiller(palette = "RdGy", direction = -1) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))




fisherTestMatrix %>% 
  gather(key = class, value = adj_p_value, -Annotation)%>% 
  ggplot( aes(class, Annotation, fill= adj_p_value)) + 
  geom_tile() +
  scale_fill_distiller(palette = "Blues", direction = 1) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
ggsave("fig3.starkEnrichment.pdf")







```


```{r get table}


Gene_STARKcorrelation = mergedToll %>% filter(class != "Background") %>% select(geneID,SYMBOL,  VTID, kmeans_UMAP,stage,Annotation) %>% distinct() %>% arrange( SYMBOL, VTID)

write.table(x = Gene_STARKcorrelation, file = "../results/Gene_STARK_enhancer_Correlation.tsv", 
            quote = F, sep = "\t",
            col.names = T,row.names = F
            )

````


