---
title: "PROseq analysis"
author: "Johan"
date: "10/25/2019"
output: html_document

params:
  workingDir: /Users/johanreimegard/git/flyEmbryogenesis
  RscriptsDir: R_Analysis/Rscripts  
  dataDir: data/PROseq
  RfunctionsDirectory: /Users/johanreimegard/GoogleDrive/git/RNAmappingPipeline/R
  proSeqDir: results/PROseq
  proSeq.data: PROseq.data.tsv
  rerun: FALSE

---



```{r setup, include=FALSE}



library(tidyverse)

library(RColorBrewer)
library(gplots)

library(knitr)
library(kableExtra)


##install.packages("viridis")  # Install
library("viridis")     

knitr::opts_chunk$set(echo = TRUE)

```

# PROseq analysis

## Pause index analysis

### Load annoation from PROseq data

```{r load annotation}

PROseqInfo = read.table( file = paste( params$workingDir, params$proSeqDir, params$proSeq.data,
                          sep = "/"),
                         quote = "", sep = "\t", 
                         header = TRUE, stringsAsFactors = F) 


GI2 = PROseqInfo %>% select(geneID, annotation, annotation2) %>% distinct()


```

####  Load the gbc and ppc transcript data

```{r Reading the pp data and the gene body data, include=FALSE}
if (!require("RColorBrewer")) {
  install.packages("RColorBrewer")
}
gbcCountFile = paste(params$workingDir, params$dataDir,"naive/counts.gbc.txt", sep = "/")

gbcData = read.table( file = gbcCountFile, header = TRUE, 
                      sep = "\t", quote = "",  
                      stringsAsFactors = FALSE)



gbcCountData = gbcData[,-1:-6]
TranscriptInfo = gbcData[,1:6]
colnames(TranscriptInfo) = c("geneID","Chr","Start","End","Strand","Length")
rownames(gbcCountData) = gbcData$Geneid

ppcCountFile = paste(params$workingDir, params$dataDir,"naive/counts.ppc.txt", sep = "/")

ppcData = read.table( file = ppcCountFile, header = TRUE, 
                      sep = "\t", quote = "",  
                      stringsAsFactors = FALSE)

ppcCountData = ppcData[,-1:-6]
ppcTrnascriptInfo = ppcData[,1:6]
rownames(ppcCountData) = ppcData$Geneid


```

#### Handle the samples


```{r handle the samples, include=FALSE }

sampleInfo = data.frame(bamFile =  colnames(ppcCountData))
sampleInfo= sampleInfo %>% separate(col = bamFile,sep = "\\.",into =  c("dir", "mutant"), remove = FALSE)
sampleInfo = sampleInfo %>% separate(col = mutant,sep = "_",into =  c("type", "naive","replicate"))


sampleInfo = sampleInfo %>% dplyr::select(bamFile,type,replicate)
sampleInfo = sampleInfo %>% unite("sampleName", type:replicate, remove = FALSE )


rownames(sampleInfo) = sampleInfo$bamFile
sampleInfo = sampleInfo[colnames(ppcCountData), ]
colnames(ppcCountData) = sampleInfo$sampleName
rownames(sampleInfo) = sampleInfo$sampleName

```

#### Merging the ppc and the gbc data
```{r Merging the ppc and the gbc data, include=FALSE }

ppcData = ppcCountData
ppcData$geneID = rownames(ppcData)
ppcData = gather(ppcData ,key = sampleName, value = ppcCount, -geneID)


trancriptData = inner_join(ppcData,sampleInfo)

head(gbcCountData)

gbcData = gbcCountData
gbcData = gbcData[,as.character(sampleInfo$bamFile)]
colnames(gbcData) = sampleInfo$sampleName
gbcData$geneID = rownames(gbcData)
gbcData = gather(gbcData ,key = sampleName, value = gbcCount, -geneID)
head(gbcData)



trancriptData = inner_join(trancriptData,gbcData)


trancriptData = inner_join(trancriptData,TranscriptInfo)



TranscriptCount = trancriptData %>% dplyr::select(geneID,  ppcCount,gbcCount) %>% 
  dplyr::group_by(geneID) %>%
  dplyr::summarise(ppcCountTot = sum(ppcCount), gbcCountTot = sum(gbcCount))

head(trancriptData)

SampleCount = trancriptData %>% dplyr::select(sampleName,  ppcCount,gbcCount) %>% 
  dplyr::group_by(sampleName) %>%
  dplyr::summarise(ppcSampleCountSum = sum(ppcCount), gbcSampleCountSum = sum(gbcCount))
```

#### Normalise the data using TPM
```{r calculate TPM }

trancriptData$ppcPK = trancriptData$ppcCount /(151/1000)
trancriptData$gbccPK = trancriptData$gbcCount /(trancriptData$Length/1000)

ScaleFactors = trancriptData %>% dplyr::select(sampleName,  ppcPK,gbccPK) %>%
  dplyr::group_by(sampleName) %>%
  dplyr::summarise(ppcScaleFactor = sum(ppcPK)/1000000, gbcScaleFactor = sum(gbccPK)/1000000)




trancriptData = inner_join(trancriptData,ScaleFactors )


trancriptData$ppcTPM = trancriptData$ppcPK / trancriptData$ppcScaleFactor
trancriptData$gbcTPM = trancriptData$gbccPK / trancriptData$gbcScaleFactor




TDATA= trancriptData

```


## QC pause index analysis

### PROseq

### Filter genes were either the ppc or the gbc expression is closer to noise

```{r analysing the data for }

trancriptData = TDATA


ggplot(trancriptData, aes( x = log(gbcTPM+1), color = replicate)) +geom_density() + facet_grid(type~. ) 
ggplot(trancriptData, aes( x = log(ppcTPM+1), color = replicate)) +geom_density() + facet_grid(type~. ) 


ggplot(trancriptData, aes( x = log(ppcTPM+1), y = log(gbcTPM+1), color = replicate)) +geom_density_2d()+ facet_grid(type~. ) 


TranscriptCount.expressed.qPRO  = trancriptData %>% filter(type == "qPRO") %>%
  filter (log(gbcTPM+1) > 3.0) %>% filter (log(ppcTPM+1) > 2) %>% distinct(geneID)

TranscriptCount.expressed.PRO  = trancriptData %>% filter(type != "qPRO") %>%
  filter (log(gbcTPM+1) > 3.0) %>% filter (log(ppcTPM+1) > 2) %>% distinct(geneID)


TranscriptCount.expressed= intersect(TranscriptCount.expressed.qPRO, TranscriptCount.expressed.PRO)

trancriptData.QC = inner_join(trancriptData,TranscriptCount.expressed) 

ggplot(trancriptData.QC, aes(x = log10(gbcTPM+1), color = type, linetype = replicate)) +
  geom_density()

ggplot(trancriptData.QC, aes( x = log(ppcTPM+1), y = log(gbcTPM+1), color = replicate)) +geom_density_2d()+ facet_grid(type~. ) 


sum = trancriptData.QC$ppcTPM+trancriptData.QC$gbcTPM
sum[sum == 0] = 1
trancriptData.QC$PI =trancriptData.QC$ppcTPM/sum 



ggplot(trancriptData.QC, aes(x = PI, color = type, linetype = replicate)) +
  geom_density() 


test = trancriptData.QC %>% dplyr::select(geneID,sampleName, ppcTPM) %>%
  spread(key = sampleName, value = ppcTPM)
testCor = cor( test[, 2:ncol(test)] ,method = "spearman")
heatmap(testCor, labCol = "")


test = trancriptData.QC %>% dplyr::select(geneID,sampleName, gbcTPM) %>%
  spread(key = sampleName, value = gbcTPM)
testCor = cor( test[, 2:ncol(test)] ,method = "spearman")
heatmap(testCor, labCol = "")
testCor


test = trancriptData.QC %>% dplyr::select(geneID,sampleName, PI) %>%
  spread(key = sampleName, value = PI)
testCor = cor( test[, 2:ncol(test)] ,method = "spearman")
testCor



ggplot(trancriptData.QC, aes( x = log(ppcTPM+1), y = log(gbcTPM+1), color = replicate)) +geom_density2d()+ facet_grid(type~. ) 



```





### add gene to transcript correlation

```{r  add  transcript to gene Information}

# Load in the fastq file that is generated as data from running multiqc on all samples. 
gtfFile = paste(params$workingDir, "annotations/Drosophila_melanogaster.BDGP6.28.99.gtf", sep = "/")
gtfInfo = read.table(file = gtfFile, header = F, sep = "\t", quote = "", stringsAsFactors = F)

gtfInfomRNA = gtfInfo %>% dplyr::filter(V3 == "transcript") 
gtfInfomRNA = gtfInfomRNA[grep(pattern = "gene_biotype \"protein_coding\"",x = gtfInfomRNA$V9 ),  ] 

gtfInfomRNA_DF= gtfInfomRNA %>% separate(col = V9,sep = ";",into =  c("geneID1", "transcriptID1"))
gtfInfomRNA_DF= gtfInfomRNA_DF %>% separate(col = geneID1,sep = "\"",into =  c("irrelevant", "geneID"))
gtfInfomRNA_DF= gtfInfomRNA_DF %>% separate(col = transcriptID1,sep = "\"",into =  c("irrelevant1", "transcriptID"))

gene2transcriptInfo  = gtfInfomRNA_DF %>% dplyr::select(geneID,transcriptID)


```


### Merge CDS, GBC and PPC and Pause index data.





```{r getting known target genes}

# make a copy 

trancriptData.QC.pauseIndex = trancriptData.QC


# change geneID to trasncriptID as it should be. 
trancriptData.QC.pauseIndex = trancriptData.QC.pauseIndex %>%  dplyr::rename(transcriptID = geneID)

# add geneID to all transript IDs.
trancriptData.QC.pauseIndex = inner_join(trancriptData.QC.pauseIndex, gene2transcriptInfo)

trancriptData.QC.pauseIndex.summary = trancriptData.QC.pauseIndex %>%
  dplyr::group_by(geneID)%>%
  dplyr::summarise(gbcTPM = max(gbcTPM))


gene2transcriptInfo.higlyTranscribed = inner_join(trancriptData.QC.pauseIndex,
                                                  trancriptData.QC.pauseIndex.summary) %>%
  dplyr::select(geneID, transcriptID)



PROseqData.PI = inner_join(trancriptData.QC.pauseIndex,gene2transcriptInfo.higlyTranscribed) 



PROseqData.PI = inner_join(PROseqData.PI,GI2)

ggplot(PROseqData.PI, aes(x = type, y = PI, color = annotation2, linetype = type))+ geom_boxplot() 



ks.test( x= PROseqData.PI$PI[PROseqData.PI$annotation2 == "DE" ],
         y = PROseqData.PI$PI[PROseqData.PI$annotation2 != "DE" ])




PI2 = PROseqData.PI %>% select(geneID, type,  ppcTPM,gbcTPM,PI,annotation2) %>%
  group_by(geneID, type,annotation2) %>%
  summarise(ppcTPM = mean(ppcTPM), gbcTPM = mean(gbcTPM),PI = mean(PI)) 


PI2
PI2.PI = PI2 %>% select(geneID, type,  PI,annotation2) %>%
  spread(key = type, value = PI) %>%
  ggplot(mapping = aes(x = PRO, y = qPRO, color =annotation2 )) + geom_density2d() +facet_grid(annotation2~. )

 PI2 %>% filter(annotation2 == "DE" ) %>% 
   select(geneID, type,  PI,annotation2) %>%
  spread(key = type, value = PI) %>%
  ggplot(mapping = aes(x = PRO, y = qPRO, color =annotation2 )) + geom_density2d()+ geom_point() +
    facet_grid(annotation2~. )

 PI2 %>% filter(annotation2 != "DE" ) %>% 
   select(geneID, type,  PI,annotation2) %>%
  spread(key = type, value = PI) %>%
  ggplot(mapping = aes(x = PRO, y = qPRO, color =annotation2 )) + geom_density2d()+ geom_point() +
    facet_grid(annotation2~. )


```

```{ }
system R failed: 256 at /proj/snic2019-30-14/private/perl/NRSA-v2/bin/pause_PROseq.pl line 426.

```