---
title: "PROseq analysis"
author: "Johan"
date: "10/25/2019"
output: html_document

params:
  workingDir: /Users/johanreimegard/git/Applied_Bioinformatics_Group_4
  RscriptsDir: R_Analysis/Rscripts  
  dataDir: data/PROseq
  RfunctionsDirectory: /Users/johanreimegard/GoogleDrive/git/RNAmappingPipeline/R
  proSeqDir: results/PROseq
  proSeq.data: PROseq.data.tsv
  rerun: FALSE

---



```{r setup, include=FALSE}
source(paste(params$RfunctionsDirectory,"ExpressionAnalysisFunctions.R",sep = "/"))



library(dplyr)
library(tidyr)
library(ggplot2)

library(knitr)
#install.packages("kableExtra")
library(kableExtra)


##install.packages("viridis")  # Install
library("viridis")     

knitr::opts_chunk$set(echo = TRUE)

```

# PROseq analysis

## Pre QC analysis
### Analysing number of reads mapping to fly vs human

```{r idxstats}
#use params that is stated at top of script
idxDataDir = paste(params$workingDir, params$dataDir,"idxstats", sep = "/")
# list files in directory
files = list.files(path = idxDataDir)

#read idxstats into table
idxstats = read.table(file = paste(idxDataDir, files[1] ,sep = "/"), header = F, sep = "\t")

sampleName = gsub(pattern = ".dmel.bowtie2.idxstats.txt", replacement = "", x = files[1])

#chonga columns name to include the samplename instead of counts. 
colnames(idxstats) = c("chromosome","Length",sampleName,"unmapped_reads" )

#selecting only three of the columns from idxstats 
idxstats = idxstats %>% dplyr::select("chromosome","Length",sampleName)


#do the same for all the other files
for(i in 2: length(files)){
  df = read.table(file = paste(idxDataDir, files[i] ,sep = "/"), header = F, sep = "\t")
  sampleName = gsub(pattern = ".dmel.bowtie2.idxstats.txt", replacement = "", x = files[i])
  colnames(df) = c("chromosome","Length",sampleName,"unmapped_reads" )
  df = df %>% dplyr::select("chromosome","Length",sampleName)
  # and join the tables
  idxstats = inner_join(idxstats,df)
}


#convert the structure so it is easier to plot in ggplot. 
idxstatsDF = idxstats %>% gather(key = Sample,value = counts ,-chromosome,-Length )

# Assign species to chromosomes. All that contains 'chr' are from fly
idxstatsDF$species = "Mouse"
idxstatsDF$species[grep(pattern = "chr",x = idxstatsDF$chromosome)] = "Drosophila"



#ggplot(idxstatsDF, aes (x = sample, y = counts, fill = species, color = chromosome)) +
#  geom_col(position = "fill")+ coord_flip() 
#theme(legend.position = "top")



#Convert to numeric
idxstatsDF$Length = as.numeric(idxstatsDF$Length)
idxstatsDF$counts = as.numeric(idxstatsDF$counts)

#summarize data using dplyr functions .really good way of cleaning data!!
idxSummary = idxstatsDF %>% 
  dplyr::select(Length,Sample,counts,species) %>%
  dplyr::group_by(species, Sample) %>%
  dplyr::summarize(n(), TotalCount = sum(counts), TotalLength = sum(Length), relativeCount = TotalCount/ TotalLength )


# Plot absolute counts between Droshphila and Mouse
ggplot(idxSummary, aes (x = Sample, y = TotalCount, fill = species, color = )) +geom_col(position = "fill")+ coord_flip() +
  theme(legend.position = "top")

ggsave("PROseq_counts_distribution.pdf")

# Plot relative counts based on genome size between Droshphila and Mouse
ggplot(idxSummary, aes (x = Sample, y = relativeCount, fill = species)) +geom_col(position = "fill")+
  coord_flip() + theme(legend.position = "top")

ggsave("PROseq_counts_relative_distribution.pdf")

```

```{r  add fastqc}

# Load in the fastq file that is generated as data from running multiqc on all samples. 
fastqcFile = paste(params$workingDir, params$dataDir,"multiqc_fastqc.txt", sep = "/")
fastqc = read.table(file = fastqcFile, header = T, sep = "\t")


# Use only some of the columns .
fastqc_info = fastqc %>% dplyr::select(Sample,avg_sequence_length,Total.Sequences )

# Spread the data from idxSummary agains so that each species gets its own column 
sampleInfo =   idxSummary %>% dplyr::select(species, Sample, TotalCount) %>% spread(key = species, value = TotalCount)

# Join idx stats and fastqc stats 
sampleInfo = inner_join(sampleInfo,fastqc_info) 

# Nr unmapped by subtraction
unmapped  = sampleInfo %>% dplyr::group_by(Sample) %>% dplyr::summarize(unmapped = Total.Sequences-Drosophila-Mouse)
sampleInfo = inner_join(sampleInfo,unmapped)


# Plot the data in a html format table. 


sampleInfo %>% dplyr::select(Sample,avg_sequence_length,Total.Sequences,Drosophila,Mouse,unmapped) %>%
  kable() %>%
  kable_styling()


```


### Comparing mapping pattern between the samples


```{r Load the data, include=FALSE }
# install RColorBrewer if missing
#if (!require("RColorBrewer")) {
#  install.packages("RColorBrewer")
#}
library(RColorBrewer)
library(gplots)
featureCountFile = paste(params$workingDir, params$dataDir,"counts.CDS.txt", sep = "/")

FCdata = read.table( file = featureCountFile, header = TRUE, 
                     sep = "\t", quote = "",  
                     stringsAsFactors = FALSE)

geneInfo  = FCdata [,1:6]
geneInfo  = geneInfo[,c(1,6)]
countData = FCdata[,-1:-6]
rownames(countData) = geneInfo$Geneid
```


```{r Create meta data table for samples , include=FALSE}

sampleInfoOld = sampleInfo
sampleInfo = data.frame(bamFile =  colnames(countData))
sampleInfo= sampleInfo %>% separate(col = bamFile,sep = "\\.",into =  c("dir", "mutant","sampleInfo"), remove = FALSE)
sampleInfo = sampleInfo %>% separate(col = sampleInfo,sep = "_",into =  c("irrelevant", "time","replicate"))

sampleInfo = sampleInfo %>% separate(col = mutant,sep = "_",into =  c("mutant"))

sampleInfo = sampleInfo %>% dplyr::select(bamFile,mutant,time,replicate)
sampleInfo = sampleInfo %>% unite("sampleName", mutant:time:replicate, remove = FALSE )
sampleInfo = sampleInfo %>% dplyr::select(bamFile,sampleName,mutant,time,replicate)


rownames(sampleInfo) = sampleInfo$bamFile
sampleInfo = sampleInfo[colnames(countData), ]
colnames(countData) = sampleInfo$sampleName
rownames(sampleInfo) = sampleInfo$sampleName

sampleInfo = sampleInfo %>% dplyr::select(sampleName,mutant,time,replicate,bamFile)



```


```{r first qc analysis}

geneExpression = log(rowSums(countData))

hist(geneExpression)

geneInfo$expression  =log(rowSums(countData))

geneInfo.QC = geneInfo %>% filter(expression > 1) 
countData.QC =  countData[geneInfo.QC$Geneid,]

distance = cor((countData.QC))

distance_matrix <- as.matrix(distance)
heatmap.2(distance_matrix, 
          col=brewer.pal(9,"Blues"), 
          density.info = "none",
          trace="none")


```


```{r script to check the known genes. }

knownGenes = read.table(file = paste(params$workingDir,"information/FlyBase_IDs.txt", sep = "/"),
                        sep = "\t", header = F, quote = "", stringsAsFactors = F)
colnames(knownGenes) = c("Submitted ID", "Current ID","Converted ID","Related record")

roshanGenes = read.table(file = paste(params$workingDir,"information/PROseq_DEG_IGB_validated.txt", sep = "/"),
                         sep = "\t", header = T, quote = "", stringsAsFactors = F)
colnames(roshanGenes) = c("SYMBOL", "Mutant")


geneInfo.QC$knownGene = "No"
geneInfo.QC$knownGene[geneInfo.QC$Geneid %in% knownGenes$`Current ID`] = "Yes"

table(geneInfo.QC$knownGene)

countData.QC.knownTargets =  countData[geneInfo.QC$Geneid[geneInfo.QC$knownGene == "Yes"],]
distance = cor((countData.QC.knownTargets))

distance_matrix <- as.matrix(distance)
pdf(file = "PROseqData.HeatMap.knownGenes.pdf")
heatmap.2(distance_matrix, 
          col=brewer.pal(9,"Blues"), 
          margins = c(8, 8),
          density.info = "none",
          trace="none")

dev.off()


```


#### Identifying that two samples have wrong lables and needs to be switched.
```{r change the two files that differ}
# Change filenames 
test = sampleInfo["toll10b_5h_rep1", ]
sampleInfo["toll10b_5h_rep1", ] = sampleInfo["gd7_5h_rep1", ]  
sampleInfo["gd7_5h_rep1", ] = test 
colnames(countData) = sampleInfo$sampleName
rownames(sampleInfo) = sampleInfo$sampleName
countData.QC =  countData[geneInfo.QC$Geneid,]

countData.QC.knownTargets =  countData[geneInfo.QC$Geneid[geneInfo.QC$knownGene == "Yes"],]
distance = cor((countData.QC.knownTargets))

distance_matrix <- as.matrix(distance)


pdf(file = "PROseqData.HeatMap.knownGenes.pdf")
heatmap.2(distance_matrix, 
          col=brewer.pal(9,"Blues"), 
          margins = c(8, 8),
          density.info = "none",
          trace="none")

dev.off()



```

## Differentiall expresssion analysis

###Normalising data and removinvg lowly expressed genes


```{r differential gene expression analysis}
library(DESeq2)

exp.data = countData.QC
metaInfo = sampleInfo


metaInfo$mutant = as.factor(metaInfo$mutant)
metaInfo$time = as.factor(metaInfo$time)

dds <- DESeqDataSetFromMatrix(countData = exp.data,
                              colData = metaInfo,
                              design = ~time + mutant )


#Remove rows with low counts and normalise samples for visualisation
dds <- dds[ rowSums(counts(dds)) > 100, ]

dim(dds)
#Normalizing and change to 2 fold log changes. 
rld <- rlog(dds)

normExpression = as.data.frame(assay(rld))
head(normExpression)

normExpression$geneID = rownames(normExpression)
normExpressionDF  = normExpression %>% gather( key = sampleName, value = rlog, -geneID)

normExpressionDF = inner_join(normExpressionDF, sampleInfo)


ggplot(normExpressionDF, aes(x = rlog, color = sampleName, linetype = time)) + geom_density()

geneInfo = normExpressionDF %>% dplyr::select(geneID, rlog, sampleName) %>% 
  dplyr::group_by (geneID) %>%
  dplyr::summarize ( min = min(rlog), max = max(rlog), mean = mean(rlog), sd = sd(rlog), overQC = length(which(rlog > 5)) ) 
geneInfo.QC2 = geneInfo %>% filter(overQC > 2)
normExpressionDF.QC = normExpressionDF %>% filter(geneID %in% geneInfo.QC2$geneID)

ggplot(normExpressionDF.QC, aes(x = rlog, color = mutant, linetype = time)) + geom_density()


normExpression.QC = normExpressionDF.QC %>%
  dplyr::select(geneID, sampleName, rlog) %>% 
  spread(key = sampleName, value = rlog)


rownames(normExpression.QC) = normExpression.QC$geneID
normExpression.QC = normExpression.QC[, -1]



```


### Visualising the data
```{r heatmap of samples}

#+ save1, include=FALSE
#png(paste(params$workingDirectory, params$proSeqDirectory,"SampleDistance.png", sep = "/"))
#plotSample2SampleDistance(assay(rld))
#dev.off()

#' 
#'
plotSample2SampleDistance(normExpression.QC)
#'  __Figure 1 Plotting sample to sample distance__ . 
#'  0 means that they are identical and 1 means that they are totally different.  
#'  The darker the blue the more similair. Also dendogram shows how similair they are. 
#' 
#' Samples do not cluster according to pre an post op. Most likely more due to difference in mapping (technical problem)
#' 
#' #### PCA analysis
#' Running PCA on the samples and plotting the different variables to see which of the parameterrs that fit the different components the best.
#' 
#' First checking how much the different PC contribute. 

```

###PCA analysis with normalised counts

```{r PCA analysis}

mir.pca <- prcomp(t(normExpression.QC), center = TRUE, scale = FALSE) 
e.var = (mir.pca[['sdev']]^2 / sum(mir.pca[['sdev']]^2))
e.var = as.data.frame( e.var )
e.var$PC = as.factor(1:nrow(e.var)) 
e.var$Variance = e.var$e.var*100
e.var$CumulativeVariance = cumsum(e.var$Variance)
qplot(PC, Variance, data=e.var, geom = c("point")
      , ylim = c(0,max(e.var$Variance)+5) 
      ,ylab = "Variance (%)")

#'  __Figure 2 Plotting PCA variance__.  
#'  Displays how much each of the PC contributes to the overall expression.    
#'  This suggest that the two first PC explains most of the variation.
#'
#+ save2, include=FALSE
#ggsave(paste(params$workingDirectory,params$proSeqDirectory, "PCAvariance.png", sep = "/"))
#'
#'
pctable = as.data.frame(mir.pca$x)[, 1:6]
pctable$sampleName = rownames(pctable)
pctable = inner_join(pctable, metaInfo)


test  = plotPCAplot(PCAinfo = pctable, n.comp = 5,
                    varianceInfo = e.var, 
                    colorComponent = "mutant", 
                    pchComponent = "time"
)
test
#ggsave("PCA_analysis.pdf")


ggplot(data = pctable, mapping = aes(x = PC1, y =  PC3, color = mutant)) + geom_point()

ggplot(data = pctable, mapping = aes(x = PC2, y =  PC5, color = mutant, shape = time)) + geom_point(aes(shape=time))

loadings = mir.pca$rotation[,1:3]


mean = pctable %>% dplyr::select(PC1,PC2,PC3,mutant) %>%
  dplyr::group_by(mutant) %>%
  dplyr::summarise(PC1 = mean(PC1) ,PC2 = mean(PC2), PC3 = mean(PC3) )

meantime = pctable %>% dplyr::select(PC1,PC2,PC3,time) %>%
  dplyr::group_by(time) %>%
  dplyr::summarise(PC1 = mean(PC1) ,PC2 = mean(PC2), PC3 = mean(PC3) )


origo = data.frame(PC1=0, PC2 = 0, PC3 = 0)


line.gd7 = rbind(mean[1,2:4],origo,-mean[1,2:4])
line.gd7$factor = "Gd7"
line.toll10b = rbind(mean[2,2:4],origo,-mean[2,2:4])
line.toll10b$factor = "Toll10b"
line.toll9 = rbind(mean[3,2:4],origo,-mean[3,2:4])
line.toll9$factor = "Toll9"
line.time = rbind(meantime[2,2:4],origo,meantime[1,2:4])
line.time$factor = "Time"


line = rbind(line.gd7,line.toll10b,line.toll9,line.time)


pctable$factor  = pctable$mutant
pctable$factor =  recode(pctable$factor, gd7 = "Gd7")
pctable$factor =  recode(pctable$factor, toll9 = "Toll9")
pctable$factor =  recode(pctable$factor, toll10b = "Toll10b")


fileNameFigCDSsamples = paste( params$workingDir, params$proSeqDir, 
                        paste("PCA_samples_with_decisionlines_CDS.pdf", sep = "_") ,
                        sep = "/")
ggplot(line, mapping = aes(x = PC2,y = PC3, color = factor))+ geom_line() + geom_point(data = pctable,  mapping = aes(x = PC2,y = PC3, color = factor, shape = time))+
  scale_color_viridis(discrete = TRUE, option = "D")
ggsave(fileNameFigCDSsamples)
```


```{r Plot the axis }

qplot(PC, Variance, data=e.var, geom = c("point")
      , ylim = c(0,max(e.var$Variance)+5) 
      ,ylab = "Variance (%)")

test
fileNameFigCDSsamples = paste( params$workingDir, params$proSeqDir, 
                        paste("PCA_samples_with_decisionlines_CDS.pdf", sep = "_") ,
                        sep = "/")
ggplot(line, mapping = aes(x = PC2,y = PC3, color = factor))+ geom_line() + geom_point(data = pctable,  mapping = aes(x = PC2,y = PC3, color = factor, shape = time))+
  scale_color_viridis(discrete = TRUE, option = "D")
ggsave(fileNameFigCDSsamples)

ggplot(line, mapping = aes(x = PC1,y = PC2, color = factor))+ geom_line() + geom_point(data = pctable,  mapping = aes(x = PC1,y = PC2, color = factor, shape = time))+
  scale_color_viridis(discrete = TRUE, option = "D")
```



```{r analysing the loadings }

#loadings1 = as.data.frame(loadings) 
loadings1 =  as.data.frame(t((t(loadings)-colMeans(loadings))/colSds(loadings)))
#loadings1 = as.data.frame(loadings)
loadings1$geneID = rownames(loadings1)

distance3d <- function(a,b){
  v1 = a - b
  
  distance = ((v1[[1]]^2 + v1[[2]]^2 + v1[[3]]^2)^0.5)
  return(distance)
}


dist3d <- function(point,b,c) {
  v1 <- b - c
  v2 <- point - b      
  v3 <- cross3d_prod(v1,v2)
  area <- sqrt(sum(v3*v3))/2
  d <- 2*area/sqrt(sum(v1*v1))
  return(d)
}

cross3d_prod <- function(v1,v2){
  v3 <- vector()
  v3[1] <- v1[2]*v2[3]-v1[3]*v2[2]
  v3[2] <- v1[3]*v2[1]-v1[1]*v2[3]
  v3[3] <- v1[1]*v2[2]-v1[2]*v2[1]
  return(v3)
}

for(j in unique(line$factor)){
  loadings1[[j]]=0
  for(i in 1:nrow(loadings1)){
  line1 = line %>% dplyr::filter(factor == j) %>% dplyr::select(PC1,PC2,PC3)
  loadings1[i,j] = dist3d(b =as.numeric(line1[1,]),
                            c = as.numeric(line1[3,]),
                            point = as.numeric(loadings1[i,1:3]))

  }

}


  




loadingsDF = gather(loadings1, key = PC, value = Loading, -geneID )

ggplot(loadingsDF, mapping = aes(x = Loading, color = PC )) + geom_density() 



```

#### Extracting interesting genes for the different timepoints and mutants based on the loading results. 

#### Annotate the loading scores

```{r  annotate the loadings scores  }


library(clusterProfiler)
library(org.Dm.eg.db)
loadingsDF = gather(loadings1, key = PC, value = Loading, PC1,PC2,PC3 )
loadingsDF  =  loadingsDF %>% dplyr::rename( ENSEMBL= geneID)

universe <- bitr(unique(loadingsDF$ENSEMBL), fromType = "ENSEMBL",
                 toType = c("ENTREZID", "SYMBOL"),
                 OrgDb = org.Dm.eg.db)


loadingsDF = inner_join(loadingsDF, universe) 

kegg = bitr_kegg(geneID = unique(loadingsDF$ENTREZID), fromType = "ncbi-geneid",
                 toType = c("kegg"),
                 organism = 'dme')
colnames(kegg) = c("ENTREZID","KEGGID") 

loadingsDF = inner_join(loadingsDF,kegg)



loadingsDF$Known = "No"
loadingsDF$Known[loadingsDF$ENSEMBL %in% knownGenes$`Current ID`  ] = "Yes"



loadingsDF = left_join(x = loadingsDF, y = roshanGenes  )
loadingsDF$Mutant[is.na(loadingsDF$Mutant)] = "No"


loadingsDF$annotation = "None"
loadingsDF$annotation[loadingsDF$Known == "Yes"] = "Literature" 
loadingsDF$annotation[loadingsDF$Mutant != "No"] = "Roshan" 
loadingsDF$annotation[loadingsDF$Mutant != "No" & loadingsDF$Known == "Yes"] = "Literature & Roshan" 



# ggo <- groupGO(gene     = unique(gene.df$ENTREZID),
#               OrgDb    = org.Dm.eg.db,
#              ont      = "CC",
#               level    = 3,
#               readable = TRUE)

```




```{r clustering the genes based on the first three PCs}



loadingsDF.all = loadingsDF %>%  spread(key = PC, value = Loading)




tmp = loadingsDF.all
tmp$distance =  ((tmp$PC1^2 + tmp$PC2^2 + tmp$PC3^2)^0.5)
loadingsDF.all = tmp
rm(tmp)

dim(loadingsDF.all)

length(unique(loadingsDF.all$ENSEMBL))
length(unique(loadingsDF.all$ENTREZID))

data.frame(table(loadingsDF.all$ENTREZID)) %>% arrange(desc(Freq))



loadingsDF.all.factorDistance =loadingsDF.all %>% gather(key = factor, value = distance2Factor,Time,Gd7,Toll10b,Toll9 )


loadingsDF.all.factorDistance$distanceFactor = sqrt(loadingsDF.all.factorDistance$distance^2 -
                                                   loadingsDF.all.factorDistance$distance2Factor^2)

loadingsDF.all.factorDistance$distanceFactor[is.na(loadingsDF.all.factorDistance$distanceFactor)] = 0


loadingsDF.all.factorDistance$direction = "Down"
line$Dir = rep(c("Up","origo","Down"), 4)


for(i in 1:nrow(loadingsDF.all.factorDistance)){
  factor2 = loadingsDF.all.factorDistance[i,]$factor
  UpPoint = line %>% dplyr::filter(factor == factor2 & Dir == "Up") %>% dplyr::select(PC1,PC2,PC3)
  DownPoint = line %>% dplyr::filter(factor == factor2 & Dir == "Down") %>% dplyr::select(PC1,PC2,PC3)
  point = loadingsDF.all.factorDistance[i,c("PC1","PC2","PC3")]
  up = distance3d(as.numeric( UpPoint),as.numeric(point))
  down = distance3d(as.numeric(DownPoint),as.numeric(point))
  
  if(up < down){
    loadingsDF.all.factorDistance[i,"direction"] = "Up"
  }else{
    loadingsDF.all.factorDistance[i,"distanceFactor"] = 
      -loadingsDF.all.factorDistance[i,"distanceFactor"]
  }
}

loadingsDF.all.factorDistance.backup = loadingsDF.all.factorDistance

```


```{r  filtering the data }



loadingsDF.all.factorDistance2 = loadingsDF.all.factorDistance %>% dplyr::select(ENSEMBL,ENTREZID,SYMBOL,KEGGID,Known,Mutant,annotation ,PC1, PC2,PC3, distance, factor,distanceFactor) %>% spread( key = factor, value = distanceFactor)




# Filter to keep only the best choice
loadingsDF.all.factorDistance.max = loadingsDF.all.factorDistance %>% group_by(ENSEMBL) %>% summarise(max = max(abs(distanceFactor)))
loadingsDF.all.factorDistance$max=abs(loadingsDF.all.factorDistance$distanceFactor) 
loadingsDF.all.factorDistance.max = inner_join(loadingsDF.all.factorDistance,loadingsDF.all.factorDistance.max)
ggplot(loadingsDF.all.factorDistance.max, mapping = aes(x=max, color = annotation ))+ geom_density() 


loadingsDF.all.filtered = loadingsDF.all.factorDistance.max %>% filter(max >3.5)  %>% dplyr::select(ENSEMBL, factor, direction)

loadingsDF.all.removed = loadingsDF.all.factorDistance.max %>% filter(max <=3.5 )%>% filter (annotation != "None")   %>% dplyr::select(ENSEMBL, factor, direction)
loadingsDF.all.removed.good = loadingsDF.all.factorDistance.max %>% filter(max <=3.5 )%>% filter (annotation == "None")   %>% dplyr::select(ENSEMBL, factor, direction)


loadingsDF.all.filtered2  = inner_join(loadingsDF.all.filtered, loadingsDF.all.factorDistance2)
loadingsDF.all.filtered2$annotation =   paste(loadingsDF.all.filtered2$annotation," & This", sep = "") 
loadingsDF.all.filtered2$annotation = recode(loadingsDF.all.filtered2$annotation, `None & This`= "This")


loadingsDF.all.removed2 = inner_join(loadingsDF.all.removed, loadingsDF.all.factorDistance2)

loadingsDF.all.removedBad = inner_join(loadingsDF.all.removed.good, loadingsDF.all.factorDistance2)




loadingsDF.all.filtered2 = rbind(loadingsDF.all.filtered2,loadingsDF.all.removed2)
loadingsDF.all.filtered2 = rbind(loadingsDF.all.filtered2,loadingsDF.all.removedBad)


loadingsDF.all.filtered2  = loadingsDF.all.filtered2 %>% dplyr::select(-ENTREZID,-KEGGID,  -Known ) %>% dplyr::rename(geneID = ENSEMBL)


max.sort = loadingsDF.all.filtered2 %>% group_by(geneID) %>% summarise(distance = max(distance))
loadingsDF.all.filtered2 = inner_join(loadingsDF.all.filtered2,max.sort)
loadingsDF.all.filtered2  = distinct(loadingsDF.all.filtered2)

loadingsDF.all.annotated  = loadingsDF.all.filtered2 %>% dplyr::select (-SYMBOL)  %>%  distinct()

dim(loadingsDF.all.annotated)


```



```{r Plot the axis }

qplot(PC, Variance, data=e.var, geom = c("point")
      , ylim = c(0,max(e.var$Variance)+5) 
      ,ylab = "Variance (%)")

test
fileNameFigCDSsamples = paste( params$workingDir, params$proSeqDir, 
                        paste("PCA_samples_with_decisionlines_CDS.pdf", sep = "_") ,
                        sep = "/")

ggplot(line, mapping = aes(x = PC2,y = PC3, color = factor))+ geom_line() + geom_point(data = pctable,  mapping = aes(x = PC2,y = PC3, color = factor, shape = time))+
  scale_color_viridis(discrete = TRUE, option = "D")

ggplot(line, mapping = aes(x = PC1,y = PC2, color = factor))+ geom_line() + geom_point(data = pctable,  mapping = aes(x = PC1,y = PC2, color = factor, shape = time))+
  scale_color_viridis(discrete = TRUE, option = "D")



ggplot(loadingsDF.all.filtered2, aes (x =PC2 , y =  PC3, color = factor, shape = direction)) + 
  geom_point()+
  scale_color_viridis(discrete = TRUE, option = "D") +
  scale_shape_manual(values=c(25, 20, 17))+ 
  ggtitle(label = "PROseq data CDS PCA loading values")+ 
  geom_line(data = line, mapping = aes(x = PC2,y = PC3, color = factor, shape = "None"))+
  coord_cartesian(xlim = c(min(loadingsDF.all.targetGenes$PC2),
            max(loadingsDF.all.targetGenes$PC2)),
       ylim=  c(min(loadingsDF.all.targetGenes$PC3),
            max(loadingsDF.all.targetGenes$PC3))
       )
  

ggplot(loadingsDF.all.filtered2, aes (x =PC1 , color = factor, linetype = direction)) + 
  geom_density()+scale_color_viridis(discrete = TRUE, option = "D")

ggplot(loadingsDF.all.factorDistance.max, mapping = aes(x=max, color = annotation ))+ stat_ecdf() 


ggplot(loadingsDF.all.targetGenes, aes (x =PC2 , y =  PC3, color = factor, shape = direction)) + 
  geom_point()+
  scale_color_viridis(discrete = TRUE, option = "D") +
  scale_shape_manual(values=c(25, 20, 17))+ 
  ggtitle(label = "PROseq data CDS PCA loading values")+ 
  geom_line(data = line, mapping = aes(x = PC2,y = PC3, color = factor, shape = "None"))+
  coord_cartesian(xlim = c(min(loadingsDF.all.targetGenes$PC2),
            max(loadingsDF.all.targetGenes$PC2)),
       ylim=  c(min(loadingsDF.all.targetGenes$PC3),
            max(loadingsDF.all.targetGenes$PC3))
       )
  



loadingsDF.all.targetGenes %>% dplyr::select(SYMBOL, factor, direction, annotation, Time, Gd7, Toll10b, Toll9)

```







#### Identifying our own candidates 

```{r Filter and cluster hits that are relevant for further anaysis}

# Filter and cluster hits that are relevant for further anaysis  based on distance from origo



loadingsDF.all.targetGenes = loadingsDF.all.filtered2 %>% filter (annotation != "None") 




ggplot(loadingsDF.all.targetGenes, aes (x =PC2 , y =  PC3, color = PC1)) + 
  geom_point()+
  facet_grid(factor~direction)+
  scale_color_viridis( option = "D")







ggplot(loadingsDF.all.targetGenes, aes (x =PC2 , y =  PC3, color = factor, shape = direction)) + 
  geom_point()+
  scale_color_viridis(discrete = TRUE, option = "D") +
  scale_shape_manual(values=c(25, 20, 17))+ 
  ggtitle(label = "PROseq data CDS PCA loading values")+ 
  geom_line(data = line, mapping = aes(x = PC2,y = PC3, color = factor, shape = "None"))+
  coord_cartesian(xlim = c(min(loadingsDF.all.targetGenes$PC2),
            max(loadingsDF.all.targetGenes$PC2)),
       ylim=  c(min(loadingsDF.all.targetGenes$PC3),
            max(loadingsDF.all.targetGenes$PC3))
       )
  


loadingsDF.all.targetGenes.DF  = loadingsDF.all.targetGenes %>% gather(key = PC, value = score, PC1, PC2, PC3) 
loadingsDF.all.targetGenes.DF$factor2 = paste(loadingsDF.all.targetGenes.DF$factor,loadingsDF.all.targetGenes.DF$direction, sep = "_")
ggplot(loadingsDF.all.targetGenes.DF,aes( x = factor2, y = score, fill =PC  )) + geom_boxplot(position=position_dodge()) + theme(axis.text.x = element_text(angle = 90, hjust = 1))




ggplot(loadingsDF.all.targetGenes, aes (x =PC2 , y =  PC3, color = as.factor(factor))) + 
  geom_point()+
  facet_wrap(annotation~direction)+
  scale_color_viridis( option = "D",discrete=TRUE)
ggsave(filename = "summary2.PCloadings_CDS.pdf", width = 15, height = 15)




```

#### Merging our candidates with Roshan and literature candidates

```{r analysing the removed samples  and merging them into the table }


#cl.known.filter <- kmeans(loadingsDF.known.trimmed[,c("PC1","PC2","PC3")],5)
#loadingsDF.known.trimmed$cluster.known.filter = as.factor(cl.known.filter$cluster)

#cl.known <- kmeans(loadingsDF.known[,c("PC1","PC2","PC3")],9)
#loadingsDF.known$cluster.known = as.factor(cl.known$cluster)


```


#### Include count data 

```{r adding rlog data for each sample }

loadingsDF.all.annotated$annotation2 = "background" 
loadingsDF.all.annotated$annotation2[loadingsDF.all.annotated$annotation != "None"] = "DE" 


length(unique(loadingsDF.all.annotated$geneID))

dim(loadingsDF.all.annotated)





PROseqData.CDS = inner_join(loadingsDF.all.annotated,normExpressionDF.QC)


```

```{r function to get GO term enrichments, include=FALSE }
enrichGO2 <- function(geneNames, backgroundList, 
                      ontology = "BP", 
                      pvalueCutoff = 0.01,
                      qvalueCutoff = 0.05){
  enriched =  enrichGO(gene          = geneNames,
                       universe      = backgroundList,
                       OrgDb         = org.Dm.eg.db,
                       ont           = "BP",
                       pAdjustMethod = "BH",
                       pvalueCutoff  = pvalueCutoff,
                       qvalueCutoff  = qvalueCutoff,
                       readable      = TRUE)
  return(enriched)
}
```
#### GOterm analysis if GOI from loadings
```{ GO}
backgroundList = unique(loadingsDF$ENTREZID)

ego.gd7 = enrichGO2(unique(loadingsGD7$ENTREZID),backgroundList)
ego.toll10b = enrichGO2(unique(loadingstoll10b$ENTREZID),backgroundList)
ego.toll9 = enrichGO2(unique(loadingstoll9$ENTREZID),backgroundList)
ego.5h = enrichGO2(unique(loadings5h$ENTREZID),backgroundList)
ego.3h = enrichGO2(unique(loadings3h$ENTREZID),backgroundList)


ego.3h  = simplify(ego.3h)
ego.5h  = simplify(ego.5h)
ego.gd7  = simplify(ego.gd7)
ego.toll10b  = simplify(ego.toll10b)
ego.toll9  = simplify(ego.toll9)


library(enrichplot)
bp.gd7 = barplot(ego.gd7, showCategory=10)
bp.toll10b = barplot(ego.toll10b, showCategory=10)
bp.toll9 = barplot(ego.toll9, showCategory=10)
bp.3h = barplot(ego.3h, showCategory=10)
bp.5h = barplot(ego.5h, showCategory=10)

goterms.mutanst = cowplot::plot_grid(bp.gd7, bp.toll10b, bp.toll9, ncol=1, labels=c("gd7","toll10b", "toll9"))
cowplot::save_plot(filename = "GOterm.mutants.pdf", plot = goterms.mutanst , base_height = 7)
goterms.timePoints = cowplot::plot_grid(bp.3h, bp.5h, ncol=1, labels=c("3h","5h"))
cowplot::save_plot(filename = "GOterm.timePoints.pdf", plot = goterms.timePoints , base_height = 7)



cnet.5h <- cnetplot(ego.5h,circular = TRUE, colorEdge = TRUE)
cnet.3h <- cnetplot(ego.3h ,circular = TRUE, colorEdge = TRUE)
cnet.gd7 <- cnetplot(ego.gd7 ,circular = TRUE, colorEdge = TRUE)
cnet.toll9 <- cnetplot(ego.toll9 ,circular = TRUE, colorEdge = TRUE)
cnet.toll10b <- cnetplot(ego.toll10b ,circular = TRUE, colorEdge = TRUE)


goterms.gd7 = cowplot::plot_grid(bp.gd7, cnet.gd7, ncol=1,labels=LETTERS[1:2])
cowplot::save_plot(filename = "GOterm.gd7.pdf", plot = goterms.gd7 , base_height = 10)

goterms.toll9 = cowplot::plot_grid(bp.toll9, cnet.toll9, ncol=1,labels=LETTERS[1:2])
cowplot::save_plot(filename = "GOterm.toll9.pdf", plot = goterms.toll9 , base_height = 10)


goterms.toll10b = cowplot::plot_grid(bp.toll10b, cnet.toll10b, ncol=1,labels=LETTERS[1:2])
cowplot::save_plot(filename = "GOterm.toll10b.pdf", plot = goterms.toll10b , base_height = 10)


```



## Pause index analysis

### Getting all the data in order 

####  Load the gbc and ppc transcript data

```{r Reading the pp data and the gene body data, include=FALSE}
if (!require("RColorBrewer")) {
  install.packages("RColorBrewer")
}
library(RColorBrewer)
library(gplots)
gbcCountFile = paste(params$workingDir, params$dataDir,"counts.gbc.txt", sep = "/")

gbcData = read.table( file = gbcCountFile, header = TRUE, 
                      sep = "\t", quote = "",  
                      stringsAsFactors = FALSE)



gbcCountData = gbcData[,-1:-6]
TranscriptInfo = gbcData[,1:6]
colnames(TranscriptInfo) = c("geneID","Chr","Start","End","Strand","Length")
rownames(gbcCountData) = gbcData$Geneid

ppcCountFile = paste(params$workingDir, params$dataDir,"counts.ppc.txt", sep = "/")

ppcData = read.table( file = ppcCountFile, header = TRUE, 
                      sep = "\t", quote = "",  
                      stringsAsFactors = FALSE)

ppcCountData = ppcData[,-1:-6]
ppcTrnascriptInfo = ppcData[,1:6]
rownames(ppcCountData) = ppcData$Geneid


```

#### Handle the samples


```{r handle the samples, include=FALSE }

sampleInfo = data.frame(bamFile =  colnames(ppcCountData))
sampleInfo= sampleInfo %>% separate(col = bamFile,sep = "\\.",into =  c("dir", "mutant","sampleInfo"), remove = FALSE)
sampleInfo = sampleInfo %>% separate(col = sampleInfo,sep = "_",into =  c("irrelevant", "time","replicate"))

sampleInfo = sampleInfo %>% separate(col = mutant,sep = "_",into =  c("mutant"))

sampleInfo = sampleInfo %>% dplyr::select(bamFile,mutant,time,replicate)
sampleInfo = sampleInfo %>% unite("sampleName", mutant:time:replicate, remove = FALSE )
sampleInfo = sampleInfo %>% dplyr::select(bamFile,sampleName,mutant,time,replicate)


rownames(sampleInfo) = sampleInfo$bamFile
sampleInfo = sampleInfo[colnames(ppcCountData), ]
colnames(ppcCountData) = sampleInfo$sampleName
rownames(sampleInfo) = sampleInfo$sampleName

sampleInfo = sampleInfo %>% dplyr::select(sampleName,mutant,time,replicate,bamFile)
```

#### Merging the ppc and the gbc data
```{r Merging the ppc and the gbc data, include=FALSE }

ppcData = ppcCountData
ppcData$geneID = rownames(ppcData)
ppcData = gather(ppcData ,key = sampleName, value = ppcCount, -geneID)


trancriptData = inner_join(ppcData,sampleInfo)

head(gbcCountData)

gbcData = gbcCountData
gbcData = gbcData[,as.character(sampleInfo$bamFile)]
colnames(gbcData) = sampleInfo$sampleName
gbcData$geneID = rownames(gbcData)
gbcData = gather(gbcData ,key = sampleName, value = gbcCount, -geneID)
head(gbcData)



trancriptData = inner_join(trancriptData,gbcData)


trancriptData = inner_join(trancriptData,TranscriptInfo)



TranscriptCount = trancriptData %>% dplyr::select(geneID,  ppcCount,gbcCount,mutant,time) %>% 
  dplyr::group_by(geneID) %>%
  dplyr::summarise(ppcCountTot = sum(ppcCount), gbcCountTot = sum(gbcCount))

head(trancriptData)

SampleCount = trancriptData %>% dplyr::select(sampleName,  ppcCount,gbcCount,mutant,time) %>% 
  dplyr::group_by(sampleName) %>%
  dplyr::summarise(ppcSampleCountSum = sum(ppcCount), gbcSampleCountSum = sum(gbcCount))
```

#### Normalise the data using TPM
```{r calculate TPM }

trancriptData$ppcPK = trancriptData$ppcCount /(151/1000)
trancriptData$gbccPK = trancriptData$gbcCount /(trancriptData$Length/1000)

ScaleFactors = trancriptData %>% dplyr::select(sampleName,  ppcPK,gbccPK) %>%
  dplyr::group_by(sampleName) %>%
  dplyr::summarise(ppcScaleFactor = sum(ppcPK)/1000000, gbcScaleFactor = sum(gbccPK)/1000000)




trancriptData = inner_join(trancriptData,ScaleFactors )


trancriptData$ppcTPM = trancriptData$ppcPK / trancriptData$ppcScaleFactor
trancriptData$gbcTPM = trancriptData$gbccPK / trancriptData$gbcScaleFactor


ggplot(trancriptData, aes( x = log(gbcTPM+1), color = replicate)) +geom_density() +facet_grid(time~mutant ) 

ggplot(trancriptData, aes( x = log(ppcTPM+1), y = log(gbcTPM+1), color = replicate)) +geom_density_2d()+ geom_smooth(method = "lm") +facet_grid(time~mutant ) 





```


## QC pause index analysis


### Filter genes were either the ppc or the gbc expression is closer to noise

```{r analysing the data for }

trancriptData %>% dplyr::select(sampleName,  ppcTPM,gbcTPM) %>%
  dplyr::group_by(sampleName) %>%
  dplyr::summarise(ppcScaleFactor = sum(ppcTPM), gbcScaleFactor = sum(gbcTPM))


ggplot(trancriptData, aes(x = log10(gbcTPM+1), color = time, linetype = replicate)) +
  geom_density()+ facet_grid(.~mutant ) 


ggplot(trancriptData, aes(x = log10(ppcTPM+1), color = time, linetype = replicate)) +
  geom_density()+ facet_grid(.~mutant ) 


TranscriptCount.expressed  = trancriptData %>% 
  filter (log10(gbcTPM+1) > 1.5) %>% filter (log10(ppcTPM+1) > 1) %>% distinct(geneID)

trancriptData.QC = inner_join(trancriptData,TranscriptCount.expressed) 

ggplot(trancriptData.QC, aes(x = log10(gbcTPM+1), color = time, linetype = replicate)) +
  geom_density()+ facet_grid(.~mutant ) 

sum = trancriptData.QC$ppcTPM+trancriptData.QC$gbcTPM
sum[sum == 0] = 1
trancriptData.QC$PI =trancriptData.QC$ppcTPM/sum 



ggplot(trancriptData.QC, aes(x = PI, color = time, linetype = replicate)) +
  geom_density()+ facet_grid(.~mutant ) 


test = trancriptData.QC %>% dplyr::select(geneID,sampleName, PI) %>%
  spread(key = sampleName, value = PI)







testCor = cor( test[, 2:ncol(test)] ,method = "spearman")

heatmap(testCor, labCol = "")


ggplot(trancriptData.QC, aes( x = log(ppcTPM+1), y = log(gbcTPM+1), color = replicate)) +geom_density2d()+ geom_smooth(method = "lm") +facet_grid(time~mutant ) 



```





### add gene to transcript correlation

```{r  add  transcript to gene Information}

# Load in the fastq file that is generated as data from running multiqc on all samples. 
gtfFile = paste(params$workingDir, "annotations/Drosophila_melanogaster.BDGP6.28.99.gtf", sep = "/")
gtfInfo = read.table(file = gtfFile, header = F, sep = "\t", quote = "", stringsAsFactors = F)

gtfInfomRNA = gtfInfo %>% dplyr::filter(V3 == "transcript") 
gtfInfomRNA = gtfInfomRNA[grep(pattern = "gene_biotype \"protein_coding\"",x = gtfInfomRNA$V9 ),  ] 

gtfInfomRNA_DF= gtfInfomRNA %>% separate(col = V9,sep = ";",into =  c("geneID1", "transcriptID1"))
gtfInfomRNA_DF= gtfInfomRNA_DF %>% separate(col = geneID1,sep = "\"",into =  c("irrelevant", "geneID"))
gtfInfomRNA_DF= gtfInfomRNA_DF %>% separate(col = transcriptID1,sep = "\"",into =  c("irrelevant1", "transcriptID"))

gene2transcriptInfo  = gtfInfomRNA_DF %>% dplyr::select(geneID,transcriptID)



```



##PRO seq total analysis

### Merge CDS, GBC and PPC and Pause index data.

```{r getting known target genes}

# make a copy 

trancriptData.QC.pauseIndex = trancriptData.QC


# change geneID to trasncriptID as it should be. 
trancriptData.QC.pauseIndex = trancriptData.QC.pauseIndex %>%  dplyr::rename(transcriptID = geneID)

# add geneID to all transript IDs.
trancriptData.QC.pauseIndex = inner_join(trancriptData.QC.pauseIndex, gene2transcriptInfo)

trancriptData.QC.pauseIndex.summary = trancriptData.QC.pauseIndex %>%
  dplyr::group_by(geneID)%>%
  dplyr::summarise(gbcTPM = max(gbcTPM))


gene2transcriptInfo.higlyTranscribed = inner_join(trancriptData.QC.pauseIndex,
                                                  trancriptData.QC.pauseIndex.summary) %>%
  dplyr::select(geneID, transcriptID)



PROseqData.PI = inner_join(trancriptData.QC.pauseIndex,gene2transcriptInfo.higlyTranscribed) 

```



```{r Join the PROseq data.}

# Join CDS with gbc and ppc data
PROseqData = inner_join(PROseqData.CDS,PROseqData.PI )
PROseq.data = PROseqData

PROseq.data$GBClog  = log2(PROseq.data$gbcTPM) 
PROseq.data$PPClog  = log2(PROseq.data$ppcTPM) 
PROseq.data = PROseq.data %>% dplyr::select(-ppcPK, -gbccPK, -ppcScaleFactor, -gbcScaleFactor, -ppcTPM, -gbcTPM)

```


### Visualize the comparison of the diffferent classes 
```{r visualising data R}

PROseq.data$factor[PROseq.data$annotation2 != "DE"] = "Background"
PROseq.data$direction[PROseq.data$annotation2 != "DE"] = "None"

## PCA plot
fileNameFigCDS = paste( params$workingDir, params$proSeqDir, 
                        paste("PCA_loading_CDS.pdf", sep = "_") ,
                        sep = "/")

line1 = line %>% filter(factor != "Time")
pctable2 = pctable
pctable2$PC2 = pctable$PC2*0.007
pctable2$PC3 = pctable$PC3*0.007
pctable2$SampleFactor = pctable$factor



PROseq.data.mutants =  PROseq.data %>% filter(factor != "Time")
ggplot(PROseq.data.mutants, aes (x =PC2 , y =  PC3, color = factor, shape = direction)) + 
  geom_point()+
  scale_color_viridis(discrete = TRUE, option = "D") +
  scale_shape_manual(values=c(25, 20, 17))+ 
  ggtitle(label = "PROseq data CDS PCA loading values")+ 
  geom_line(data = line1, mapping = aes(x = PC2,y = PC3, color = factor, shape = "None"))+
  coord_cartesian(xlim = c(min(loadingsDF.all.filtered$PC2),
            max(loadingsDF.all.filtered$PC2)),
       ylim=  c(min(loadingsDF.all.filtered$PC3),
            max(loadingsDF.all.filtered$PC3))
       )+
   facet_wrap(annotation~. )
 


  ggsave(filename = fileNameFigCDS ,width = 15,height = 15)

fileNameFigCDS2 = paste( params$workingDir, params$proSeqDir, 
                        paste("PCA_loading_CDS2.pdf", sep = "_") ,
                        sep = "/")

PROseq.data.mutants =  PROseq.data %>% filter(factor != "Time")
ggplot(PROseq.data.mutants, aes (x =PC2 , y =  PC3, color = factor, shape = direction)) + 
  geom_point()+
  scale_color_viridis(discrete = TRUE, option = "D") +
  scale_shape_manual(values=c(25, 20, 17))+ 
  ggtitle(label = "PROseq data CDS PCA loading values")+ 
  geom_line(data = line1, mapping = aes(x = PC2,y = PC3, color = factor, shape = "None"))+
  coord_cartesian(xlim = c(min(loadingsDF.all.filtered$PC2),
            max(loadingsDF.all.filtered$PC2)),
       ylim=  c(min(loadingsDF.all.filtered$PC3),
            max(loadingsDF.all.filtered$PC3))
       )+
   facet_wrap(annotation~. )+
  geom_point(data = pctable2,  mapping = aes(x = PC2,y = PC3, fill = SampleFactor),shape=23, color="darkred", size=3)
 
  ggsave(filename = fileNameFigCDS2 ,width = 15,height = 15)


  
  
  
## GBC plot
fileNameFigGBC = paste( params$workingDir, params$proSeqDir, 
                        paste("Distribution_log_GBC.pdf", sep = "_") ,
                        sep = "/")
ggplot(PROseq.data, mapping = aes(x = factor , y =  GBClog , color= mutant )) + geom_boxplot()  + 
  facet_grid(time ~ direction)+ theme(axis.text.x = element_text(angle = 90, hjust = 1))

ggsave(filename = fileNameFigGBC ,width = 15,height = 15)


## PPC plot
fileNameFigPPC = paste( params$workingDir, params$proSeqDir, 
                        paste("Distribution_log_PPC.pdf", sep = "_") ,
                        sep = "/")

ggplot(PROseq.data, mapping = aes(x = factor , y =  PPClog , color= mutant  )) + geom_boxplot() + 
  facet_grid(time ~ direction)+ theme(axis.text.x = element_text(angle = 90, hjust = 1))

ggsave(filename = fileNameFigPPC )


fileNameFigCDS = paste( params$workingDir, params$proSeqDir, 
                        paste("Distribution_log_CDS.pdf", sep = "_") ,
                        sep = "/")


ggplot(PROseq.data, mapping = aes(x = factor , y =  rlog , color= mutant  )) + geom_boxplot() + 
  facet_grid(time ~ direction )+ theme(axis.text.x = element_text(angle = 90, hjust = 1))
ggsave(filename = fileNameFigCDS )


fileNameFigPI = paste( params$workingDir, params$proSeqDir, 
                       paste("Distribution_PI.pdf", sep = "_") ,
                       sep = "/")

ggplot(PROseq.data, mapping = aes(x = factor , y =  PI , color= mutant  )) + geom_boxplot() + 
  facet_grid(time ~ direction)+ theme(axis.text.x = element_text(angle = 90, hjust = 1))
ggsave(filename = fileNameFigPI )


PROseq.data %>% dplyr::group_by(geneID, annotation) %>% dplyr::summarise(nrOfAnnoations = n())
ggplot(PROseq.data, mapping = aes(x = annotation   )) + geom_bar() + 
  facet_grid(time ~ direction)+ theme(axis.text.x = element_text(angle = 90, hjust = 1))
ggsave(filename = fileNameFigPI )



```


### Compare the individual genes  for the different classes

```{r compare genes in classes}


PROseq.data.GOI = PROseq.data %>% 
  filter(factor != "Background" )%>% 
  filter(factor != "Known" )


PROseq.data.GOI$Mutant = recode(PROseq.data.GOI$Mutant, gd7 = "Gd7")


PROseq.data.GOI.geneInfo = PROseq.data.GOI %>% 
  dplyr::select(geneID, transcriptID, factor, direction,Mutant) %>% distinct() %>% arrange(factor, direction) %>% dplyr::rename ("Roshan" = Mutant)
  
write.table(x = PROseq.data.GOI.geneInfo, 
            file = paste( params$workingDir, params$proSeqDir, 
                          "selectedGenes.tsv" ,sep = "/"),
            quote = F, sep = "\t", col.names = TRUE, row.names = F)  



factors = as.character(unique(PROseq.data.GOI$factor))
directions = as.character(unique(PROseq.data.GOI$direction))


for(i in factors ){
  if(!dir.exists(paste( params$workingDir, params$proSeqDir, i ,
                        sep = "/"))){
    dir.create(paste( params$workingDir, params$proSeqDir, i ,
                      sep = "/"))
  }
  for(j in directions){
    
    PROseq.data.GOI.factor.dir = PROseq.data.GOI %>%
      filter( factor == i &  direction == j )
    if(nrow(PROseq.data.GOI.factor.dir) > 0){
      #CDS
      fileNameFigCDS = paste( params$workingDir, params$proSeqDir, i, 
                              paste(i,j, "expression_CDS.pdf", sep = "_") ,
                              sep = "/")
      ggplot(PROseq.data.GOI.factor.dir, aes(x = mutant, 
                                             y = rlog, 
                                             color = time)) + 
        geom_boxplot()+
        facet_wrap(SYMBOL~.)
      
      ggsave(filename = fileNameFigCDS , width = 20, height = 20)
      
      #PI
      fileNameFigPI = paste( params$workingDir, params$proSeqDir, i, 
                             paste(i,j, "pausingIndex.pdf", sep = "_") ,
                             sep = "/")
      ggplot(PROseq.data.GOI.factor.dir, aes(x = mutant,  y = PI, color = time)) + 
        geom_boxplot()+
        facet_wrap(SYMBOL~.) 
      ggsave(filename = fileNameFigPI , width = 20, height = 20)
      
      #GBC
      fileNameFigGBC = paste( params$workingDir, params$proSeqDir, i, 
                              paste(i,j, "expression_GBC.pdf", sep = "_") ,
                              sep = "/")
      ggplot(PROseq.data.GOI.factor.dir, 
             aes(x = mutant,  y = GBClog, color = time)) + 
        geom_boxplot()+
        facet_wrap(SYMBOL~.) 
      ggsave(filename = fileNameFigGBC , width = 20, height = 20)
      
      #PPC
      fileNameFigPPC = paste( params$workingDir, params$proSeqDir, i, 
                              paste(i,j, "expression_PPC.pdf", sep = "_") ,
                              sep = "/")
      ggplot(PROseq.data.GOI.factor.dir, 
             aes(x = mutant,  y = PPClog, color = time)) + 
        geom_boxplot()+
        facet_wrap(SYMBOL~.) 
      ggsave(filename = fileNameFigPPC , width = 20, height = 20)
    }
  }
}

```




### Visualize the individual genes 

```{r visualise all individual genes in interesting groups}


rerun = params$rerun
PROseq.data.interesting = PROseq.data 


factors = as.character(unique(PROseq.data.interesting$factor))
directions = as.character(unique(PROseq.data.interesting$direction))



for(i in factors ){
  if(!dir.exists(paste( params$workingDir, params$proSeqDir, i ,
                        sep = "/"))){
    dir.create(paste( params$workingDir, params$proSeqDir, i ,
                      sep = "/"))
  }
  for(j in directions){
    if(!dir.exists(paste( params$workingDir, params$proSeqDir, i ,j,
                          sep = "/"))){
      dir.create(paste( params$workingDir, params$proSeqDir, i ,j ,
                        sep = "/"))}
    
    
    PROseq.data.interesting.GOI.factor.dir = PROseq.data.interesting %>%
      filter( factor == i &  direction == j )
    if(nrow(PROseq.data.interesting.GOI.factor.dir) > 0){
      # write PC1 values 
      fileNameTable = paste( params$workingDir, params$proSeqDir, i,
                             paste(i,j, "summary.tab.tsv", sep = "_") ,
                             sep = "/")
      
      df = PROseq.data.interesting.GOI.factor.dir %>% dplyr::group_by(SYMBOL ) %>% 
        dplyr::summarise(PC1 = mean(PC1), PC2 = mean(PC2), 
                         PC3 = mean(PC3), distance = mean(distance)) %>% 
        dplyr::arrange(desc(distance) )
      
      
      write.table(x = df, file = fileNameTable,
                  quote = F, sep = "\t", col.names = T, row.names = F) 
      
      
      PROseq.data.interesting.GOI.factor.dir = 
        PROseq.data.interesting.GOI.factor.dir %>% 
        gather(rlog,PI,GBClog, PPClog, key = Expressions,value = score)
      
      fileNameTable = paste( params$workingDir, params$proSeqDir, i,j,
                             paste(i,j, "gene.expression.pdf", sep = "_") ,
                             sep = "/")
      if(!file.exists(fileNameTable) | rerun){
        
        plist = lapply(split(PROseq.data.interesting.GOI.factor.dir, 
                             PROseq.data.interesting.GOI.factor.dir$SYMBOL),
                       function(g) {
                         ggplot(g, mapping = aes(x = factor , y =  score , color= mutant )) +
                           geom_boxplot() +  scale_y_continuous(limits = c(0,NA))+
                           facet_grid(Expressions ~ time ,scales = "free_y")+
                           theme(axis.title.x=element_blank(),
                                 axis.title.y=element_blank(),
                                 axis.text.x=element_blank(),
                                 axis.ticks.x=element_blank())+
                           
                           ggtitle(unique(g$SYMBOL))
                         
                       })
        
        
        # Four separate single-page PDF files, each with six plots
        
        pdf(fileNameTable)
        for (k in seq(1, length(plist), 4)) {
          grid.arrange(grobs=plist[k:min((k+3),length(plist)) ], 
                       ncol=2, left="Measurements", bottom=class)
        }
        dev.off()
      }
    }
    
  }
  
}





```





### Print PROseq data to file 

```{r visualise all individual genes2}

PROseq.data.print  = PROseq.data

PROseq.data.print =  PROseq.data.print %>% dplyr::select(-Mutant)


fileNamePROseq = paste( params$workingDir, params$proSeqDir,  
                        params$proSeq.data,
                             sep = "/")
write.table(x = PROseq.data.print, file = fileNamePROseq,
                  quote = F, sep = "\t", col.names = T, row.names = F) 
      
PROseq.data.selected  = PROseq.data %>%filter (annotation2 == "DE") %>%dplyr::select(geneID,transcriptID, factor,direction) %>% distinct()



write.table(x = PROseq.data.selected, 
            file = paste( params$workingDir, params$proSeqDir, 
                          "selectedGenes.tsv" ,sep = "/"),
            quote = F, sep = "\t", col.names = TRUE, row.names = F)  



```










```{ }
system R failed: 256 at /proj/snic2019-30-14/private/perl/NRSA-v2/bin/pause_PROseq.pl line 426.

```