---
title: "R Notebook"
output: html_notebook

params:
  workingDir: /Users/johanreimegard/git/flyEmbryogenesis
  #workingDir: /Users/roshan/projects/tollTranscription/m_mannervik_19
  synonymsFileIn: information/fb_synonym_fb_2019_06.tsv  
  knownTargetsFile: information/pipe_Toll_known_Targets.txt
  EisenInfoFile: Eisen-ChIP_data-info.txt
  ZetlingerInfoFile: Zetlinger-ChIP-DATA-info.txt
  WieshausInfoFile: Weishaus-ATAC_data-info.txt
  ATACseqFile: ChIP_ATAC_sampleInfo.tab.txt
  dataDir: data/PROseq
  scriptDir: bash_scripts
  

---


The steps on uppmax to create the correct file 
# 

```{r}
library(tidyverse)

sampleInfoFile  = paste(params$workingDir,"information", params$ATACseqFile , sep = "/") 

sampleInfo = read.table(file = sampleInfoFile, header = T, sep = "\t", quote = "", stringsAsFactors = F)

ATACsampleInfo = sampleInfo %>% filter(Assaytype == "ATAC-Seq")
recode(ATACsampleInfo$SampleInfo, `Toll9/10`="Toll9_10"  )

ATACsampleInfo$SampleInfo = gsub(pattern = "9/10", replacement = "9_10", x = ATACsampleInfo$SampleInfo)
class(ATACsampleInfo)

```



## sort the bamfiles according to name
```{r sort the files according to name in bam ATAC seq }

nrOfCores = 8
bamFilesInfo = ATACsampleInfo %>% select(bamFiles)
bamFilesInfo$NameBamFiles = gsub(pattern = "sorted.bam","sorted.name.bam", bamFilesInfo$bamFiles)

bamFilesInfo$sorting = paste("if [ ! -f ", 
                             bamFilesInfo$NameBamFiles,
                             " ]; then samtools sort --threads ",nrOfCores," -n -o ",
                             bamFilesInfo$NameBamFiles,
                             " ",
                             bamFilesInfo$bamFiles,
                             "; fi",
                             sep = "")




fileConn<-file(paste(params$workingDir, params$scriptDir, "samtools.sort.name.ATACseq1.sh", sep = "/"))
writeLines(text = c("cd /proj/snic2019-30-14/nobackup"," ",
                    "module load bioinfo-tools",
                    "module load samtools"," "," "
                    ),con =  fileConn, sep = "\n")

write.table(x = bamFilesInfo$sorting, file = paste(params$workingDir, params$scriptDir, "samtools.sort.name.ATACseq1.sh", sep = "/"), append = TRUE, quote = F, sep = "\t", row.names = F,col.names = F)



```





##Create the correct genrichFile

```{r generich ATAC seq }





Genotypes = unique(ATACsampleInfo$Genotype)
timepoints = unique(ATACsampleInfo$Time)

blackListFile = "/proj/snic2019-30-14/private/genomes/dm6/dm6-blacklist.v2.bed"


GenerichCalls = data.frame(test = 1:100, GenrichCall = "test", sortBam = "test" ,stringsAsFactors = F)
counts = 1
for(i in 1:length(timepoints)){
  for (j in 1:length(Genotypes)){
    experiment = ATACsampleInfo %>%  filter( Genotype ==Genotypes[j] & Time == timepoints[i])
    if(nrow(experiment) > 0){
      bamFiles = experiment$bamFiles
      NameBamFiles = gsub(pattern = "sorted.bam","sorted.name.bam", bamFiles)

      GenerichCall = paste( "Genrich",
                            "-j -r  ",
                            paste( "-f",paste("Genrich",paste(Genotypes[j],
                                                              timepoints[i] ,
                                                              "genrich.encode.narrowPeak.log",sep = "_"), sep  ="/")),
                            paste( "-k",paste("Genrich",paste(Genotypes[j],
                                                              timepoints[i] ,
                                                              "genrich.encode.narrowPeak.pileups",sep = "_"), sep  ="/")),
                            paste( "-b",paste("Genrich",paste(Genotypes[j],
                                                              timepoints[i] ,
                                                              "genrich.encode.narrowPeak.intervals",sep = "_"), sep  ="/")),
                            paste( "-R",paste("Genrich",paste(Genotypes[j],
                                                              timepoints[i] ,
                                                              "genrich.encode.narrowPeak.pcrDuplicates",sep = "_"), sep  ="/")),
                            paste( "-E",blackListFile, sep = " "),
                            paste( "-t",paste(experiment$NameBamFiles, collapse = ","), sep  =" "),
                            paste( "-o",paste("Genrich",paste(Genotypes[j],
                                                              timepoints[i] ,
                                                              "genrich.encode.narrowPeak",sep = "_"), sep  ="/")),
                            sep = " ")
      GenerichCall
      GenerichCalls[counts ,2 ] = GenerichCall
      counts = counts + 1
    }
    
  }
}

GenerichCalls = GenerichCalls[1:counts-1,2]
info = data.frame(calls = GenerichCalls)

dim(info)



fileConn<-file(paste(params$workingDir, params$scriptDir, "Genrich.ATACseq1.sh", sep = "/"))
writeLines(text = c("cd /proj/snic2019-30-14/nobackup"," "),con =  fileConn, sep = "\n")
close(fileConn)

write.table(x = GenerichCalls, file = paste(params$workingDir, params$scriptDir, "Genrich.ATACseq1.sh", sep = "/"), append = TRUE, quote = F, sep = "\t", row.names = F,col.names = F)
writeLines(GenerichCalls[[i]], fileConn)
 



```


## Merge all the files into one


```{r merge the bedfiles into one }


load_module = "module load bioinfo-tools BEDTools/2.27.1"
correctDir =  "cd /proj/snic2019-30-14/nobackup/Genrich"


line1 = "cat gd7_*h_genrich.encode.narrowPeak  Toll10B_*h_genrich.encode.narrowPeak Toll9_10_*h_genrich.encode.narrowPeak|sort -k1,1 -k2,2n   >ATAC.genrich.encode.narrowPeak"
line2 = "bedtools merge -d 200 -i ATAC.genrich.encode.narrowPeak > ATAC.genrich.encode.narrowPeak.merged.bed"
line3 = 'ATAC.genrich.encode.narrowPeak.merged.bed |awk \'BEGIN{FS=OFS="\\t"; print "GeneID\\tChr\\tStart\\tEnd\\tStrand"}{print "Region"NR, $1, $2+1, $3, "."}\' >ATAC.genrich.encode.narrowPeak.merged.SAF'





fileConn<-file(paste(params$workingDir, params$scriptDir, "Genrich.ATACmerge.sh", sep = "/"))
writeLines(text = c(load_module," ",correctDir," ",line1,line2,line3),con =  fileConn, sep = "\n")
close(fileConn)




```

```{r run featureCount on chipseq data}


#      featureCounts -F SAF -a Genrich/K27me3.genrich.encode.narrowPeak.merged.SAF -o ChIPseq/K27me3.counts.regions.txt ChIP_seq_mapped_dm6/bam_dmel/gd7_K27me3_2_R1.dmel.bowtie2.sorted.bam  ChIP_seq_mapped_dm6/bam_dmel/gd7_K27me3_1_R1.dmel.bowtie2.sorted.bam




load_module = "module load bioinfo-tools subread/2.0.0"
correctDir =  "cd /proj/snic2019-30-14/nobackup"

featureCountCall = paste( "featureCounts  -F SAF -T 5",
                              paste( "-a",paste("Genrich",paste("ATAC.genrich.encode.narrowPeak.merged.SAF",sep = "."), sep  ="/"),  sep  =" "),
                              paste( "-o",paste("ChIPseq",paste("ATAC.peak.featureCount.count.tab.txt",sep = "."), sep  ="/"), sep  =" "),
                              paste(ATACsampleInfo$bamFiles, collapse = " ") )


fileConn<-file(paste(params$workingDir, params$scriptDir, "ATAC.featureCount.sh", sep = "/"))
writeLines(text = c(load_module," ",correctDir," ",featureCountCall),con =  fileConn, sep = "\n")
close(fileConn)



```


Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Cmd+Option+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Cmd+Shift+K* to preview the HTML file).
