---
title: "PROseq analysis"
author: "Joha
date: "10/25/2019"
output: html_document

params:
  workingDir: /Users/johanreimegard/git/flyEmbryogenesis
  RscriptsDir: R_Analysis/Rscripts  
  dataDir: data/PROseq
  ATACfile: data/ATACseq/ATAC_PPC.featureCount.count.tab.txt
  CBPfile: data/ChIPseq/CBP_PPC.featureCount.count.tab.txt
  RfunctionsDirectory: /Users/johanreimegard/GoogleDrive/git/RNAmappingPipeline/R
  proSeqDir: results/PROseq
  proSeq.data: PROseq.data.tsv
  proSeq.data.PPC: PROseq.data.PPC.tsv
  rerun: FALSE
---



```{r setup, include=FALSE}
source(paste(params$RfunctionsDirectory,"ExpressionAnalysisFunctions.R",sep = "/"))



library(tidyverse)
library(knitr)
#install.packages("kableExtra")
library(kableExtra)


library("viridis")     

knitr::opts_chunk$set(echo = TRUE)


library(RColorBrewer)
library(gplots)


```

# PROseq analysis


```{r Load the data, include=FALSE }
# install RColorBrewer if missing
#if (!require("RColorBrewer")) {
#  install.packages("RColorBrewer")
#}
PROseqFile = paste(params$workingDir, params$proSeqDir,params$proSeq.data, sep = "/")


PROseq = read.table( file = PROseqFile, header = TRUE, 
                     sep = "\t", quote = "",  
                     stringsAsFactors = FALSE)

head(PROseq)

metaInfo = PROseq.PI %>% dplyr::select( sampleName, mutant, time, replicate) %>% distinct() 
geneInfo = PROseq %>% dplyr::select(geneID, factor, direction, annotation, annotation2) %>% distinct()
```



# ATACseq analysis


```{r Load the data, include=FALSE }
ATACseqFile = paste(params$workingDir, params$ATACfile, sep = "/")


ATACseq = read.table( file = ATACseqFile, header = TRUE, 
                     sep = "\t", quote = "",  
                     stringsAsFactors = FALSE)

head(PROseq)

metaInfo = PROseq.PI %>% dplyr::select( sampleName, mutant, time, replicate) %>% distinct() 
geneInfo = PROseq %>% dplyr::select(geneID, factor, direction, annotation, annotation2) %>% distinct()
```




### Comparing mapping pattern between the samples


```{r Load the data, include=FALSE }
# install RColorBrewer if missing
#if (!require("RColorBrewer")) {
#  install.packages("RColorBrewer")
#}
library(RColorBrewer)
library(gplots)
PROseqFile = paste(params$workingDir, params$proSeqDir,params$proSeq.data, sep = "/")


PROseq = read.table( file = PROseqFile, header = TRUE, 
                     sep = "\t", quote = "",  
                     stringsAsFactors = FALSE)

head(PROseq)

metaInfo = PROseq.PI %>% dplyr::select( sampleName, mutant, time, replicate) %>% distinct() 
geneInfo = PROseq %>% dplyr::select(geneID, factor, direction, annotation, annotation2) %>% distinct()
```



#PI analysis


```{r Create meta data table for samples , include=FALSE}


PROseq.PI = PROseq %>% dplyr::select(geneID, factor, direction, sampleName, mutant, time, replicate, PI,   GBClog,    PPClog ) 
head(PROseq.PI)

PROseq.PI.data = PROseq.PI %>% 
  dplyr::select(geneID, sampleName, PI) %>%
  spread(key = sampleName,value = PI)
rownames(PROseq.PI.data)   = PROseq.PI.data$geneID
PROseq.PI.data = PROseq.PI.data[,-1]

PROseq.GBC.data = PROseq.PI %>% 
  dplyr::select(geneID, sampleName, GBClog) %>%
  spread(key = sampleName,value = GBClog)
rownames(PROseq.GBC.data)   = PROseq.GBC.data$geneID
PROseq.GBC.data = PROseq.GBC.data[,-1]
PROseq.GBC.data[PROseq.GBC.data == -Inf] = 0  

PROseq.PPC.data = PROseq.PI %>% 
  dplyr::select(geneID, sampleName, PPClog) %>%
  spread(key = sampleName,value = PPClog)
rownames(PROseq.PPC.data)   = PROseq.PPC.data$geneID
PROseq.PPC.data = PROseq.PPC.data[,-1]
PROseq.PPC.data[PROseq.PPC.data == -Inf] = 0  



```
## Differentiall expresssion analysis

###Normalising data and removinvg lowly expressed genes


### Visualising the data
```{r heatmap of samples}

#+ save1, include=FALSE
#png(paste(params$workingDirectory, params$proSeqDirectory,"SampleDistance.png", sep = "/"))
#plotSample2SampleDistance(assay(rld))
#dev.off()

#' 
#'
plotSample2SampleDistance(PROseq.PI.data)
plotSample2SampleDistance(PROseq.GBC.data)
plotSample2SampleDistance(PROseq.PPC.data)

#'  __Figure 1 Plotting sample to sample distance__ . 
#'  0 means that they are identical and 1 means that they are totally different.  
#'  The darker the blue the more similair. Also dendogram shows how similair they are. 
#' 
#' Samples do not cluster according to pre an post op. Most likely more due to difference in mapping (technical problem)
#' 
#' #### PCA analysis
#' Running PCA on the samples and plotting the different variables to see which of the parameterrs that fit the different components the best.
#' 
#' First checking how much the different PC contribute. 

```

###PCA analysis with normalised counts

```{r PCA analysis}

mir.pca <- prcomp(t(PROseq.PPC.data), center = TRUE, scale = FALSE) 
e.var = (mir.pca[['sdev']]^2 / sum(mir.pca[['sdev']]^2))
e.var = as.data.frame( e.var )
e.var$PC = as.factor(1:nrow(e.var)) 
e.var$Variance = e.var$e.var*100
e.var$CumulativeVariance = cumsum(e.var$Variance)
qplot(PC, Variance, data=e.var, geom = c("point")
      , ylim = c(0,max(e.var$Variance)+5) 
      ,ylab = "Variance (%)")

pctable = as.data.frame(mir.pca$x)[, 1:6]
pctable$sampleName = rownames(pctable)
pctable = inner_join(pctable, metaInfo)


test  = plotPCAplot(PCAinfo = pctable, n.comp = 5,
                    varianceInfo = e.var, 
                    colorComponent = "mutant", 
                    pchComponent = "time"
)
test


ggplot(data = pctable, mapping = aes(x = PC2, y =  PC3, color = mutant, shape = time)) + geom_point(aes(shape=time))

```

```{r get the axis } 

mean = pctable %>% dplyr::select(PC1,PC2,PC3,mutant) %>%
  dplyr::group_by(mutant) %>%
  dplyr::summarise(PC1 = mean(PC1) ,PC2 = mean(PC2), PC3 = mean(PC3) )

meantime = pctable %>% dplyr::select(PC1,PC2,PC3,time) %>%
  dplyr::group_by(time) %>%
  dplyr::summarise(PC1 = mean(PC1) ,PC2 = mean(PC2), PC3 = mean(PC3) )


origo = data.frame(PC1=0, PC2 = 0, PC3 = 0)


line.gd7 = rbind(mean[1,2:4],origo,-mean[1,2:4])
line.gd7$factor = "Gd7"
line.toll10b = rbind(mean[2,2:4],origo,-mean[2,2:4])
line.toll10b$factor = "Toll10b"
line.toll9 = rbind(mean[3,2:4],origo,-mean[3,2:4])
line.toll9$factor = "Toll9"
line.time = rbind(meantime[2,2:4],origo,meantime[1,2:4])
line.time$factor = "Time"


line = rbind(line.gd7,line.toll10b,line.toll9,line.time)


pctable$factor  = pctable$mutant
pctable$factor =  recode(pctable$factor, gd7 = "Gd7")
pctable$factor =  recode(pctable$factor, toll9 = "Toll9")
pctable$factor =  recode(pctable$factor, toll10b = "Toll10b")


fileNameFigCDSsamples = paste( params$workingDir, params$proSeqDir, 
                        paste("PCA_samples_with_decisionlines_CDS.pdf", sep = "_") ,
                        sep = "/")

ggplot(line, mapping = aes(x = PC2,y = PC3, color = factor))+ geom_line() + geom_point(data = pctable,  mapping = aes(x = PC2,y = PC3, color = factor, shape = time))+
  scale_color_viridis(discrete = TRUE, option = "D")
ggsave(fileNameFigCDSsamples)


ggplot(line, mapping = aes(x = PC1,y = PC2, color = factor))+ geom_line() + geom_point(data = pctable,  mapping = aes(x = PC1,y = PC2, color = factor, shape = time))+
  scale_color_viridis(discrete = TRUE, option = "D")
```



```{r analysing the loadings }

loadings = mir.pca$rotation[,c("PC1","PC2","PC3")] 
loadings1 = as.data.frame(t((t(loadings)-colMeans(loadings))/colSds(loadings)))
#loadings1 = as.data.frame(loadings)
loadings1$geneID = rownames(loadings1)

distance3d <- function(a,b){
  v1 = a - b
  
  distance = ((v1[[1]]^2 + v1[[2]]^2 + v1[[3]]^2)^0.5)
  return(distance)
}


dist3d <- function(point,b,c) {
  v1 <- b - c
  v2 <- point - b      
  v3 <- cross3d_prod(v1,v2)
  area <- sqrt(sum(v3*v3))/2
  d <- 2*area/sqrt(sum(v1*v1))
  return(d)
}

cross3d_prod <- function(v1,v2){
  v3 <- vector()
  v3[1] <- v1[2]*v2[3]-v1[3]*v2[2]
  v3[2] <- v1[3]*v2[1]-v1[1]*v2[3]
  v3[3] <- v1[1]*v2[2]-v1[2]*v2[1]
  return(v3)
}

for(j in unique(line$factor)){
  loadings1[[j]]=0
  for(i in 1:nrow(loadings1)){
  line1 = line %>% dplyr::filter(factor == j) %>% dplyr::select(PC1,PC2,PC3)
  loadings1[i,j] = dist3d(b =as.numeric(line1[1,]),
                            c = as.numeric(line1[3,]),
                            point = as.numeric(loadings1[i,1:3]))

  }

}


  




loadingsDF = gather(loadings1, key = PC, value = Loading, -geneID )


loadingsDFannotated = inner_join(loadingsDF, geneInfo) 

loadingsDFannotated = loadingsDFannotated %>% filter( PC %in% c("PC1","PC2","PC3")) 
ggplot(loadingsDFannotated, mapping = aes(x = Loading, color = PC )) + geom_density() + facet_grid(factor~.) 



```

#### Extracting interesting genes for the different timepoints and mutants based on the loading results. 


```{r clustering the genes based on the first three PCs}



loadingsDF.all = loadingsDF %>%  spread(key = PC, value = Loading)




tmp = loadingsDF.all
tmp$distance =  ((tmp$PC1^2 + tmp$PC2^2 + tmp$PC3^2)^0.5)
loadingsDF.all = tmp

loadingsDF.all.factorDistance =loadingsDF.all %>% gather(key = factor, value = distance2Factor,Time,Gd7,Toll10b,Toll9 )


loadingsDF.all.factorDistance$distanceFactor = sqrt(loadingsDF.all.factorDistance$distance^2 -
                                                   loadingsDF.all.factorDistance$distance2Factor^2)

loadingsDF.all.factorDistance$distanceFactor[is.na(loadingsDF.all.factorDistance$distanceFactor)] = 0


loadingsDF.all.factorDistance$direction = "Down"
line$Dir = rep(c("Up","origo","Down"), 4)


for(i in 1:nrow(loadingsDF.all.factorDistance)){
  factor2 = loadingsDF.all.factorDistance[i,]$factor
  UpPoint = line %>% dplyr::filter(factor == factor2 & Dir == "Up") %>% dplyr::select(PC1,PC2,PC3)
  DownPoint = line %>% dplyr::filter(factor == factor2 & Dir == "Down") %>% dplyr::select(PC1,PC2,PC3)
  point = loadingsDF.all.factorDistance[i,c("PC1","PC2","PC3")]
  up = distance3d(as.numeric( UpPoint),as.numeric(point))
  down = distance3d(as.numeric(DownPoint),as.numeric(point))
  
  if(up < down){
    loadingsDF.all.factorDistance[i,"direction"] = "Up"
  }else{
    loadingsDF.all.factorDistance[i,"distanceFactor"] = 
      -loadingsDF.all.factorDistance[i,"distanceFactor"]
  }
}


loadingsDF.all.factorDistance.max = loadingsDF.all.factorDistance %>% group_by(geneID) %>% summarise(max = max(abs(distanceFactor)))
 


loadingsDF.all.factorDistance$max=abs(loadingsDF.all.factorDistance$distanceFactor) 



loadingsDF.all.factorDistance.max2 = inner_join(loadingsDF.all.factorDistance,loadingsDF.all.factorDistance.max)

test  =colnames(loadingsDF.all.factorDistance)
test[6] = "PPC_factor"
test[9]= "PPC_direction"
test[10]= "max"

colnames(loadingsDF.all.factorDistance) = test
colnames(loadingsDF.all.factorDistance.max2) = test

unique(geneInfo$factor)
loadingsDF.all.factorDistance.annotated = inner_join(loadingsDF.all.factorDistance, geneInfo) 


ggplot(loadingsDF.all.factorDistance.annotated, mapping = aes(x = distanceFactor, color = PPC_factor )) + geom_density() + facet_grid(factor~direction) 


test = loadingsDF.all.factorDistance.annotated %>%  mutate(PPC_annotation = paste(PPC_factor, PPC_direction, sep = "_"), CDS_annotation = paste(factor, direction, sep = "_")) %>% filter(factor != "Background") %>% dplyr::select(geneID, distanceFactor, PPC_annotation,CDS_annotation)
test = inner_join(test,loadingsDF.all.factorDistance.max) %>% filter(abs(distanceFactor) == max)


ggplot(test, aes(x = CDS_annotation, fill = PPC_annotation)) + geom_bar()





```


### Print PROseq PPC data to file 

```{r }

loadingsDF.all.factorDistance.annotated$PPC_factor = paste("PPC", loadingsDF.all.factorDistance.annotated$PPC_factor, sep = "_")
PROseq.data.PPC = loadingsDF.all.factorDistance.annotated %>% dplyr::select(geneID, PPC_factor,distanceFactor ) %>% spread(key = PPC_factor, value = distanceFactor)
PPC_annotation = loadingsDF.all.factorDistance.max2 %>% dplyr::select(geneID, PPC_factor, PPC_direction) %>% distinct()

PROseq.data.PPC.annotated = inner_join(PROseq.data.PPC,PPC_annotation)



fileNamePROseqPPC = paste( params$workingDir, params$proSeqDir,  
                        params$proSeq.data.PPC,
                             sep = "/")
write.table(x = PROseq.data.PPC.annotated, file = fileNamePROseqPPC,
                  quote = F, sep = "\t", col.names = T, row.names = F) 
      



```










```{ }
system R failed: 256 at /proj/snic2019-30-14/private/perl/NRSA-v2/bin/pause_PROseq.pl line 426.

```