---
title: "Joint peak analysis analysis"
author: "Johan"
date: "10/25/2019"
output: html_document

params:
  workingDir: /Users/johanreimegard/git/Applied_Bioinformatics_Group_4
  proSeq.data.CBP: PROseq.CBP.peak.tsv
  proSeqDir: results/PROseq
  ATACSeqDir: results/ATACseq
  ATACSeq.data: ATACseq.data.raw.tsv
  chipSeqDir: results/chipSeq
  enhancerSeqDir: results/enhancers
  k27ac.data: K27ac.ChIPseq.peak.pattern.raw.tsv
  CBP.data: CBP-CBP_dm6.ChIPseq.peak.pattern.raw.tsv
  GD7enhancer: results/enhancers/h3k27ac_gd7_csaw_intersect.bed
  Toll10benhancer: results/enhancers/h3k27ac_Toll10B_csaw_intersect.bed
  Toll9enhancer: results/enhancers/h3k27ac_Tollrm910_csaw_intersect.bed
  
  
---




```{r setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)

library(gplots)

#install.packages("kableExtra")
library(kableExtra)
library("viridis")     
library(GenomicRanges)
library(patchwork)
library(umap)


```


## Annotation

### K27ac ChiPseq data identified enhancers. 

```{r enhancer info, include=FALSE}

# Load in the fastq file that is generated as data from running multiqc on all samples. 
GD7enhancersFile = paste(params$workingDir, params$GD7enhancer, sep = "/")
GD7enhancers = read.table(file = GD7enhancersFile, header = F, sep = "\t", quote = "", stringsAsFactors = F)
GD7enhancers = GD7enhancers[,1:3]
colnames(GD7enhancers) = c("Chr","Start","End")
GD7enhancers$Genotype = "Gd7"

Toll10benhancersFile = paste(params$workingDir, params$Toll10benhancer, sep = "/")
Toll10benhancers = read.table(file = Toll10benhancersFile, header = F, sep = "\t", quote = "", stringsAsFactors = F)
Toll10benhancers = Toll10benhancers[,1:3]
colnames(Toll10benhancers) = c("Chr","Start","End")
Toll10benhancers$Genotype = "Toll10b"


Toll9enhancersFile = paste(params$workingDir, params$Toll9enhancer, sep = "/")
Toll9enhancers = read.table(file = Toll9enhancersFile, header = F, sep = "\t", quote = "", stringsAsFactors = F)
Toll9enhancers = Toll9enhancers[,1:3]
colnames(Toll9enhancers) = c("Chr","Start","End")
Toll9enhancers$Genotype = "Toll9"

Enhancers = rbind(GD7enhancers,Toll10benhancers,Toll9enhancers)
Enhancers$Strand = "."
Enhancers$Chr = paste("chr",Enhancers$Chr, sep = "")
Enhancers$enhancerID = paste("region",1:nrow(Enhancers))
rownames(Enhancers) = Enhancers$enhancerID

EnhancersInfo = Enhancers %>% 
  dplyr::select(Genotype,enhancerID) %>%  
  dplyr::rename(enhancerGenotype = Genotype)
 

EnhancersPeakData = makeGRangesFromDataFrame(Enhancers,
                                      seqnames.field = "Chr", 
                                      start.field = "Start", 
                                      end.field = "End", 
                                      strand.field = "Strand", 
                                      keep.extra.columns = TRUE)
names(EnhancersPeakData) = EnhancersPeakData$enhancerID

```

### In situ

```{r insitu info, include=FALSE}

# Load in the fastq file that is generated as data from running multiqc on all samples. 

inSituFile = paste(params$workingDir, "results/inSitu/inSitu.Stark.dm6.tsv", sep = "/")
inSituDF = read.table(file = inSituFile, header = T, sep = "\t", quote = "", stringsAsFactors = F)

inSituDF2 = inSituDF %>% dplyr::select(seqnames, start, end, stg4_6) %>% 
  separate_rows( stg4_6, sep = "\\|") %>%
  separate(stg4_6, into = c("Location","score"), sep = ";")

inSituDF2 = inSituDF2 %>% replace_na(list(score = 0, Location = "None")) 
table(inSituDF2$Location)
inSituDF2$insituID = paste("region",1:nrow(inSituDF2))


inSituInfo = inSituDF2 %>%
  dplyr::select(insituID ,Location,score) %>%  
  dplyr::rename(inSituLocation = Location, inSituScore = score)


inSituData = makeGRangesFromDataFrame(inSituDF2,
                                      seqnames.field = c("Chr","seqnames"), 
                                      start.field = c("Start","start"), 
                                      end.field = c("End","end"), 
                                      strand.field = "Strand", 
                                      keep.extra.columns = TRUE)



names(inSituData) = inSituDF2$insituID
```




## CBP peak data 


### PROseq

```{r  PROseq CBP peaks data, include=FALSE}

CBP_file = paste( params$workingDir, 
              params$proSeqDir, 
              params$proSeq.data.CBP,
                          sep = "/")

PROseqInfo.CBP = read.table(file = CBP_file ,
                         quote = "", sep = "\t", 
                         header = TRUE, stringsAsFactors = F) 

PROseqInfo.CBP = PROseqInfo.CBP %>% 
  dplyr::select(geneID, Gd7, Toll10b,Toll9, Time) %>%
  dplyr::rename("PRO_Gd7" = Gd7, "PRO_Toll10b" = Toll10b, 
                "PRO_Toll9" = Toll9, "PRO_Time" = Time, Geneid = geneID)



# ChIPseq analysisdi
```

### ATAC seq

```{r ATACseq peak analysis, include=FALSE}



ATACseqInfo = read.table( file = paste( params$workingDir, params$ATACSeqDir, params$ATACSeq.data,
                          sep = "/"),
                         quote = "", sep = "\t", 
                         header = TRUE, stringsAsFactors = F) 

ATACseqInfo = ATACseqInfo %>% dplyr::rename("CBP_Start" = Start, "CBP_End" = End ,
                                    "CBP_factor" = factor, "CBP_direction" = direction,
                                    ) 


CBPPeakData = makeGRangesFromDataFrame(ATACseqInfo,
                                      seqnames.field = "Chr", 
                                      start.field = "CBP_Start", 
                                      end.field = "CBP_End", 
                                      strand.field = "Strand")
names(CBPPeakData) = ATACseqInfo$Geneid




ATACseqInfo = ATACseqInfo %>%  
  dplyr::select(Geneid, Gd7, Toll10b,Toll9, Time) %>%
  dplyr::rename("ATAC_Gd7" = Gd7, "ATAC_Toll10b" = Toll10b, 
                "ATAC_Toll9" = Toll9, "ATAC_Time" = Time)





```





### CBP seq

```{r CBP peak analysis , include=FALSE}


CBPInfo =  read.table( file = paste( params$workingDir, params$chipSeqDir,
                                        params$CBP.data,
                          sep = "/"),
                         quote = "", sep = "\t", 
                         header = TRUE, stringsAsFactors = F) 

CBPInfo = CBPInfo %>% 
  dplyr::select(Geneid, Gd7, Toll10b,Toll9) %>%
  dplyr::rename("CBP_Gd7" = Gd7, "CBP_Toll10b" = Toll10b, 
                "CBP_Toll9" = Toll9)


```


## K27ac peak  data


```{r K27ac peak analysis, include=FALSE}





K27acInfo =  read.table( file = paste( params$workingDir, params$chipSeqDir,
                                        params$k27ac.data,
                          sep = "/"),
                         quote = "", sep = "\t", 
                         header = TRUE, stringsAsFactors = F) 



K27acInfo = K27acInfo %>% dplyr::rename("K27ac_Geneid" = Geneid, "K27ac_Start" = Start, "K27ac_End" = End ,
                                    "K27ac_factor" = factor, "K27ac_direction" = direction)



K27acPeakData = makeGRangesFromDataFrame(K27acInfo,
                                      seqnames.field = "Chr", 
                                      start.field = "K27ac_Start", 
                                      end.field = "K27ac_End", 
                                      strand.field = "Strand" 
                                      )
names(K27acPeakData) = K27acInfo$K27ac_Geneid


K27acInfo =  K27acInfo %>% 
  dplyr::select(K27ac_Geneid, Gd7, Toll10b,Toll9) %>%
  dplyr::rename("K27ac_Gd7" = Gd7, "K27ac_Toll10b" = Toll10b, 
                "K27ac_Toll9" = Toll9) 

```


## Merge the data

###  Join the CBP peak data

```{r CBP peak data joining , include=FALSE}


CBPpeak.ATAC.PRO  = inner_join(ATACseqInfo, PROseqInfo.CBP)
CBPpeak.ATAC.PRO.CBP  = inner_join(CBPpeak.ATAC.PRO, CBPInfo)



```




### Join the CBP peak data with the H3K27ac data
```{r CBP peak and K27ac peak data joining , include=FALSE}



overlap_ATAC_K27ac = data.frame(findOverlaps(CBPPeakData,K27acPeakData, maxgap = 0))
length(unique(overlap_ATAC_K27ac$subjectHits))
overlap_ATAC_K27ac$K27ac_Geneid = names(K27acPeakData)[overlap_ATAC_K27ac$subjectHits]
overlap_ATAC_K27ac$Geneid = names(CBPPeakData)[overlap_ATAC_K27ac$queryHits]
overlap_ATAC_K27ac = overlap_ATAC_K27ac %>% dplyr::select(K27ac_Geneid,Geneid )


CBPpeak.ATAC.PRO.CBP  = inner_join(CBPpeak.ATAC.PRO.CBP,  overlap_ATAC_K27ac)
CBPpeak.ATAC.PRO.CBP.K27ac =  inner_join(CBPpeak.ATAC.PRO.CBP,  K27acInfo)



```



### Join the annotation
```{r Adding annotation from enhancers and inSitu, include=FALSE}

overlap_CBP_enhancers = data.frame(findOverlaps(CBPPeakData,EnhancersPeakData, maxgap = 0))
length(unique(overlap_CBP_enhancers$subjectHits))

overlap_CBP_enhancers$enhancerID = EnhancersPeakData$enhancerID[overlap_CBP_enhancers$subjectHits]
overlap_CBP_enhancers$Geneid = names(CBPPeakData)[overlap_CBP_enhancers$queryHits]
overlap_CBP_enhancers = overlap_CBP_enhancers %>% dplyr::select(enhancerID,Geneid )


CBPpeak.ATAC.PRO.CBP.K27ac.enhancer = left_join(CBPpeak.ATAC.PRO.CBP.K27ac, overlap_CBP_enhancers)

CBPpeak.ATAC.PRO.CBP.K27ac.enhancer = left_join(CBPpeak.ATAC.PRO.CBP.K27ac.enhancer, EnhancersInfo)



overlap_CBP_inSitu = data.frame(findOverlaps(CBPPeakData,inSituData, maxgap = 0))
overlap_CBP_inSitu$insituID = inSituData$insituID[overlap_CBP_inSitu$subjectHits]
overlap_CBP_inSitu$Geneid = names(CBPPeakData)[overlap_CBP_inSitu$queryHits]
overlap_CBP_inSitu = overlap_CBP_inSitu %>% dplyr::select(insituID,Geneid )


CBPpeak.ATAC.PRO.CBP.K27ac.enhancer.inSitu = left_join(CBPpeak.ATAC.PRO.CBP.K27ac.enhancer, overlap_CBP_inSitu)

CBPpeak.ATAC.PRO.CBP.K27ac.enhancer.inSitu = left_join(CBPpeak.ATAC.PRO.CBP.K27ac.enhancer.inSitu, inSituInfo)





```





```{r adding annotation as 0 and 1, include=FALSE}


EnhancerInfo = CBPpeak.ATAC.PRO.CBP.K27ac.enhancer.inSitu
EnhancerInfo$Geneid_unique = make.names(EnhancerInfo$Geneid, unique = TRUE)



EnhancerInfo$enhancer=1
EnhancerInfo$enhancer[is.na(EnhancerInfo$enhancerGenotype)]  = 0 

EnhancerInfo$inSitu=0
EnhancerInfo$inSitu[EnhancerInfo$inSituLocation %in% c("dorsal_ectoderm_AISN_broad", "mesoderm_AISN", "ventral_ectoderm_AISN")] = 1

PE =  EnhancerInfo%>% filter(enhancer == 1) %>% ggplot(aes(PRO_Toll9, PRO_Toll10b, color = enhancerGenotype)) + geom_point()+ geom_density_2d(EnhancerInfo,mapping = aes(PRO_Toll9, PRO_Toll10b), color= "grey")
PI =  EnhancerInfo %>% filter(inSitu == 1) %>% ggplot(aes(PRO_Toll9, PRO_Toll10b, color = inSituLocation )) + geom_point()+ geom_density_2d(EnhancerInfo,mapping = aes(PRO_Toll9, PRO_Toll10b), color= "grey")
AE =  EnhancerInfo%>% filter(enhancer == 1) %>% ggplot(aes(ATAC_Toll9, ATAC_Toll10b, color = enhancerGenotype)) + geom_point()+ geom_density_2d(EnhancerInfo,mapping = aes(ATAC_Toll9, ATAC_Toll10b), color= "grey")
AI =  EnhancerInfo %>% filter(inSitu == 1) %>% ggplot(aes(ATAC_Toll9, ATAC_Toll10b, color = inSituLocation )) + geom_point()+ geom_density_2d(EnhancerInfo,mapping = aes(ATAC_Toll9, ATAC_Toll10b), color= "grey")
CE =  EnhancerInfo%>% filter(enhancer == 1) %>% ggplot(aes(CBP_Toll9, CBP_Toll10b, color = enhancerGenotype)) + geom_point()+ geom_density_2d(EnhancerInfo,mapping = aes(CBP_Toll9, CBP_Toll10b), color= "grey") 
CI =  EnhancerInfo %>% filter(inSitu == 1) %>% ggplot(aes(CBP_Toll9, CBP_Toll10b, color = inSituLocation )) + geom_point()+ geom_density_2d(EnhancerInfo,mapping = aes(CBP_Toll9, CBP_Toll10b), color= "grey")
KE =  EnhancerInfo%>% filter(enhancer == 1) %>% ggplot(aes(K27ac_Toll9, K27ac_Toll10b, color = enhancerGenotype)) + geom_point()+ geom_density_2d(EnhancerInfo,mapping = aes(K27ac_Toll9, K27ac_Toll10b), color= "grey")
KI =  EnhancerInfo %>% filter(inSitu == 1) %>% ggplot(aes(K27ac_Toll9, K27ac_Toll10b, color = inSituLocation )) + geom_point()+ geom_density_2d(EnhancerInfo,mapping = aes(K27ac_Toll9, K27ac_Toll10b), color= "grey")

```










```{r ROC info function, include=FALSE}
library(ROCit)




getROCinfo <- function(enhancerData, scoreColumn = "distance2", classColumn = "enhancer" ,  FPR = 0.1){
  ROCit_obj <- rocit(score=enhancerData[[scoreColumn]],
                     class=enhancerData[[classColumn]],)
  
  ROCplotInfo = data.frame(score = scoreColumn, classifier = classColumn, TPR = ROCit_obj$TPR, FPR = ROCit_obj$FPR)
  
  InfoPlot = plot(ROCit_obj)
  ksInfo = ksplot(ROCit_obj)

  
  Youden  = data.frame(score = scoreColumn, classifier = classColumn,  
                       TPR = InfoPlot$`optimal Youden Index point`[3] ,
                       FPR = InfoPlot$`optimal Youden Index point`[2],
                       YoudenCutoff = InfoPlot$`optimal Youden Index point`[4],
                       ksCutoff = ksInfo$`KS Cutoff`,
                       AUC = ROCit_obj$AUC,
                       Neg_count = ROCit_obj$neg_count,
                       hardCutOff = ROCit_obj$Cutoff[sum(ROCit_obj$FPR <FPR)],
                       hardTPR = ROCit_obj$TPR[sum(ROCit_obj$FPR < FPR)], 
                       FalsePositive = sum(ROCit_obj$FPR < FPR)
  )
  ROCInfo = list(ROCplotInfo=ROCplotInfo, Youden =  Youden, ROCplot = InfoPlot, KSinfo = ksInfo) 
  return (ROCInfo)

}




```











```{r PCA analysis, include=FALSE}

allEnhancerInfo2 = EnhancerInfo %>% dplyr::select(-Geneid,-enhancerID,-enhancerGenotype,-insituID,-inSituLocation,-inSituScore,-Geneid_unique,-enhancer,-inSitu,-K27ac_Geneid, PRO_Time, ATAC_Time)
rownames(allEnhancerInfo2) = EnhancerInfo$Geneid_unique


#test = apply(allEnhancerInfo2,1,var)
#test = test[order(test, decreasing = T)]
#test.top = test[1:500]
#allEnhancerInfo2 = allEnhancerInfo2[names(test.top),]

mir.pca <- prcomp(allEnhancerInfo2, center = TRUE, scale = FALSE) 
e.var = (mir.pca[['sdev']]^2 / sum(mir.pca[['sdev']]^2))
e.var = as.data.frame( e.var )
e.var$PC = as.factor(1:nrow(e.var)) 
e.var$Variance = e.var$e.var*100
e.var$CumulativeVariance = cumsum(e.var$Variance)
qplot(PC, Variance, data=e.var, geom = c("point")
      , ylim = c(0,max(e.var$Variance)+5) 
      ,ylab = "Variance (%)")

pctable = as.data.frame(mir.pca$x)[, 1:5]




pctable$Geneid_unique = rownames(pctable)




pctable$Geneid_unique = rownames(pctable)

AllData = inner_join(EnhancerInfo,pctable)


AllData = AllData %>% mutate(distance5 = (PC1^2+PC2^2+PC3^2+PC4^2++PC5^2)^0.5)
AllData = AllData %>% mutate(distance4 = (PC1^2+PC2^2+PC3^2+PC4^2)^0.5)
AllData = AllData %>% mutate(distance3 = (PC1^2+PC2^2+PC3^2)^0.5)
AllData = AllData %>% mutate(distance2 = (PC1^2+PC2^2)^0.5)
AllData = AllData %>% mutate(distance1 = abs(PC1))

FE =  AllData %>% filter(enhancer == 1) %>% ggplot(aes(PC1, PC2, color = enhancerGenotype )) + geom_point()+ geom_density_2d(AllData,mapping = aes(PC1, PC2), color= "grey")
FI =  AllData %>% filter(inSitu == 1) %>% ggplot(aes(PC1, PC2, color = inSituLocation )) + geom_point()+ geom_density_2d(AllData,mapping = aes(PC1, PC2), color= "grey")
```

```{r plot this steps }

PE+PI+ plot_layout(ncol = 1)
AE+AI+ plot_layout(ncol = 1)
CE+CI+ plot_layout(ncol = 1)
KE+KI+ plot_layout(ncol = 1)
FE + FI+ plot_layout(ncol = 1)

PE+PI+AE+AI+CE+CI+KE+KI+FE+FI +plot_layout(ncol = 2)


```
### Assesment and selection

```{r roc analysis  , include=FALSE}
enhancerData = AllData

d1ROCi = getROCinfo(AllData,scoreColumn = "distance1", classColumn = "inSitu")
d2ROCi = getROCinfo(AllData,scoreColumn = "distance2", classColumn = "inSitu")
d3ROCi = getROCinfo(AllData,scoreColumn = "distance3", classColumn = "inSitu")
d4ROCi = getROCinfo(AllData,scoreColumn = "distance4", classColumn = "inSitu")
d5ROCi = getROCinfo(AllData,scoreColumn = "distance5", classColumn = "inSitu")
d1ROCe = getROCinfo(AllData,scoreColumn = "distance1", classColumn = "enhancer")
d2ROCe = getROCinfo(AllData,scoreColumn = "distance2", classColumn = "enhancer")
d3ROCe = getROCinfo(AllData,scoreColumn = "distance3", classColumn = "enhancer")
d4ROCe = getROCinfo(AllData,scoreColumn = "distance4", classColumn = "enhancer")
d5ROCe = getROCinfo(AllData,scoreColumn = "distance5", classColumn = "enhancer")



ROCplotInfo = rbind (d1ROCi$ROCplotInfo, d2ROCi$ROCplotInfo, d3ROCi$ROCplotInfo,d4ROCi$ROCplotInfo, d5ROCi$ROCplotInfo,
                     d1ROCe$ROCplotInfo, d2ROCe$ROCplotInfo, d3ROCe$ROCplotInfo,d4ROCe$ROCplotInfo, d5ROCe$ROCplotInfo)
ROCplotSummary = rbind(d1ROCi$Youden, d2ROCi$Youden, d3ROCi$Youden,d4ROCi$Youden, d5ROCi$Youden,
                     d1ROCe$Youden, d2ROCe$Youden, d3ROCe$Youden,d4ROCe$Youden, d5ROCe$Youden)


```

```{r plot ROC}

ROCplotSummary

qplot(PC, Variance, data=e.var, geom = c("point")
      , ylim = c(0,max(e.var$Variance)+5) 
      ,ylab = "Variance (%)")


ggplot(ROCplotInfo, mapping =  aes(x = FPR, y = TPR, color = score))+ 
  geom_line()+ 
  geom_abline()+ 
  geom_point(data = ROCplotSummary, mapping = aes(y = TPR, x = FPR, color = score ), size = 4) +
    facet_grid(.~classifier)

  










```
# Enhancer selection


Based on ROC statistics  it is best if we include all five PC  

```{r plot classification }


scores = ROCplotSummary %>% dplyr::filter(classifier == "enhancer" & score == "distance5")

AllData.enhancers = AllData %>% filter(distance5 > scores$YoudenCutoff)

AllData.enhancers = AllData.enhancers %>% replace_na(list(enhancerID = "None", enhancerGenotype = "None",insituID = "None",inSituLocation = "None", inSituScore = -1 )) 


peakInfo = data.frame(CBPPeakData)
peakInfo$Geneid = names(CBPPeakData)

AllData.enhancers = inner_join(peakInfo, AllData.enhancers)

fileName = paste(params$workingDir, params$enhancerSeqDir,"enhancers.PCA.5dim.tsv", sep = "/")

write.table(x = AllData.enhancers, file = fileName, sep = "\t", quote = F, row.names = F, col.names = T )



```







